<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Usage documentation &mdash; BioBlend 0.5.3-dev documentation</title>
    
    <link rel="stylesheet" href="../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.5.3-dev',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <link rel="top" title="BioBlend 0.5.3-dev documentation" href="../../index.html" />
    <link rel="next" title="Configuration documents for BioBlend" href="../lib_config.html" />
    <link rel="prev" title="API documentation for interacting with Galaxy" href="all.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../lib_config.html" title="Configuration documents for BioBlend"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="all.html" title="API documentation for interacting with Galaxy"
             accesskey="P">previous</a> |</li>
        <li><a href="../../index.html">BioBlend 0.5.3-dev documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="usage-documentation">
<h1>Usage documentation<a class="headerlink" href="#usage-documentation" title="Permalink to this headline">¶</a></h1>
<p>This page describes some sample use cases for the Galaxy API and provides
examples for these API calls.
In addition to this page, there are functional examples of complete scripts in the
<code class="docutils literal"><span class="pre">docs/examples</span></code> directory of the BioBlend source code repository.</p>
<div class="section" id="connect-to-a-galaxy-server">
<h2>Connect to a Galaxy server<a class="headerlink" href="#connect-to-a-galaxy-server" title="Permalink to this headline">¶</a></h2>
<p>To connect to a running Galaxy server, you will need an account on that Galaxy instance and an API key for the account. Instructions on getting an API key can be found at <a class="reference external" href="http://wiki.galaxyproject.org/Learn/API">http://wiki.galaxyproject.org/Learn/API</a> .</p>
<p>To open a connection call:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">bioblend.galaxy</span> <span class="kn">import</span> <span class="n">GalaxyInstance</span>

<span class="n">gi</span> <span class="o">=</span> <span class="n">GalaxyInstance</span><span class="p">(</span><span class="n">url</span><span class="o">=</span><span class="s">&#39;http://example.galaxy.url&#39;</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="s">&#39;your-API-key&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>We now have a <code class="docutils literal"><span class="pre">GalaxyInstance</span></code> object which allows us to interact with the Galaxy server under our account, and access our data. If the account is a Galaxy admin account we also will be able to use this connection to carry out admin actions.</p>
</div>
<div class="section" id="view-histories-and-datasets">
<span id="id1"></span><h2>View Histories and Datasets<a class="headerlink" href="#view-histories-and-datasets" title="Permalink to this headline">¶</a></h2>
<p>Methods for accessing histories and datasets are grouped under <code class="docutils literal"><span class="pre">GalaxyInstance.histories.*</span></code> and <code class="docutils literal"><span class="pre">GalaxyInstance.datasets.*</span></code> respectively.</p>
<p>To get information on the Histories currently in your account, call:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">histories</span><span class="o">.</span><span class="n">get_histories</span><span class="p">()</span>
<span class="go">[{u&#39;id&#39;: u&#39;f3c2b0f3ecac9f02&#39;,</span>
<span class="go">  u&#39;name&#39;: u&#39;RNAseq_DGE_BASIC_Prep&#39;,</span>
<span class="go">  u&#39;url&#39;: u&#39;/api/histories/f3c2b0f3ecac9f02&#39;},</span>
<span class="go"> {u&#39;id&#39;: u&#39;8a91dcf1866a80c2&#39;,</span>
<span class="go">  u&#39;name&#39;: u&#39;June demo&#39;,</span>
<span class="go">  u&#39;url&#39;: u&#39;/api/histories/8a91dcf1866a80c2&#39;}]</span>
</pre></div>
</div>
<p>This returns a list of dictionaries containing basic metadata, including the id and name of each History. In this case, we have two existing Histories in our account, &#8216;RNAseq_DGE_BASIC_Prep&#8217; and &#8216;June demo&#8217;. To get more detailed information about a History we can pass its id to the <code class="docutils literal"><span class="pre">show_history</span></code> method:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">histories</span><span class="o">.</span><span class="n">show_history</span><span class="p">(</span><span class="s">&#39;f3c2b0f3ecac9f02&#39;</span><span class="p">,</span> <span class="n">contents</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="go">{u&#39;annotation&#39;: u&#39;&#39;,</span>
<span class="go"> u&#39;contents_url&#39;: u&#39;/api/histories/f3c2b0f3ecac9f02/contents&#39;,</span>
<span class="go"> u&#39;id&#39;: u&#39;f3c2b0f3ecac9f02&#39;,</span>
<span class="go"> u&#39;name&#39;: u&#39;RNAseq_DGE_BASIC_Prep&#39;,</span>
<span class="go"> u&#39;nice_size&#39;: u&#39;93.5 MB&#39;,</span>
<span class="go"> u&#39;state&#39;: u&#39;ok&#39;,</span>
<span class="go"> u&#39;state_details&#39;: {u&#39;discarded&#39;: 0,</span>
<span class="go">      u&#39;empty&#39;: 0,</span>
<span class="go">      u&#39;error&#39;: 0,</span>
<span class="go">      u&#39;failed_metadata&#39;: 0,</span>
<span class="go">      u&#39;new&#39;: 0,</span>
<span class="go">      u&#39;ok&#39;: 7,</span>
<span class="go">      u&#39;paused&#39;: 0,</span>
<span class="go">      u&#39;queued&#39;: 0,</span>
<span class="go">      u&#39;running&#39;: 0,</span>
<span class="go">      u&#39;setting_metadata&#39;: 0,</span>
<span class="go">      u&#39;upload&#39;: 0 },</span>
<span class="go"> u&#39;state_ids&#39;: {u&#39;discarded&#39;: [],</span>
<span class="go">      u&#39;empty&#39;: [],</span>
<span class="go">      u&#39;error&#39;: [],</span>
<span class="go">      u&#39;failed_metadata&#39;: [],</span>
<span class="go">      u&#39;new&#39;: [],</span>
<span class="go">      u&#39;ok&#39;: [u&#39;d6842fb08a76e351&#39;,</span>
<span class="go">              u&#39;10a4b652da44e82a&#39;,</span>
<span class="go">              u&#39;81c601a2549966a0&#39;,</span>
<span class="go">              u&#39;a154f05e3bcee26b&#39;,</span>
<span class="go">              u&#39;1352fe19ddce0400&#39;,</span>
<span class="go">              u&#39;06d549c52d753e53&#39;,</span>
<span class="go">              u&#39;9ec54455d6279cc7&#39;],</span>
<span class="go">      u&#39;paused&#39;: [],</span>
<span class="go">      u&#39;queued&#39;: [],</span>
<span class="go">      u&#39;running&#39;: [],</span>
<span class="go">      u&#39;setting_metadata&#39;: [],</span>
<span class="go">      u&#39;upload&#39;: []</span>
<span class="go">      }</span>
<span class="go">  }</span>
</pre></div>
</div>
<p id="example-dataset">This gives us a dictionary containing the History&#8217;s metadata. With <code class="docutils literal"><span class="pre">contents=False</span></code> (the default), we only get a list of ids of the datasets contained within the History; with <code class="docutils literal"><span class="pre">contents=True</span></code> we would get metadata on each dataset. We can also directly access more detailed information on a particular dataset by passing its id to the <code class="docutils literal"><span class="pre">show_dataset</span></code> method:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">show_dataset</span><span class="p">(</span><span class="s">&#39;10a4b652da44e82a&#39;</span><span class="p">)</span>
<span class="go">{u&#39;data_type&#39;: u&#39;fastqsanger&#39;,</span>
<span class="go"> u&#39;deleted&#39;: False,</span>
<span class="go"> u&#39;file_size&#39;: 16527060,</span>
<span class="go"> u&#39;genome_build&#39;: u&#39;dm3&#39;,</span>
<span class="go"> u&#39;id&#39;: 17499,</span>
<span class="go"> u&#39;metadata_data_lines&#39;: None,</span>
<span class="go"> u&#39;metadata_dbkey&#39;: u&#39;dm3&#39;,</span>
<span class="go"> u&#39;metadata_sequences&#39;: None,</span>
<span class="go"> u&#39;misc_blurb&#39;: u&#39;15.8 MB&#39;,</span>
<span class="go"> u&#39;misc_info&#39;: u&#39;Noneuploaded fastqsanger file&#39;,</span>
<span class="go"> u&#39;model_class&#39;: u&#39;HistoryDatasetAssociation&#39;,</span>
<span class="go"> u&#39;name&#39;: u&#39;C1_R2_1.chr4.fq&#39;,</span>
<span class="go"> u&#39;purged&#39;: False,</span>
<span class="go"> u&#39;state&#39;: u&#39;ok&#39;,</span>
<span class="go"> u&#39;visible&#39;: True}</span>
</pre></div>
</div>
</div>
<div class="section" id="view-data-libraries">
<h2>View Data Libraries<a class="headerlink" href="#view-data-libraries" title="Permalink to this headline">¶</a></h2>
<p>Methods for accessing Data Libraries are grouped under <code class="docutils literal"><span class="pre">GalaxyInstance.libraries.*</span></code>. Most Data Library methods are available to all users, but as only administrators can create new Data Libraries within Galaxy, the <code class="docutils literal"><span class="pre">create_folder</span></code> and <code class="docutils literal"><span class="pre">create_library</span></code> methods can only be called using an API key belonging to an admin account.</p>
<p>We can view the Data Libraries available to our account using:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">libraries</span><span class="o">.</span><span class="n">get_libraries</span><span class="p">()</span>
<span class="go">[{u&#39;id&#39;: u&#39;8e6f930d00d123ea&#39;,</span>
<span class="go">  u&#39;name&#39;: u&#39;RNA-seq workshop data&#39;,</span>
<span class="go">  u&#39;url&#39;: u&#39;/api/libraries/8e6f930d00d123ea&#39;},</span>
<span class="go"> {u&#39;id&#39;: u&#39;f740ab636b360a70&#39;,</span>
<span class="go">  u&#39;name&#39;: u&#39;1000 genomes&#39;,</span>
<span class="go">  u&#39;url&#39;: u&#39;/api/libraries/f740ab636b360a70&#39;}]</span>
</pre></div>
</div>
<p>This gives a list of metadata dictionaries with basic information on each library. We can get more information on a particular Data Library by passing its id to the <code class="docutils literal"><span class="pre">show_library</span></code> method:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">libraries</span><span class="o">.</span><span class="n">show_library</span><span class="p">(</span><span class="s">&#39;8e6f930d00d123ea&#39;</span><span class="p">)</span>
<span class="go">{u&#39;contents_url&#39;: u&#39;/api/libraries/8e6f930d00d123ea/contents&#39;,</span>
<span class="go"> u&#39;description&#39;: u&#39;RNA-Seq workshop data&#39;,</span>
<span class="go"> u&#39;name&#39;: u&#39;RNA-Seq&#39;,</span>
<span class="go"> u&#39;synopsis&#39;: u&#39;Data for the RNA-Seq tutorial&#39;}</span>
</pre></div>
</div>
</div>
<div class="section" id="upload-files-to-a-data-library">
<h2>Upload files to a Data Library<a class="headerlink" href="#upload-files-to-a-data-library" title="Permalink to this headline">¶</a></h2>
<p>We can get files into Data Libraries in several ways: by uploading from our local machine, by retrieving from a URL, by passing the new file content directly into the method, or by importing a file from the filesystem on the Galaxy server.</p>
<p>For instance, to upload a file from our machine we might call:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">libraries</span><span class="o">.</span><span class="n">upload_file_from_local_path</span><span class="p">(</span><span class="s">&#39;8e6f930d00d123ea&#39;</span><span class="p">,</span> <span class="s">&#39;/local/path/to/mydata.fastq&#39;</span><span class="p">,</span> <span class="n">file_type</span><span class="o">=</span><span class="s">&#39;fastqsanger&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Note that we have provided the id of the destination Data Library, and in this case we have specified the type that Galaxy should assign to the new dataset. The default value for <code class="docutils literal"><span class="pre">file_type</span></code> is &#8216;auto&#8217;, in which case Galaxy will attempt to guess the dataset type.</p>
</div>
<div class="section" id="view-workflows">
<h2>View Workflows<a class="headerlink" href="#view-workflows" title="Permalink to this headline">¶</a></h2>
<p>Methods for accessing workflows are grouped under <code class="docutils literal"><span class="pre">GalaxyInstance.workflows.*</span></code>.</p>
<p>To get information on the Workflows currently in your account, use:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">workflows</span><span class="o">.</span><span class="n">get_workflows</span><span class="p">()</span>
<span class="go">[{u&#39;id&#39;: u&#39;e8b85ad72aefca86&#39;,</span>
<span class="go">  u&#39;name&#39;: u&quot;TopHat + cufflinks part 1&quot;,</span>
<span class="go">  u&#39;url&#39;: u&#39;/api/workflows/e8b85ad72aefca86&#39;},</span>
<span class="go"> {u&#39;id&#39;: u&#39;b0631c44aa74526d&#39;,</span>
<span class="go">  u&#39;name&#39;: u&#39;CuffDiff&#39;,</span>
<span class="go">  u&#39;url&#39;: u&#39;/api/workflows/b0631c44aa74526d&#39;}]</span>
</pre></div>
</div>
<p>This returns a list of metadata dictionaries. We can get the details of a particular Workflow, including its steps, by passing its id to the <code class="docutils literal"><span class="pre">show_workflow</span></code> method:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">workflows</span><span class="o">.</span><span class="n">show_workflow</span><span class="p">(</span><span class="s">&#39;e8b85ad72aefca86&#39;</span><span class="p">)</span>
<span class="go">{u&#39;id&#39;: u&#39;e8b85ad72aefca86&#39;,</span>
<span class="go"> u&#39;inputs&#39;:</span>
<span class="go">    {u&#39;252&#39;:</span>
<span class="go">       {u&#39;label&#39;: u&#39;Input RNA-seq fastq&#39;,</span>
<span class="go">        u&#39;value&#39;: u&#39;&#39;</span>
<span class="go">        }</span>
<span class="go">     },</span>
<span class="go"> u&#39;name&#39;: u&quot;TopHat + cufflinks part 1&quot;,</span>
<span class="go"> u&#39;steps&#39;:</span>
<span class="go">    {u&#39;250&#39;:</span>
<span class="go">       {u&#39;id&#39;: 250,</span>
<span class="go">        u&#39;input_steps&#39;:</span>
<span class="go">           {u&#39;input1&#39;:</span>
<span class="go">              {u&#39;source_step&#39;: 252,</span>
<span class="go">               u&#39;step_output&#39;: u&#39;output&#39;</span>
<span class="go">               }</span>
<span class="go">           },</span>
<span class="go">        u&#39;tool_id&#39;: u&#39;tophat&#39;,</span>
<span class="go">        u&#39;type&#39;: u&#39;tool&#39;</span>
<span class="go">        },</span>
<span class="go">     u&#39;251&#39;:</span>
<span class="go">        {u&#39;id&#39;: 251,</span>
<span class="go">         u&#39;input_steps&#39;:</span>
<span class="go">            {u&#39;input&#39;:</span>
<span class="go">               {u&#39;source_step&#39;: 250,</span>
<span class="go">                u&#39;step_output&#39;: u&#39;accepted_hits&#39;</span>
<span class="go">                }</span>
<span class="go">            },</span>
<span class="go">         u&#39;tool_id&#39;: u&#39;cufflinks&#39;,</span>
<span class="go">         u&#39;type&#39;: u&#39;tool&#39;</span>
<span class="go">         },</span>
<span class="go">     u&#39;252&#39;:</span>
<span class="go">        {u&#39;id&#39;: 252,</span>
<span class="go">         u&#39;input_steps&#39;: {},</span>
<span class="go">         u&#39;tool_id&#39;: None,</span>
<span class="go">         u&#39;type&#39;: u&#39;data_input&#39;</span>
<span class="go">         }</span>
<span class="go">     },</span>
<span class="go"> u&#39;url&#39;: u&#39;/api/workflows/e8b85ad72aefca86&#39;</span>
<span class="go"> }</span>
</pre></div>
</div>
</div>
<div class="section" id="export-or-import-a-workflow">
<h2>Export or import a Workflow<a class="headerlink" href="#export-or-import-a-workflow" title="Permalink to this headline">¶</a></h2>
<p>Workflows can be exported from or imported into Galaxy as JSON. This makes it possible to archive Workflows, or to move them between Galaxy instances.</p>
<p>To export a workflow, we can call:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">workflow_string</span> <span class="o">=</span> <span class="n">gi</span><span class="o">.</span><span class="n">workflows</span><span class="o">.</span><span class="n">export_workflow_json</span><span class="p">(</span><span class="s">&#39;e8b85ad72aefca86&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>This gives us a (rather long) string with a JSON-encoded representation of the Workflow. We can import this string as a new Workflow with:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">workflows</span><span class="o">.</span><span class="n">import_workflow_json</span><span class="p">(</span><span class="n">workflow_string</span><span class="p">)</span>
<span class="go">{u&#39;id&#39;: u&#39;c0bacafdfe211f9a&#39;,</span>
<span class="go"> u&#39;name&#39;: u&#39;TopHat + cufflinks part 1 (imported from API)&#39;,</span>
<span class="go"> u&#39;url&#39;: u&#39;/api/workflows/c0bacafdfe211f9a&#39;}</span>
</pre></div>
</div>
<p>This call returns a dictionary containing basic metadata on the new Workflow object. Since in this case we have imported the JSON string into the original Galaxy instance, we now have a duplicate of the original Workflow in our account:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">workflows</span><span class="o">.</span><span class="n">get_workflows</span><span class="p">()</span>
<span class="go">[{u&#39;id&#39;: u&#39;c0bacafdfe211f9a&#39;,</span>
<span class="go">  u&#39;name&#39;: u&#39;TopHat + cufflinks part 1 (imported from API)&#39;,</span>
<span class="go">  u&#39;url&#39;: u&#39;/api/workflows/c0bacafdfe211f9a&#39;},</span>
<span class="go"> {u&#39;id&#39;: u&#39;e8b85ad72aefca86&#39;,</span>
<span class="go">  u&#39;name&#39;: u&quot;TopHat + cufflinks part 1&quot;,</span>
<span class="go">  u&#39;url&#39;: u&#39;/api/workflows/e8b85ad72aefca86&#39;},</span>
<span class="go"> {u&#39;id&#39;: u&#39;b0631c44aa74526d&#39;,</span>
<span class="go">  u&#39;name&#39;: u&#39;CuffDiff&#39;,</span>
<span class="go">  u&#39;url&#39;: u&#39;/api/workflows/b0631c44aa74526d&#39;}]</span>
</pre></div>
</div>
<p>Instead of using JSON strings directly, Workflows can be exported to or imported from files on the local disk using the <code class="docutils literal"><span class="pre">export_workflow_to_local_path</span></code> and <code class="docutils literal"><span class="pre">import_workflow_from_local_path</span></code> methods. See the <a class="reference internal" href="all.html#workflows-api"><em>API reference</em></a> for details.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If we export a Workflow from one Galaxy instance and import it into another, Galaxy will only run it without modification if it has the same versions of the tool wrappers installed. This is to ensure reproducibility. Otherwise, we will need to manually update the Workflow to use the new tool versions.</p>
</div>
</div>
<div class="section" id="run-a-workflow">
<h2>Run a Workflow<a class="headerlink" href="#run-a-workflow" title="Permalink to this headline">¶</a></h2>
<p>To run a Workflow, we need to tell Galaxy which datasets to use for which workflow inputs. We can use datasets from Histories or Data Libraries.</p>
<p>Examine the Workflow above. We can see that it takes only one input file. That is:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">wf</span> <span class="o">=</span> <span class="n">gi</span><span class="o">.</span><span class="n">workflows</span><span class="o">.</span><span class="n">show_workflow</span><span class="p">(</span><span class="s">&#39;e8b85ad72aefca86&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">wf</span><span class="p">[</span><span class="s">&#39;inputs&#39;</span><span class="p">]</span>
<span class="go">{u&#39;252&#39;:</span>
<span class="go">    {u&#39;label&#39;:</span>
<span class="go">        u&#39;Input RNA-seq fastq&#39;,</span>
<span class="go">        u&#39;value&#39;: u&#39;&#39;</span>
<span class="go">    }</span>
<span class="go">}</span>
</pre></div>
</div>
<p>There is one input, labelled &#8216;Input RNA-seq fastq&#8217;. This input is passed to the Tophat tool and should be a fastq file. We will use the dataset we examined above, under <a class="reference internal" href="#view-histories-and-datasets"><em>View Histories and Datasets</em></a>, which had name &#8216;C1_R2_1.chr4.fq&#8217; and id &#8216;10a4b652da44e82a&#8217;.</p>
<p>To specify the inputs, we build a data map and pass this to the <code class="docutils literal"><span class="pre">run_workflow</span></code> method. This data map is a nested dictionary object which maps inputs to datasets. We call:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">datamap</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">datamap</span><span class="p">[</span><span class="s">&#39;252&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span> <span class="s">&#39;src&#39;</span><span class="p">:</span><span class="s">&#39;hda&#39;</span><span class="p">,</span> <span class="s">&#39;id&#39;</span><span class="p">:</span><span class="s">&#39;10a4b652da44e82a&#39;</span> <span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">workflows</span><span class="o">.</span><span class="n">run_workflow</span><span class="p">(</span><span class="s">&#39;e8b85ad72aefca86&#39;</span><span class="p">,</span> <span class="n">datamap</span><span class="p">,</span> <span class="n">history_name</span><span class="o">=</span><span class="s">&#39;New output history&#39;</span><span class="p">)</span>
<span class="go">{u&#39;history&#39;: u&#39;0a7b7992a7cabaec&#39;,</span>
<span class="go"> u&#39;outputs&#39;: [u&#39;33be8ad9917d9207&#39;,</span>
<span class="go">              u&#39;fbee1c2dc793c114&#39;,</span>
<span class="go">              u&#39;85866441984f9e28&#39;,</span>
<span class="go">              u&#39;1c51aa78d3742386&#39;,</span>
<span class="go">              u&#39;a68e8770e52d03b4&#39;,</span>
<span class="go">              u&#39;c54baf809e3036ac&#39;,</span>
<span class="go">              u&#39;ba0db8ce6cd1fe8f&#39;,</span>
<span class="go">              u&#39;c019e4cf08b2ac94&#39;</span>
<span class="go">              ]</span>
<span class="go">}</span>
</pre></div>
</div>
<p>In this case the only input id is &#8216;252&#8217; and the corresponding dataset id is &#8216;10a4b652da44e82a&#8217;. We have specified the dataset source to be &#8216;hda&#8217; (HistoryDatasetAssociation) since the dataset is stored in a History. See the <a class="reference internal" href="all.html#workflows-api"><em>API reference</em></a> for allowed dataset specifications. We have also requested that a new History be created and used to store the results of the run, by setting <code class="docutils literal"><span class="pre">history_name='New</span> <span class="pre">output</span> <span class="pre">history'</span></code>.</p>
<p>The <code class="docutils literal"><span class="pre">run_workflow</span></code> call submits all the jobs which need to be run to the Galaxy workflow engine, with the appropriate dependencies so that they will run in order. The call returns immediately, so we can continue to submit new jobs while waiting for this workflow to execute. <code class="docutils literal"><span class="pre">run_workflow</span></code> returns the id of the output History and of the datasets that will be created as a result of this run. Note that these dataset ids are valid immediately, so we can specify these datasets as inputs to new jobs even before the files have been created, and the new jobs will be added to the queue with the appropriate dependencies.</p>
<p>If we view the output History immediately after calling <code class="docutils literal"><span class="pre">run_workflow</span></code>, we will see something like:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">histories</span><span class="o">.</span><span class="n">show_history</span><span class="p">(</span><span class="s">&#39;0a7b7992a7cabaec&#39;</span><span class="p">)</span>
<span class="go">{u&#39;annotation&#39;: u&#39;&#39;,</span>
<span class="go"> u&#39;contents_url&#39;: u&#39;/api/histories/0a7b7992a7cabaec/contents&#39;,</span>
<span class="go"> u&#39;id&#39;: u&#39;0a7b7992a7cabaec&#39;,</span>
<span class="go"> u&#39;name&#39;: u&#39;New output history&#39;,</span>
<span class="go"> u&#39;nice_size&#39;: u&#39;0 bytes&#39;,</span>
<span class="go"> u&#39;state&#39;: u&#39;queued&#39;,</span>
<span class="go"> u&#39;state_details&#39;: {u&#39;discarded&#39;: 0,</span>
<span class="go">     u&#39;empty&#39;: 0,</span>
<span class="go">     u&#39;error&#39;: 0,</span>
<span class="go">     u&#39;failed_metadata&#39;: 0,</span>
<span class="go">     u&#39;new&#39;: 0,</span>
<span class="go">     u&#39;ok&#39;: 0,</span>
<span class="go">     u&#39;paused&#39;: 0,</span>
<span class="go">     u&#39;queued&#39;: 8,</span>
<span class="go">     u&#39;running&#39;: 0,</span>
<span class="go">     u&#39;setting_metadata&#39;: 0,</span>
<span class="go">     u&#39;upload&#39;: 0},</span>
<span class="go"> u&#39;state_ids&#39;: {u&#39;discarded&#39;: [],</span>
<span class="go">     u&#39;empty&#39;: [],</span>
<span class="go">     u&#39;error&#39;: [],</span>
<span class="go">     u&#39;failed_metadata&#39;: [],</span>
<span class="go">     u&#39;new&#39;: [],</span>
<span class="go">     u&#39;ok&#39;: [],</span>
<span class="go">     u&#39;paused&#39;: [],</span>
<span class="go">     u&#39;queued&#39;: [u&#39;33be8ad9917d9207&#39;,</span>
<span class="go">                 u&#39;fbee1c2dc793c114&#39;,</span>
<span class="go">                 u&#39;85866441984f9e28&#39;,</span>
<span class="go">                 u&#39;1c51aa78d3742386&#39;,</span>
<span class="go">                 u&#39;a68e8770e52d03b4&#39;,</span>
<span class="go">                 u&#39;c54baf809e3036ac&#39;,</span>
<span class="go">                 u&#39;ba0db8ce6cd1fe8f&#39;,</span>
<span class="go">                 u&#39;c019e4cf08b2ac94&#39;],</span>
<span class="go">     u&#39;running&#39;: [],</span>
<span class="go">     u&#39;setting_metadata&#39;: [],</span>
<span class="go">     u&#39;upload&#39;: []</span>
<span class="go">    }</span>
<span class="go">}</span>
</pre></div>
</div>
<p>In this case, because the submitted jobs have not had time to run, the output History contains 8 datasets in the &#8216;queued&#8217; state and has a total size of 0 bytes. If we make this call again later we should instead see completed output files.</p>
</div>
<div class="section" id="view-users">
<h2>View Users<a class="headerlink" href="#view-users" title="Permalink to this headline">¶</a></h2>
<p>Methods for managing users are grouped under <code class="docutils literal"><span class="pre">GalaxyInstance.users.*</span></code>. User management is only available to Galaxy administrators, that is, the API key used to connect to Galaxy must be that of an admin account.</p>
<p>To get a list of users, call:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi</span><span class="o">.</span><span class="n">users</span><span class="o">.</span><span class="n">get_users</span><span class="p">()</span>
<span class="go">[{u&#39;email&#39;: u&#39;userA@unimelb.edu.au&#39;,</span>
<span class="go">  u&#39;id&#39;: u&#39;975a9ce09b49502a&#39;,</span>
<span class="go">  u&#39;quota_percent&#39;: None,</span>
<span class="go">  u&#39;url&#39;: u&#39;/api/users/975a9ce09b49502a&#39;},</span>
<span class="go"> {u&#39;email&#39;: u&#39;userB@student.unimelb.edu.au&#39;,</span>
<span class="go">  u&#39;id&#39;: u&#39;0193a95acf427d2c&#39;,</span>
<span class="go">  u&#39;quota_percent&#39;: None,</span>
<span class="go">  u&#39;url&#39;: u&#39;/api/users/0193a95acf427d2c&#39;}]</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Usage documentation</a><ul>
<li><a class="reference internal" href="#connect-to-a-galaxy-server">Connect to a Galaxy server</a></li>
<li><a class="reference internal" href="#view-histories-and-datasets">View Histories and Datasets</a></li>
<li><a class="reference internal" href="#view-data-libraries">View Data Libraries</a></li>
<li><a class="reference internal" href="#upload-files-to-a-data-library">Upload files to a Data Library</a></li>
<li><a class="reference internal" href="#view-workflows">View Workflows</a></li>
<li><a class="reference internal" href="#export-or-import-a-workflow">Export or import a Workflow</a></li>
<li><a class="reference internal" href="#run-a-workflow">Run a Workflow</a></li>
<li><a class="reference internal" href="#view-users">View Users</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="all.html"
                        title="previous chapter">API documentation for interacting with Galaxy</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="../lib_config.html"
                        title="next chapter">Configuration documents for BioBlend</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../../_sources/api_docs/galaxy/docs.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../lib_config.html" title="Configuration documents for BioBlend"
             >next</a> |</li>
        <li class="right" >
          <a href="all.html" title="API documentation for interacting with Galaxy"
             >previous</a> |</li>
        <li><a href="../../index.html">BioBlend 0.5.3-dev documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2012, Enis Afgan.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3b2.
    </div>
  </body>
</html>