<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>samsifter.util.genetics &mdash; SamSifter 0.5.3 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.5.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <link rel="top" title="SamSifter 0.5.3 documentation" href="../../../index.html" />
    <link rel="up" title="Module code" href="../../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../../index.html">SamSifter 0.5.3 documentation</a> &raquo;</li>
          <li><a href="../../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for samsifter.util.genetics</h1><div class="highlight"><pre>
<span class="c"># -*- coding: utf-8 -*-</span>
<span class="sd">&quot;&quot;&quot;Library of standard methods for manipulation of genetic sequences.</span>

<span class="sd">.. moduleauthor:: Florian Aldehoff (f.aldehoff@student.uni-tuebingen.de)</span>
<span class="sd">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="is_iupac"><a class="viewcode-back" href="../../../samsifter.util.html#samsifter.util.genetics.is_iupac">[docs]</a><span class="k">def</span> <span class="nf">is_iupac</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Check nucleotide sequence for invalid (non-IUPAC) codes.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seq : str</span>
<span class="sd">        Nucleotide sequence.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bool</span>
<span class="sd">        True if sequence contains only IUPAC ambiguity codes, otherwise False.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">iupac</span> <span class="o">=</span> <span class="nb">set</span><span class="p">((</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span>
                 <span class="s">&#39;Y&#39;</span><span class="p">,</span> <span class="s">&#39;R&#39;</span><span class="p">,</span>
                 <span class="s">&#39;W&#39;</span><span class="p">,</span> <span class="s">&#39;S&#39;</span><span class="p">,</span>
                 <span class="s">&#39;K&#39;</span><span class="p">,</span> <span class="s">&#39;M&#39;</span><span class="p">,</span>
                 <span class="s">&#39;B&#39;</span><span class="p">,</span> <span class="s">&#39;H&#39;</span><span class="p">,</span> <span class="s">&#39;D&#39;</span><span class="p">,</span> <span class="s">&#39;V&#39;</span><span class="p">,</span>
                 <span class="s">&#39;X&#39;</span><span class="p">,</span> <span class="s">&#39;N&#39;</span><span class="p">,</span>
                 <span class="s">&#39;-&#39;</span><span class="p">,</span>
                 <span class="s">&#39;a&#39;</span><span class="p">,</span> <span class="s">&#39;t&#39;</span><span class="p">,</span> <span class="s">&#39;c&#39;</span><span class="p">,</span> <span class="s">&#39;g&#39;</span><span class="p">,</span>
                 <span class="s">&#39;y&#39;</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">,</span>
                 <span class="s">&#39;w&#39;</span><span class="p">,</span> <span class="s">&#39;s&#39;</span><span class="p">,</span>
                 <span class="s">&#39;k&#39;</span><span class="p">,</span> <span class="s">&#39;m&#39;</span><span class="p">,</span>
                 <span class="s">&#39;b&#39;</span><span class="p">,</span> <span class="s">&#39;h&#39;</span><span class="p">,</span> <span class="s">&#39;d&#39;</span><span class="p">,</span> <span class="s">&#39;v&#39;</span><span class="p">,</span>
                 <span class="s">&#39;x&#39;</span><span class="p">,</span> <span class="s">&#39;n&#39;</span><span class="p">))</span>
    <span class="k">return</span> <span class="nb">set</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="o">.</span><span class="n">issubset</span><span class="p">(</span><span class="n">iupac</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="opposite"><a class="viewcode-back" href="../../../samsifter.util.html#samsifter.util.genetics.opposite">[docs]</a><span class="k">def</span> <span class="nf">opposite</span><span class="p">(</span><span class="n">base</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Mirror base code into the exact opposite IUPAC ambiguity code.</span>

<span class="sd">    &#39;A&#39; becomes &#39;not A&#39;, gap becomes any base, etc.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    base : str</span>
<span class="sd">        Single one-letter IUPAC nucleotide code.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Opposite one-letter IUPAC ambiguity code.</span>
<span class="sd">    None</span>
<span class="sd">        If base is invalid.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">opposites</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span>    <span class="s">&#39;B&#39;</span><span class="p">,</span>
                 <span class="s">&#39;G&#39;</span><span class="p">:</span>    <span class="s">&#39;H&#39;</span><span class="p">,</span>
                 <span class="s">&#39;C&#39;</span><span class="p">:</span>    <span class="s">&#39;D&#39;</span><span class="p">,</span>
                 <span class="s">&#39;T&#39;</span><span class="p">:</span>    <span class="s">&#39;V&#39;</span><span class="p">,</span>
                 <span class="s">&#39;Y&#39;</span><span class="p">:</span>    <span class="s">&#39;R&#39;</span><span class="p">,</span>    <span class="c"># pyrimidine | purin</span>
                 <span class="s">&#39;R&#39;</span><span class="p">:</span>    <span class="s">&#39;Y&#39;</span><span class="p">,</span>    <span class="c"># purine | pyrimidine</span>
                 <span class="s">&#39;W&#39;</span><span class="p">:</span>    <span class="s">&#39;S&#39;</span><span class="p">,</span>    <span class="c"># weak | strong</span>
                 <span class="s">&#39;S&#39;</span><span class="p">:</span>    <span class="s">&#39;W&#39;</span><span class="p">,</span>    <span class="c"># strong | weak</span>
                 <span class="s">&#39;K&#39;</span><span class="p">:</span>    <span class="s">&#39;M&#39;</span><span class="p">,</span>    <span class="c"># keto | amino</span>
                 <span class="s">&#39;M&#39;</span><span class="p">:</span>    <span class="s">&#39;K&#39;</span><span class="p">,</span>    <span class="c"># amino | keto</span>
                 <span class="s">&#39;B&#39;</span><span class="p">:</span>    <span class="s">&#39;A&#39;</span><span class="p">,</span>
                 <span class="s">&#39;H&#39;</span><span class="p">:</span>    <span class="s">&#39;G&#39;</span><span class="p">,</span>
                 <span class="s">&#39;D&#39;</span><span class="p">:</span>    <span class="s">&#39;C&#39;</span><span class="p">,</span>
                 <span class="s">&#39;V&#39;</span><span class="p">:</span>    <span class="s">&#39;T&#39;</span><span class="p">,</span>
                 <span class="s">&#39;X&#39;</span><span class="p">:</span>    <span class="s">&#39;-&#39;</span><span class="p">,</span>
                 <span class="s">&#39;N&#39;</span><span class="p">:</span>    <span class="s">&#39;-&#39;</span><span class="p">,</span>
                 <span class="s">&#39;-&#39;</span><span class="p">:</span>    <span class="s">&#39;N&#39;</span><span class="p">,</span>    <span class="c"># ____________________</span>
                 <span class="s">&#39;a&#39;</span><span class="p">:</span>    <span class="s">&#39;b&#39;</span><span class="p">,</span>
                 <span class="s">&#39;g&#39;</span><span class="p">:</span>    <span class="s">&#39;h&#39;</span><span class="p">,</span>
                 <span class="s">&#39;c&#39;</span><span class="p">:</span>    <span class="s">&#39;d&#39;</span><span class="p">,</span>
                 <span class="s">&#39;t&#39;</span><span class="p">:</span>    <span class="s">&#39;v&#39;</span><span class="p">,</span>
                 <span class="s">&#39;y&#39;</span><span class="p">:</span>    <span class="s">&#39;t&#39;</span><span class="p">,</span>    <span class="c"># pyrimidine | purin</span>
                 <span class="s">&#39;r&#39;</span><span class="p">:</span>    <span class="s">&#39;y&#39;</span><span class="p">,</span>    <span class="c"># purine | pyrimidine</span>
                 <span class="s">&#39;w&#39;</span><span class="p">:</span>    <span class="s">&#39;s&#39;</span><span class="p">,</span>    <span class="c"># weak | strong</span>
                 <span class="s">&#39;s&#39;</span><span class="p">:</span>    <span class="s">&#39;w&#39;</span><span class="p">,</span>    <span class="c"># strong | weak</span>
                 <span class="s">&#39;k&#39;</span><span class="p">:</span>    <span class="s">&#39;m&#39;</span><span class="p">,</span>    <span class="c"># keto | amino</span>
                 <span class="s">&#39;m&#39;</span><span class="p">:</span>    <span class="s">&#39;k&#39;</span><span class="p">,</span>    <span class="c"># amino | keto</span>
                 <span class="s">&#39;b&#39;</span><span class="p">:</span>    <span class="s">&#39;a&#39;</span><span class="p">,</span>
                 <span class="s">&#39;h&#39;</span><span class="p">:</span>    <span class="s">&#39;g&#39;</span><span class="p">,</span>
                 <span class="s">&#39;d&#39;</span><span class="p">:</span>    <span class="s">&#39;c&#39;</span><span class="p">,</span>
                 <span class="s">&#39;v&#39;</span><span class="p">:</span>    <span class="s">&#39;t&#39;</span><span class="p">,</span>
                 <span class="s">&#39;x&#39;</span><span class="p">:</span>    <span class="s">&#39;-&#39;</span><span class="p">,</span>
                 <span class="s">&#39;n&#39;</span><span class="p">:</span>    <span class="s">&#39;-&#39;</span><span class="p">}</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">opposite</span> <span class="o">=</span> <span class="n">opposites</span><span class="p">[</span><span class="n">base</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">None</span>
    <span class="k">return</span> <span class="n">opposite</span>

</div>
<div class="viewcode-block" id="complement"><a class="viewcode-back" href="../../../samsifter.util.html#samsifter.util.genetics.complement">[docs]</a><span class="k">def</span> <span class="nf">complement</span><span class="p">(</span><span class="n">base</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Returns a base&#39;s complement according to IUPAC or None.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    base : str</span>
<span class="sd">        Single one-letter IUPAC nucleotide code.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Complementary one-letter IUPAC ambiguity code.</span>
<span class="sd">    None</span>
<span class="sd">        If base is invalid.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">complements</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span>    <span class="s">&#39;T&#39;</span><span class="p">,</span>
                   <span class="s">&#39;G&#39;</span><span class="p">:</span>    <span class="s">&#39;C&#39;</span><span class="p">,</span>
                   <span class="s">&#39;C&#39;</span><span class="p">:</span>    <span class="s">&#39;A&#39;</span><span class="p">,</span>
                   <span class="s">&#39;T&#39;</span><span class="p">:</span>    <span class="s">&#39;G&#39;</span><span class="p">,</span>
                   <span class="s">&#39;Y&#39;</span><span class="p">:</span>    <span class="s">&#39;R&#39;</span><span class="p">,</span>    <span class="c"># pyrimidine &gt; purin</span>
                   <span class="s">&#39;R&#39;</span><span class="p">:</span>    <span class="s">&#39;Y&#39;</span><span class="p">,</span>    <span class="c"># purine &gt; pyrimidine</span>
                   <span class="s">&#39;W&#39;</span><span class="p">:</span>    <span class="s">&#39;W&#39;</span><span class="p">,</span>    <span class="c"># weak: A or T</span>
                   <span class="s">&#39;S&#39;</span><span class="p">:</span>    <span class="s">&#39;S&#39;</span><span class="p">,</span>    <span class="c"># strong: G or C</span>
                   <span class="s">&#39;K&#39;</span><span class="p">:</span>    <span class="s">&#39;M&#39;</span><span class="p">,</span>    <span class="c"># keto &gt; amino</span>
                   <span class="s">&#39;M&#39;</span><span class="p">:</span>    <span class="s">&#39;K&#39;</span><span class="p">,</span>    <span class="c"># amino &gt; keto</span>
                   <span class="s">&#39;D&#39;</span><span class="p">:</span>    <span class="s">&#39;H&#39;</span><span class="p">,</span>
                   <span class="s">&#39;V&#39;</span><span class="p">:</span>    <span class="s">&#39;B&#39;</span><span class="p">,</span>
                   <span class="s">&#39;H&#39;</span><span class="p">:</span>    <span class="s">&#39;D&#39;</span><span class="p">,</span>
                   <span class="s">&#39;B&#39;</span><span class="p">:</span>    <span class="s">&#39;V&#39;</span><span class="p">,</span>
                   <span class="s">&#39;X&#39;</span><span class="p">:</span>    <span class="s">&#39;X&#39;</span><span class="p">,</span>
                   <span class="s">&#39;N&#39;</span><span class="p">:</span>    <span class="s">&#39;N&#39;</span><span class="p">,</span>
                   <span class="s">&#39;-&#39;</span><span class="p">:</span>    <span class="s">&#39;-&#39;</span><span class="p">,</span>    <span class="c"># ____________________</span>
                   <span class="s">&#39;a&#39;</span><span class="p">:</span>    <span class="s">&#39;t&#39;</span><span class="p">,</span>
                   <span class="s">&#39;g&#39;</span><span class="p">:</span>    <span class="s">&#39;c&#39;</span><span class="p">,</span>
                   <span class="s">&#39;c&#39;</span><span class="p">:</span>    <span class="s">&#39;a&#39;</span><span class="p">,</span>
                   <span class="s">&#39;t&#39;</span><span class="p">:</span>    <span class="s">&#39;g&#39;</span><span class="p">,</span>
                   <span class="s">&#39;y&#39;</span><span class="p">:</span>    <span class="s">&#39;r&#39;</span><span class="p">,</span>
                   <span class="s">&#39;r&#39;</span><span class="p">:</span>    <span class="s">&#39;y&#39;</span><span class="p">,</span>
                   <span class="s">&#39;w&#39;</span><span class="p">:</span>    <span class="s">&#39;w&#39;</span><span class="p">,</span>
                   <span class="s">&#39;s&#39;</span><span class="p">:</span>    <span class="s">&#39;s&#39;</span><span class="p">,</span>
                   <span class="s">&#39;k&#39;</span><span class="p">:</span>    <span class="s">&#39;m&#39;</span><span class="p">,</span>
                   <span class="s">&#39;m&#39;</span><span class="p">:</span>    <span class="s">&#39;k&#39;</span><span class="p">,</span>
                   <span class="s">&#39;d&#39;</span><span class="p">:</span>    <span class="s">&#39;h&#39;</span><span class="p">,</span>
                   <span class="s">&#39;v&#39;</span><span class="p">:</span>    <span class="s">&#39;b&#39;</span><span class="p">,</span>
                   <span class="s">&#39;h&#39;</span><span class="p">:</span>    <span class="s">&#39;d&#39;</span><span class="p">,</span>
                   <span class="s">&#39;b&#39;</span><span class="p">:</span>    <span class="s">&#39;v&#39;</span><span class="p">,</span>
                   <span class="s">&#39;x&#39;</span><span class="p">:</span>    <span class="s">&#39;x&#39;</span><span class="p">,</span>
                   <span class="s">&#39;n&#39;</span><span class="p">:</span>    <span class="s">&#39;n&#39;</span><span class="p">}</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">complement</span> <span class="o">=</span> <span class="n">complements</span><span class="p">[</span><span class="n">base</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">None</span>
    <span class="k">return</span> <span class="n">complement</span>

</div>
<div class="viewcode-block" id="reverse"><a class="viewcode-back" href="../../../samsifter.util.html#samsifter.util.genetics.reverse">[docs]</a><span class="k">def</span> <span class="nf">reverse</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reverses a sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seq : str</span>
<span class="sd">        Any sequence.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Reverse sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">seq</span><span class="p">[::</span><span class="mi">1</span><span class="p">]</span>

</div>
<div class="viewcode-block" id="reverse_complement"><a class="viewcode-back" href="../../../samsifter.util.html#samsifter.util.genetics.reverse_complement">[docs]</a><span class="k">def</span> <span class="nf">reverse_complement</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reverse complement a nucleotide sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seq : str</span>
<span class="sd">        A nucleotide sequence of IUPAC ambuigity codes.</span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Reverse complement of the nucleotide sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">rev</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">revcomp</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
    <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">rev</span><span class="p">:</span>
        <span class="n">revcomp</span> <span class="o">+=</span> <span class="n">complement</span><span class="p">(</span><span class="n">base</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">revcomp</span>

</div>
<div class="viewcode-block" id="transcribe"><a class="viewcode-back" href="../../../samsifter.util.html#samsifter.util.genetics.transcribe">[docs]</a><span class="k">def</span> <span class="nf">transcribe</span><span class="p">(</span><span class="n">dna</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Transcription of DNA sequence to corresponding RNA sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    dna : str</span>
<span class="sd">        DNA nucleotide sequence.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        RNA nucleotide sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">dna</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;U&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;t&#39;</span><span class="p">,</span> <span class="s">&#39;u&#39;</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="reverse_transcribe"><a class="viewcode-back" href="../../../samsifter.util.html#samsifter.util.genetics.reverse_transcribe">[docs]</a><span class="k">def</span> <span class="nf">reverse_transcribe</span><span class="p">(</span><span class="n">rna</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reverse transcription of RNA sequence to corresponding cDNA sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    rna : str</span>
<span class="sd">        RNA nucleotide sequence.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        DNA nucleotide sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">rna</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;U&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">&#39;u&#39;</span><span class="p">,</span> <span class="s">&#39;t&#39;</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="gc"><a class="viewcode-back" href="../../../samsifter.util.html#samsifter.util.genetics.gc">[docs]</a><span class="k">def</span> <span class="nf">gc</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Calculate GC content of a nucleotide sequence.</span>

<span class="sd">    Considers all IUPAC codes using 6x counts to deal with 1/2 and 1/3</span>
<span class="sd">    probabilities from ambiguity codes.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seq : str</span>
<span class="sd">        A nucleotide sequence.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        GC content as fraction of 1.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="mf">0.0</span>

    <span class="n">gc_at_values</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>
                    <span class="s">&#39;G&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
                    <span class="s">&#39;C&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
                    <span class="s">&#39;T&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>
                    <span class="s">&#39;Y&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>    <span class="c"># pyrimidine (C or T)</span>
                    <span class="s">&#39;R&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>    <span class="c"># purine (A or G)</span>
                    <span class="s">&#39;W&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>    <span class="c"># weak (A or T)</span>
                    <span class="s">&#39;S&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>    <span class="c"># strong (G or C)</span>
                    <span class="s">&#39;K&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>    <span class="c"># keto (T or G)</span>
                    <span class="s">&#39;M&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>    <span class="c"># amino (A or C)</span>
                    <span class="s">&#39;B&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>    <span class="c"># not A</span>
                    <span class="s">&#39;H&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>    <span class="c"># not G</span>
                    <span class="s">&#39;D&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>    <span class="c"># not C</span>
                    <span class="s">&#39;V&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>    <span class="c"># not T</span>
                    <span class="s">&#39;X&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>    <span class="c"># any</span>
                    <span class="s">&#39;N&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>    <span class="c"># any</span>
                    <span class="s">&#39;-&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>    <span class="c"># ____________________</span>
                    <span class="s">&#39;a&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>
                    <span class="s">&#39;g&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
                    <span class="s">&#39;c&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
                    <span class="s">&#39;t&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>
                    <span class="s">&#39;y&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
                    <span class="s">&#39;r&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
                    <span class="s">&#39;w&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>
                    <span class="s">&#39;s&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
                    <span class="s">&#39;k&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
                    <span class="s">&#39;m&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
                    <span class="s">&#39;b&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
                    <span class="s">&#39;h&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
                    <span class="s">&#39;d&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
                    <span class="s">&#39;v&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
                    <span class="s">&#39;x&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
                    <span class="s">&#39;n&#39;</span><span class="p">:</span>    <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)}</span>
    <span class="n">gc</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">at</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">:</span>
        <span class="n">gc</span> <span class="o">+=</span> <span class="n">gc_at_values</span><span class="p">[</span><span class="n">base</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">at</span> <span class="o">+=</span> <span class="n">gc_at_values</span><span class="p">[</span><span class="n">base</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">gc</span> <span class="o">/</span> <span class="p">((</span><span class="n">at</span> <span class="o">+</span> <span class="n">gc</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.0</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="aln_identity"><a class="viewcode-back" href="../../../samsifter.util.html#samsifter.util.genetics.aln_identity">[docs]</a><span class="k">def</span> <span class="nf">aln_identity</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">include_indels</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">include_deamination</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
                 <span class="n">include_unknown</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Determines identity of two sequences in an alignment.</span>

<span class="sd">    Calculates modified Hamming distance of an alignment with gaps with</span>
<span class="sd">    optional exclusion of possibly deaminated T&gt;C and A&gt;G as well as indels.</span>

<span class="sd">    Enabling all three optional parameters will calculate values identical to</span>
<span class="sd">    MALT while the default settings will calculate values identical to</span>
<span class="sd">    PMDtools.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    read : str</span>
<span class="sd">        Full read sequence.</span>
<span class="sd">    ref : str</span>
<span class="sd">        Corresponding reference sequence (only the aligned part, but including</span>
<span class="sd">        any gaps, skips and indels).</span>
<span class="sd">    include_indels : bool, optional</span>
<span class="sd">        Consider insertions and deletions, defaults to False.</span>
<span class="sd">    include_deamination : bool, optional</span>
<span class="sd">        Consider any potentially deaminated bases, defaults to False.</span>
<span class="sd">    include_unknown : bool, optional</span>
<span class="sd">        Consider any unknown bases (N), defaults to False.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float</span>
<span class="sd">        Identity of read and reference as fraction of 1.</span>
<span class="sd">    str</span>
<span class="sd">        Mismatch string (match = ``|``, mismatch = ``x``, indel = ``-``).</span>

<span class="sd">    Example</span>
<span class="sd">    -------</span>
<span class="sd">    Showing differences between PMDtools and MALT::</span>

<span class="sd">        TCCAGCAGGTCGATGACCTTGATGCCGGTCTCGAACATCTTCA</span>
<span class="sd">        ||-|||||||||||||||||||||||||||||||||||||-|x</span>
<span class="sd">        TC-AGCAGGTCGATGACCTTGATGCCGGTCTCGAACATCT-CG</span>
<span class="sd">        ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]</span>

<span class="sd">        43 bases</span>
<span class="sd">        2 gaps</span>
<span class="sd">        1 total mismatches</span>
<span class="sd">            1 G&gt;A mismatches (A in read but G in reference)</span>
<span class="sd">            0 C&gt;T mismatches (T in read but C in reference)</span>
<span class="sd">            0 other mismatches</span>
<span class="sd">        40 matches</span>

<span class="sd">        PMD:    40 / 40 = 100%</span>
<span class="sd">        identity = 1.0 * matches / (matches + other mismatches)</span>
<span class="sd">                 = 1.0 * 40 / (40 + 0)</span>
<span class="sd">                 = 1.0</span>

<span class="sd">        MALT:   40 / 43 = 93%</span>
<span class="sd">        identity = 1.0 * matches / (matches + gaps + total mismatches)</span>
<span class="sd">                 = 1.0 * 40 / (40 + 2 + 1)</span>
<span class="sd">                 = 0.930232</span>
<span class="sd">                 ~ 0.93</span>

<span class="sd">    Thus it is required to include indels, potentially deaminated bases and</span>
<span class="sd">    unknown bases to produce identity values similar to MALT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># this should never happen if reference was correctly reconstructed</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">read</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&quot;Undefined for sequences of unequal length&quot;</span><span class="p">)</span>

    <span class="n">match</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">mismatch</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">mismatch_string</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
    <span class="k">for</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="n">ref</span><span class="p">):</span>
        <span class="n">thesebases</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">]</span>
        <span class="k">if</span> <span class="s">&#39;-&#39;</span> <span class="ow">in</span> <span class="n">thesebases</span><span class="p">:</span>
            <span class="n">mismatch_string</span> <span class="o">+=</span> <span class="s">&#39;-&#39;</span>
            <span class="c"># default: exclude indels from mismatches</span>
            <span class="k">if</span> <span class="n">include_indels</span><span class="p">:</span>
                <span class="n">mismatch</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="s">&#39;N&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">include_unknown</span><span class="p">:</span>
                <span class="n">mismatch</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">continue</span>
        <span class="k">if</span> <span class="n">b</span> <span class="o">==</span> <span class="s">&#39;N&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">include_unknown</span><span class="p">:</span>
                <span class="n">mismatch</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">continue</span>
        <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="n">b</span><span class="p">:</span>
            <span class="n">mismatch_string</span> <span class="o">+=</span> <span class="s">&#39;|&#39;</span>
            <span class="n">match</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">a</span> <span class="o">!=</span> <span class="n">b</span><span class="p">:</span>
            <span class="n">mismatch_string</span> <span class="o">+=</span> <span class="s">&#39;x&#39;</span>
            <span class="c"># default: exclude possible deamination from mismatches</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">include_deamination</span><span class="p">:</span>
                <span class="k">if</span> <span class="s">&#39;C&#39;</span> <span class="o">==</span> <span class="n">b</span> <span class="ow">and</span> <span class="s">&#39;T&#39;</span> <span class="o">==</span> <span class="n">a</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">if</span> <span class="s">&#39;G&#39;</span> <span class="o">==</span> <span class="n">b</span> <span class="ow">and</span> <span class="s">&#39;A&#39;</span> <span class="o">==</span> <span class="n">a</span><span class="p">:</span>
                    <span class="k">continue</span>
            <span class="n">mismatch</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">perc_identity</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">*</span> <span class="n">match</span> <span class="o">/</span> <span class="p">(</span><span class="n">match</span> <span class="o">+</span> <span class="n">mismatch</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ZeroDivisionError</span><span class="p">:</span>
        <span class="n">perc_identity</span> <span class="o">=</span> <span class="mf">0.0</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">perc_identity</span><span class="p">,</span> <span class="n">mismatch_string</span><span class="p">)</span></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../../index.html">SamSifter 0.5.3 documentation</a> &raquo;</li>
          <li><a href="../../index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2015, Florian Aldehoff.
      Last updated on Mar 09, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
  </body>
</html>