<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Annotation with snpEff or VEP</title>
    
    <link rel="stylesheet" href="../_static/labibi.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.12.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/gemini.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="Preprocessing and Loading a VCF file into GEMINI" href="preprocessing.html" />
    <link rel="prev" title="Quick start" href="quick_start.html" />


<script src="../_static/labibi.js"></script>



<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', '']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>


  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="preprocessing.html" title="Preprocessing and Loading a VCF file into GEMINI"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="quick_start.html" title="Quick start"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  
  <div class="section" id="annotation-with-snpeff-or-vep">
<h1>Annotation with snpEff or VEP<a class="headerlink" href="#annotation-with-snpeff-or-vep" title="Permalink to this headline">¶</a></h1>
<p><cite>GEMINI</cite> depends upon external tools to predict the functional consequence of variants in a VCF file.
We currently support annotations produced by either <a class="reference external" href="http://snpeff.sourceforge.net/">SnpEff</a>
or <a class="reference external" href="http://www.ensembl.org/info/docs/variation/vep/index.html">VEP</a>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Versions tested: VEP versions 73 through 75 and core SnpEff versions 3.0 through 3.6.
GEMINI supports ENSEMBL annotations hence users are expected to download genome databases
for these tools as represented in the examples below.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Version support would be subsequently updated here, as we test along
and add or edit changes available with the latest version of these tools.</p>
</div>
<p>Recommended instructions for annotating existing VCF files with these tools are summarized here.</p>
<div class="section" id="stepwise-installation-and-usage-of-vep">
<h2>Stepwise installation and usage of VEP<a class="headerlink" href="#stepwise-installation-and-usage-of-vep" title="Permalink to this headline">¶</a></h2>
<p>Download the Variant Effect Predictor &#8220;standalone perl script&#8221; from
<a class="reference external" href="http://www.ensembl.org/info/docs/tools/vep/script/vep_download.html">Ensembl</a>.
You can choose a specific version of VEP to download here</p>
<p>Example:</p>
<div class="highlight-bash"><div class="highlight"><pre>Download version 74
</pre></div>
</div>
<p>Untar the tarball into the current directory</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>tar -zxvf variant_effect_predictor.tar.gz
</pre></div>
</div>
<p>This will create the variant_effect_predictor directory. Now do the following for install:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span><span class="nb">cd </span>variant_effect_predictor
<span class="nv">$ </span>perl INSTALL.pl <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
<p>By default this would install the API&#8217;s, bioperl-1.2.3 and the cache files (in
the $HOME/.vep directory).</p>
<div class="section" id="homebrew-vep-installation">
<h3>Homebrew VEP installation<a class="headerlink" href="#homebrew-vep-installation" title="Permalink to this headline">¶</a></h3>
<p>If you are a <a class="reference external" href="http://brew.sh/">Homebrew</a> or <a class="reference external" href="http://brew.sh/linuxbrew/">Linuxbrew</a>
user, there is an automated recipe to install the main VEP script and plugins in
the <a class="reference external" href="https://github.com/chapmanb/homebrew-cbl">CloudBioLinux homebrew repository</a>:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>brew tap chapmanb/cbl
<span class="nv">$ </span>brew update
<span class="nv">$ </span>brew install vep
</pre></div>
</div>
</div>
<div class="section" id="manual-installation-of-vep">
<h3>Manual installation of VEP<a class="headerlink" href="#manual-installation-of-vep" title="Permalink to this headline">¶</a></h3>
<p>For those (e.g mac users) who have a problem installing through this install script, try a
manual installation of the API&#8217;s, BioPerl-1.2.3 and set all pre-requisites for running VEP
(<tt class="docutils literal"><span class="pre">DBI</span></tt> and <tt class="docutils literal"><span class="pre">DBD::mysql</span></tt> modules required). The appropriate pre-build caches should be
downloaded for Human to the <tt class="docutils literal"><span class="pre">$HOME/.vep</span></tt> directory and then untar.</p>
<p>You may follow instructions at <a class="reference external" href="http://www.ensembl.org/info/docs/api/api_installation.html">http://www.ensembl.org/info/docs/api/api_installation.html</a>
which provides alternate options for the API installation and additional tips for windows/mac
users. It also has information for setting up your environment to run VEP.</p>
<p>Example download of the cache files</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>wget ftp://ftp.ensembl.org/pub/release-73/variation/VEP/homo_sapiens_vep_73.tar.gz
</pre></div>
</div>
<p>You may change the release date in this example to get the appropriate cache files for your version
of VEP that you have installed.</p>
<p>Example</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>wget ftp://ftp.ensembl.org/pub/release-74/variation/VEP/homo_sapiens_vep_74.tar.gz
</pre></div>
</div>
<p>Cache requires the <tt class="docutils literal"><span class="pre">gzip</span></tt> and <tt class="docutils literal"><span class="pre">zcat</span></tt> utilities. VEP uses <tt class="docutils literal"><span class="pre">zcat</span></tt>
to decompress cached files. For systems where zcat may not be installed or may
not work, the following option needs to be added along with the <tt class="docutils literal"><span class="pre">--cache</span></tt> option:</p>
<div class="highlight-bash"><div class="highlight"><pre>--compress <span class="s2">&quot;gunzip -c&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="running-vep">
<h3>Running VEP<a class="headerlink" href="#running-vep" title="Permalink to this headline">¶</a></h3>
<p>You may now run VEP as:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>perl variant_effect_predictor.pl <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p>We recommend running VEP with the following options as currently we support
VEP fields specified as below:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>perl variant_effect_predictor.pl -i example.vcf <span class="se">\</span>
    --cache <span class="se">\</span>
    --sift b <span class="se">\</span>
    --polyphen b <span class="se">\</span>
    --symbol <span class="se">\</span>
    --numbers <span class="se">\</span>
    --biotype <span class="se">\</span>
    --total_length <span class="se">\</span>
    -o output <span class="se">\</span>
    --vcf <span class="se">\</span>
    --fields Consequence,Codons,Amino_acids,Gene,SYMBOL,Feature,EXON,PolyPhen,SIFT,Protein_position,BIOTYPE
</pre></div>
</div>
<p>A documentation for the above specified options may be found at
<a class="reference external" href="http://www.ensembl.org/info/docs/tools/vep/script/vep_options.html">http://www.ensembl.org/info/docs/tools/vep/script/vep_options.html</a></p>
<p>As of GEMINI version 0.8.0, you can also run VEP with additional fields, which
will be automatically added to the variants table as columns. As an example,
run VEP on your VCF with the <a class="reference external" href="https://github.com/ensembl-variation/VEP_plugins/blob/master/dbNSFP.pm">dbNSFP</a>
and <a class="reference external" href="https://github.com/konradjk/loftee">LOFTEE</a> plugins to annotate potential high
impact variations:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>variant_effect_predictor.pl --sift b --polyphen b --symbol --numbers --biotype <span class="se">\</span>
--total_length --canonical --ccds <span class="se">\</span>
--fields Consequence,Codons,Amino_acids,Gene,SYMBOL,Feature,EXON,PolyPhen,SIFT,Protein_position,BIOTYPE,CANONICAL,CCDS,RadialSVM_score,RadialSVM_pred,LR_score,LR_pred,CADD_raw,CADD_phred,Reliability_index,LoF,LoF_filter,LoF_flags <span class="se">\</span>
--plugin dbNSFP,/path/to/dbNSFP_v2.5.gz,RadialSVM_score,RadialSVM_pred,LR_score,LR_pred,CADD_raw,CADD_phred,Reliability_index <span class="se">\</span>
--plugin LoF,human_ancestor_fa:/path/to/human_ancestor.fa
</pre></div>
</div>
<p>Feeding this into GEMINI produces a variants table with columns for each of the
additional VEP metrics. The annotation loader names each column by prefixing
<tt class="docutils literal"><span class="pre">vep_</span></tt> to the origin VEP name, so select on <tt class="docutils literal"><span class="pre">vep_radialsvm_score</span></tt> or
<tt class="docutils literal"><span class="pre">vep_lof_filter</span></tt> in the final database.</p>
</div>
</div>
<div class="section" id="stepwise-installation-and-usage-of-snpeff">
<h2>Stepwise installation and usage of SnpEff<a class="headerlink" href="#stepwise-installation-and-usage-of-snpeff" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Basic Requirements: Java v1.7 or later; at least 4GB of memory</p>
</div>
<p>Download the supported versions of SnpEff from <a class="reference external" href="http://snpeff.sourceforge.net/download.html">http://snpeff.sourceforge.net/download.html</a></p>
<p>Example:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>wget http://sourceforge.net/projects/snpeff/files/snpEff_v3_6_core.zip
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">SnpEff should be installed preferably in <tt class="docutils literal"><span class="pre">snpEff</span></tt> directory in your
home directory. Else, you must update the <tt class="docutils literal"><span class="pre">data_dir</span></tt> parameter in
your snpEff.config file. For e.g. if the installation of snpEff has been done
in <tt class="docutils literal"><span class="pre">~/src</span></tt> instead of <tt class="docutils literal"><span class="pre">~/</span></tt> then change the data_dir parameter in
snpEff.config to <tt class="docutils literal"><span class="pre">data_dir</span> <span class="pre">=</span> <span class="pre">~/src/snpEff/data/</span></tt></p>
</div>
<p>Unzip the downloaded package.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>unzip snpEff_v3_6_core.zip
</pre></div>
</div>
<p>Change to the <tt class="docutils literal"><span class="pre">snpEff</span></tt> directory and download the genome database.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span><span class="nb">cd </span>snpEff_v3_6_core
<span class="nv">$ </span>java -jar snpEff.jar download GRCh37.69
</pre></div>
</div>
<p>Unzip the downloaded genome database. This will create and place the genome
in the &#8216;data&#8217; directory</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>unzip snpEff_v3_6_GRCh37.69.zip
</pre></div>
</div>
<p>To annotate a vcf using snpEff, use the <tt class="docutils literal"><span class="pre">default</span></tt> options as below:</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Memory options for the run may be specified as <tt class="docutils literal"><span class="pre">-Xmx4G</span></tt> (4GB)</p>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>java -Xmx4G -jar snpEff.jar -i vcf -o vcf GRCh37.69 example.vcf &gt; example_snpeff.vcf
</pre></div>
</div>
<p>If running from a directory different from the installation directory, the
complete path needs to be specified as,  e.g.:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>java -Xmx4G -jar path/to/snpEff/snpEff.jar -c path/to/snpEff/snpEff.config GRCh37.69 path/to/example.vcf &gt; example_snpeff.vcf
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">When using the latest versions of snpEff (e.g. 4.1) annotate your VCF with the additional parameters <tt class="docutils literal"><span class="pre">-classic</span></tt> and <tt class="docutils literal"><span class="pre">-formatEff</span></tt>.
This would ensure proper loading of the gene info columns in the variants table.</p>
</div>
</div>
<div class="section" id="columns-populated-by-snpeff-vep-tools">
<h2>Columns populated by snpEff/VEP tools<a class="headerlink" href="#columns-populated-by-snpeff-vep-tools" title="Permalink to this headline">¶</a></h2>
<p>The following variant consequence columns in the variant/variant_impacts table,
are populated with these annotations, which are otherwise set to null.</p>
<ul class="simple">
<li>anno_id</li>
<li>gene</li>
<li>transcript</li>
<li>exon</li>
<li>is_exonic</li>
<li>is_lof</li>
<li>is_coding</li>
<li>codon_change</li>
<li>aa_change</li>
<li>aa_length</li>
<li>biotype</li>
<li>impact</li>
<li>impact_so</li>
<li>impact_severity</li>
<li>polyphen_pred</li>
<li>polyphen_score</li>
<li>sift_pred</li>
<li>sift_score</li>
</ul>
</div>
<div class="section" id="standardizing-impact-definitions-for-gemini">
<h2>Standardizing <tt class="docutils literal"><span class="pre">impact</span></tt> definitions for GEMINI<a class="headerlink" href="#standardizing-impact-definitions-for-gemini" title="Permalink to this headline">¶</a></h2>
<p>GEMINI uses slightly modified impact terms (for ease) to describe the functional consequence of a given variant as provided by
snpEff/VEP.</p>
<p>The table below shows the alternate <cite>GEMINI</cite> terms used for <cite>snpEff/VEP</cite>.</p>
<table border="1" class="docutils">
<colgroup>
<col width="25%" />
<col width="20%" />
<col width="55%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Gemini terms</th>
<th class="head">snpEff terms</th>
<th class="head">VEP terms (uses SO by default)</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>splice_acceptor</td>
<td>SPLICE_SITE_ACCEPTOR</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001574">splice_acceptor_variant</a></td>
</tr>
<tr class="row-odd"><td>splice_donor</td>
<td>SPLICE_SITE_DONOR</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001575">splice_donor_variant</a></td>
</tr>
<tr class="row-even"><td>stop_gain</td>
<td>STOP_GAINED</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001587">stop_gained</a></td>
</tr>
<tr class="row-odd"><td>stop_loss</td>
<td>STOP_LOST</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001578">stop_lost</a></td>
</tr>
<tr class="row-even"><td>frame_shift</td>
<td>FRAME_SHIFT</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001589">frameshift_variant</a></td>
</tr>
<tr class="row-odd"><td>start_loss</td>
<td>START_LOST</td>
<td>null</td>
</tr>
<tr class="row-even"><td>exon_deleted</td>
<td>EXON_DELETED</td>
<td>null</td>
</tr>
<tr class="row-odd"><td>non_synonymous_start</td>
<td>NON_SYNONYMOUS_START</td>
<td>null</td>
</tr>
<tr class="row-even"><td>transcript_codon_change</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001582">initiator_codon_variant</a></td>
</tr>
<tr class="row-odd"><td>chrom_large_del</td>
<td>CHROMOSOME_LARGE_DELETION</td>
<td>null</td>
</tr>
<tr class="row-even"><td>rare_amino_acid</td>
<td>RARE_AMINO_ACID</td>
<td>null</td>
</tr>
<tr class="row-odd"><td>non_syn_coding</td>
<td>NON_SYNONYMOUS_CODING</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001583">missense_variant</a></td>
</tr>
<tr class="row-even"><td>inframe_codon_gain</td>
<td>CODON_INSERTION</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001821">inframe_insertion</a></td>
</tr>
<tr class="row-odd"><td>inframe_codon_loss</td>
<td>CODON_DELETION</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001822">inframe_deletion</a></td>
</tr>
<tr class="row-even"><td>inframe_codon_change</td>
<td>CODON_CHANGE</td>
<td>null</td>
</tr>
<tr class="row-odd"><td>codon_change_del</td>
<td>CODON_CHANGE_PLUS_CODON_DELETION</td>
<td>null</td>
</tr>
<tr class="row-even"><td>codon_change_ins</td>
<td>CODON_CHANGE_PLUS_CODON_INSERTION</td>
<td>null</td>
</tr>
<tr class="row-odd"><td>UTR_5_del</td>
<td>UTR_5_DELETED</td>
<td>null</td>
</tr>
<tr class="row-even"><td>UTR_3_del</td>
<td>UTR_3_DELETED</td>
<td>null</td>
</tr>
<tr class="row-odd"><td>splice_region</td>
<td>SPLICE_SITE_REGION</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001630">splice_region_variant</a></td>
</tr>
<tr class="row-even"><td>mature_miRNA</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001620">mature_miRNA_variant</a></td>
</tr>
<tr class="row-odd"><td>regulatory_region</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001566">regulatory_region_variant</a></td>
</tr>
<tr class="row-even"><td>TF_binding_site</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001782">TF_binding_site_variant</a></td>
</tr>
<tr class="row-odd"><td>regulatory_region_ablation</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001894">regulatory_region_ablation</a></td>
</tr>
<tr class="row-even"><td>regulatory_region_amplification</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001891">regulatory_region_amplification</a></td>
</tr>
<tr class="row-odd"><td>TFBS_ablation</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001895">TFBS_ablation</a></td>
</tr>
<tr class="row-even"><td>TFBS_amplification</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001892">TFBS_amplification</a></td>
</tr>
<tr class="row-odd"><td>synonymous_stop</td>
<td>SYNONYMOUS_STOP</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001567">stop_retained_variant</a></td>
</tr>
<tr class="row-even"><td>synonymous_coding</td>
<td>SYNONYMOUS_CODING</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001819">synonymous_variant</a></td>
</tr>
<tr class="row-odd"><td>UTR_5_prime</td>
<td>UTR_5_PRIME</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001623">5_prime_UTR_variant</a></td>
</tr>
<tr class="row-even"><td>UTR_3_prime</td>
<td>UTR_3_PRIME</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001624">3_prime_UTR_variant</a></td>
</tr>
<tr class="row-odd"><td>intron</td>
<td>INTRON</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001627">intron_variant</a></td>
</tr>
<tr class="row-even"><td>CDS</td>
<td>CDS</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001580">coding_sequence_variant</a></td>
</tr>
<tr class="row-odd"><td>upstream</td>
<td>UPSTREAM</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001631">upstream_gene_variant</a></td>
</tr>
<tr class="row-even"><td>downstream</td>
<td>DOWNSTREAM</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001632">downstream_gene_variant</a></td>
</tr>
<tr class="row-odd"><td>intergenic</td>
<td>INTERGENIC</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001628">intergenic_variant</a></td>
</tr>
<tr class="row-even"><td>intergenic_conserved</td>
<td>INTERGENIC_CONSERVED</td>
<td>null</td>
</tr>
<tr class="row-odd"><td>intragenic</td>
<td>INTRAGENIC</td>
<td>null</td>
</tr>
<tr class="row-even"><td>gene</td>
<td>GENE</td>
<td>null</td>
</tr>
<tr class="row-odd"><td>transcript</td>
<td>TRANSCRIPT</td>
<td>null</td>
</tr>
<tr class="row-even"><td>exon</td>
<td>EXON</td>
<td>null</td>
</tr>
<tr class="row-odd"><td>start_gain</td>
<td>START_GAINED</td>
<td>null</td>
</tr>
<tr class="row-even"><td>synonymous_start</td>
<td>SYNONYMOUS_START</td>
<td>null</td>
</tr>
<tr class="row-odd"><td>intron_conserved</td>
<td>INTRON_CONSERVED</td>
<td>null</td>
</tr>
<tr class="row-even"><td>nc_transcript</td>
<td>null</td>
<td><a class="reference external" href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001619">nc_transcript_variant</a> (should have been returned by VEP as: non_coding_transcript_variant)</td>
</tr>
<tr class="row-odd"><td>NMD_transcript</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001621">NMD_transcript_variant</a></td>
</tr>
<tr class="row-even"><td>incomplete_terminal_codon</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001626">incomplete_terminal_codon_variant</a></td>
</tr>
<tr class="row-odd"><td>nc_exon</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001792">non_coding_exon_variant</a> (should have been returned by VEP as: non_coding_transcript_exon_variant)</td>
</tr>
<tr class="row-even"><td>transcript_ablation</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001893">transcript_ablation</a></td>
</tr>
<tr class="row-odd"><td>transcript_amplification</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001889">transcript_amplification</a></td>
</tr>
<tr class="row-even"><td>feature elongation</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001907">feature_elongation</a></td>
</tr>
<tr class="row-odd"><td>feature truncation</td>
<td>null</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001906">feature_truncation</a></td>
</tr>
</tbody>
</table>
<p><em>Note: &#8220;null&#8221; refers to the absence of the corresponding term in the alternate database</em></p>
</div>
<div class="section" id="so-impact-definitions-in-gemini">
<h2>SO <tt class="docutils literal"><span class="pre">impact</span></tt> definitions in GEMINI<a class="headerlink" href="#so-impact-definitions-in-gemini" title="Permalink to this headline">¶</a></h2>
<p>The below table shows the <cite>Sequence Ontology (SO)</cite> term mappings for the GEMINI impacts, which is otherwise contained in the
<tt class="docutils literal"><span class="pre">impact_so</span></tt> column of the variants/variant_impacts table of the GEMINI database. The last column shows the severity terms
defined in GEMINI for these impacts.</p>
<table border="1" class="docutils">
<colgroup>
<col width="38%" />
<col width="47%" />
<col width="15%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Gemini terms (column: impact)</th>
<th class="head">Sequence Ontology terms (column: impact_so)</th>
<th class="head">Impact severity</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>splice_acceptor</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001574">splice_acceptor_variant</a></td>
<td>HIGH</td>
</tr>
<tr class="row-odd"><td>splice_donor</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001575">splice_donor_variant</a></td>
<td>HIGH</td>
</tr>
<tr class="row-even"><td>stop_gain</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001587">stop_gained</a></td>
<td>HIGH</td>
</tr>
<tr class="row-odd"><td>stop_loss</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001578">stop_lost</a></td>
<td>HIGH</td>
</tr>
<tr class="row-even"><td>frame_shift</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001589">frameshift_variant</a></td>
<td>HIGH</td>
</tr>
<tr class="row-odd"><td>start_loss</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002012">start_lost</a></td>
<td>HIGH</td>
</tr>
<tr class="row-even"><td>exon_deleted</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001572">exon_loss_variant</a></td>
<td>HIGH</td>
</tr>
<tr class="row-odd"><td>non_synonymous_start</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001582">initiator_codon_variant</a></td>
<td>HIGH</td>
</tr>
<tr class="row-even"><td>transcript_codon_change</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001582">initiator_codon_variant</a></td>
<td>HIGH</td>
</tr>
<tr class="row-odd"><td>chrom_large_del</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:1000029">chromosomal_deletion</a></td>
<td>HIGH</td>
</tr>
<tr class="row-even"><td>rare_amino_acid</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002008">rare_amino_acid_variant</a></td>
<td>HIGH</td>
</tr>
<tr class="row-odd"><td>non_syn_coding</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001583">missense_variant</a></td>
<td>MED</td>
</tr>
<tr class="row-even"><td>inframe_codon_gain</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001821">inframe_insertion</a></td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>inframe_codon_loss</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001822">inframe_deletion</a></td>
<td>MED</td>
</tr>
<tr class="row-even"><td>inframe_codon_change</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001580">coding_sequence_variant</a></td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>codon_change_del</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001826">disruptive_inframe_deletion</a></td>
<td>MED</td>
</tr>
<tr class="row-even"><td>codon_change_ins</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001824">disruptive_inframe_insertion</a></td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>UTR_5_del</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002013">5_prime_UTR_truncation</a> + <a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001572">exon_loss_variant</a></td>
<td>MED</td>
</tr>
<tr class="row-even"><td>UTR_3_del</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002015">3_prime_UTR_truncation</a> + <a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001572">exon_loss_variant</a></td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>splice_region</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001630">splice_region_variant</a></td>
<td>MED</td>
</tr>
<tr class="row-even"><td>mature_miRNA</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001620">mature_miRNA_variant</a></td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>regulatory_region</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001566">regulatory_region_variant</a></td>
<td>MED</td>
</tr>
<tr class="row-even"><td>TF_binding_site</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001782">TF_binding_site_variant</a></td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>regulatory_region_ablation</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001894">regulatory_region_ablation</a></td>
<td>MED</td>
</tr>
<tr class="row-even"><td>regulatory_region_amplification</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001891">regulatory_region_amplification</a></td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>TFBS_ablation</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001895">TFBS_ablation</a></td>
<td>MED</td>
</tr>
<tr class="row-even"><td>TFBS_amplification</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001892">TFBS_amplification</a></td>
<td>MED</td>
</tr>
<tr class="row-odd"><td>synonymous_stop</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001567">stop_retained_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>synonymous_coding</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001819">synonymous_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>UTR_5_prime</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001623">5_prime_UTR_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>UTR_3_prime</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001624">3_prime_UTR_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>intron</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001627">intron_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>CDS</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001580">coding_sequence_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>upstream</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001631">upstream_gene_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>downstream</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001632">downstream_gene_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>intergenic</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001628">intergenic_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>intergenic_conserved</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002017">conserved_intergenic_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>intragenic</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002011">intragenic_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>gene</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001564">gene_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>transcript</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001576">transcript_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>exon</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001791">exon_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>start_gain</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001988">5_prime_UTR_premature_start_codon_gain_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>synonymous_start</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002019">start_retained_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>intron_conserved</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002018">conserved_intron_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>nc_transcript</td>
<td><a class="reference external" href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001619">nc_transcript_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>NMD_transcript</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001621">NMD_transcript_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>incomplete_terminal_codon</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001626">incomplete_terminal_codon_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>nc_exon</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001792">non_coding_exon_variant</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>transcript_ablation</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001893">transcript_ablation</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>transcript_amplification</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001889">transcript_amplification</a></td>
<td>LOW</td>
</tr>
<tr class="row-even"><td>feature elongation</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001907">feature_elongation</a></td>
<td>LOW</td>
</tr>
<tr class="row-odd"><td>feature truncation</td>
<td><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001906">feature_truncation</a></td>
<td>LOW</td>
</tr>
</tbody>
</table>
</div>
</div>




  

<div id="disqus_thread"></div>
 <script type="text/javascript">
     /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
     var disqus_shortname = 'geminidocs'; // required: replace example with your forum shortname

     /* * * DON'T EDIT BELOW THIS LINE * * */
     (function() {
         var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
         dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
         (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
     })();
 </script>
 <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
 <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>






          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/gemini.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Annotation with snpEff or VEP</a><ul>
<li><a class="reference internal" href="#stepwise-installation-and-usage-of-vep">Stepwise installation and usage of VEP</a><ul>
<li><a class="reference internal" href="#homebrew-vep-installation">Homebrew VEP installation</a></li>
<li><a class="reference internal" href="#manual-installation-of-vep">Manual installation of VEP</a></li>
<li><a class="reference internal" href="#running-vep">Running VEP</a></li>
</ul>
</li>
<li><a class="reference internal" href="#stepwise-installation-and-usage-of-snpeff">Stepwise installation and usage of SnpEff</a></li>
<li><a class="reference internal" href="#columns-populated-by-snpeff-vep-tools">Columns populated by snpEff/VEP tools</a></li>
<li><a class="reference internal" href="#standardizing-impact-definitions-for-gemini">Standardizing <tt class="docutils literal"><span class="pre">impact</span></tt> definitions for GEMINI</a></li>
<li><a class="reference internal" href="#so-impact-definitions-in-gemini">SO <tt class="docutils literal"><span class="pre">impact</span></tt> definitions in GEMINI</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="quick_start.html"
                        title="previous chapter">Quick start</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="preprocessing.html"
                        title="next chapter">Preprocessing and Loading a VCF file into GEMINI</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/functional_annotation.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="preprocessing.html" title="Preprocessing and Loading a VCF file into GEMINI"
             >next</a> |</li>
        <li class="right" >
          <a href="quick_start.html" title="Quick start"
             >previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2012,2013.
      Last updated on Mar 19, 2015.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>

<!-- 
<div id="editor-trap">
      <h3>Edit and improve this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/arq5x/gemini/blob/master/docs/content/functional_annotation.rst">
          Annotation with snpEff or VEP
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/arq5x/gemini/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>
 -->
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', UA-39150563-1]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>


  </body>
</html>