<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>The Gemini database schema</title>
    
    <link rel="stylesheet" href="../_static/labibi.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.12.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/gemini.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="Using the GEMINI API" href="api.html" />
    <link rel="prev" title="The GEMINI browser interface" href="browser.html" />


<script src="../_static/labibi.js"></script>



<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', '']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>


  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="api.html" title="Using the GEMINI API"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="browser.html" title="The GEMINI browser interface"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  
  <div class="section" id="the-gemini-database-schema">
<h1>The Gemini database schema<a class="headerlink" href="#the-gemini-database-schema" title="Permalink to this headline">¶</a></h1>
<div class="section" id="the-variants-table">
<h2>The <tt class="docutils literal"><span class="pre">variants</span></tt> table<a class="headerlink" href="#the-variants-table" title="Permalink to this headline">¶</a></h2>
<div class="section" id="core-vcf-fields">
<h3>Core VCF fields<a class="headerlink" href="#core-vcf-fields" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="6%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">column_name</th>
<th class="head">type</th>
<th class="head">notes</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>chrom</td>
<td>STRING</td>
<td>The chromosome on which the variant resides (from VCF <tt class="docutils literal"><span class="pre">CHROM</span></tt> field).</td>
</tr>
<tr class="row-odd"><td>start</td>
<td>INTEGER</td>
<td>The 0-based start position. (from VCF <tt class="docutils literal"><span class="pre">POS</span></tt> field, but converted to 0-based coordinates)</td>
</tr>
<tr class="row-even"><td>end</td>
<td>INTEGER</td>
<td>The 1-based end position. (from VCF <tt class="docutils literal"><span class="pre">POS</span></tt> field, yet inferred based on the size of the variant)</td>
</tr>
<tr class="row-odd"><td>vcf_id</td>
<td>STRING</td>
<td>The VCF <tt class="docutils literal"><span class="pre">ID</span></tt> field.</td>
</tr>
<tr class="row-even"><td>variant_id</td>
<td>INTEGER</td>
<td>PRIMARY_KEY</td>
</tr>
<tr class="row-odd"><td>anno_id</td>
<td>INTEGER</td>
<td>Variant transcript number for the most severely affected transcript</td>
</tr>
<tr class="row-even"><td>ref</td>
<td>STRING</td>
<td>Reference allele (from VCF <tt class="docutils literal"><span class="pre">REF</span></tt> field)</td>
</tr>
<tr class="row-odd"><td>alt</td>
<td>STRING</td>
<td>Alternate alele for the variant (from VCF <tt class="docutils literal"><span class="pre">ALT</span></tt> field)</td>
</tr>
<tr class="row-even"><td>qual</td>
<td>INTEGER</td>
<td>Quality score for the assertion made in ALT (from VCF <tt class="docutils literal"><span class="pre">QUAL</span></tt> field)</td>
</tr>
<tr class="row-odd"><td>filter</td>
<td>STRING</td>
<td>A string of filters passed/failed in variant calling (from VCF <tt class="docutils literal"><span class="pre">FILTER</span></tt> field)</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="variant-and-popgen-info">
<h3>Variant and PopGen info<a class="headerlink" href="#variant-and-popgen-info" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="6%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>type</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">The type of variant.</div>
<div class="line">Any of: [<em>snp</em>, <em>indel</em>]</div>
</div>
</td>
</tr>
<tr class="row-odd"><td>sub_type</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">The variant sub-type.</div>
<div class="line">If <tt class="docutils literal"><span class="pre">type</span></tt> is <em>snp</em>:   [<em>ts</em>, (transition), <em>tv</em> (transversion)]</div>
<div class="line">If <tt class="docutils literal"><span class="pre">type</span></tt> is <em>indel</em>: [<em>ins</em>, (insertion), <em>del</em> (deletion)]</div>
</div>
</td>
</tr>
<tr class="row-even"><td>call_rate</td>
<td>FLOAT</td>
<td>The fraction of samples with a valid genotype</td>
</tr>
<tr class="row-odd"><td>num_hom_ref</td>
<td>INTEGER</td>
<td>The total number of of homozygotes for the reference (<tt class="docutils literal"><span class="pre">ref</span></tt>) allele</td>
</tr>
<tr class="row-even"><td>num_het</td>
<td>INTEGER</td>
<td>The total number of heterozygotes observed.</td>
</tr>
<tr class="row-odd"><td>num_hom_alt</td>
<td>INTEGER</td>
<td>The total number of homozygotes for the reference (<tt class="docutils literal"><span class="pre">alt</span></tt>) allele</td>
</tr>
<tr class="row-even"><td>num_unknown</td>
<td>INTEGER</td>
<td>The total number of of unknown genotypes</td>
</tr>
<tr class="row-odd"><td>aaf</td>
<td>FLOAT</td>
<td>The observed allele frequency for the alternate allele</td>
</tr>
<tr class="row-even"><td>hwe</td>
<td>FLOAT</td>
<td>The Chi-square probability of deviation from HWE (assumes random mating)</td>
</tr>
<tr class="row-odd"><td>inbreeding_coeff</td>
<td>FLOAT</td>
<td>The inbreeding co-efficient that expresses the likelihood of effects due to inbreeding</td>
</tr>
<tr class="row-even"><td>pi</td>
<td>FLOAT</td>
<td>The computed nucleotide diversity (pi) for the site</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="genotype-information">
<h3>Genotype information<a class="headerlink" href="#genotype-information" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="6%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>gts</td>
<td>BLOB</td>
<td><div class="first last line-block">
<div class="line">A compressed binary vector of sample genotypes (e.g., &#8220;A/A&#8221;, &#8220;A|G&#8221;, &#8220;G/G&#8221;)</div>
<div class="line">- Extracted from the VCF <tt class="docutils literal"><span class="pre">GT</span></tt> genotype tag.</div>
</div>
</td>
</tr>
<tr class="row-odd"><td>gt_types</td>
<td>BLOB</td>
<td><div class="first last line-block">
<div class="line">A compressed binary vector of numeric genotype &#8220;types&#8221; (e.g., 0, 1, 2)</div>
<div class="line">- Inferred from the VCF <tt class="docutils literal"><span class="pre">GT</span></tt> genotype tag.</div>
</div>
</td>
</tr>
<tr class="row-even"><td>gt_phases</td>
<td>BLOB</td>
<td><div class="first last line-block">
<div class="line">A compressed binary vector of sample genotype phases (e.g., False, True, False)</div>
<div class="line">- Extracted from the VCF <tt class="docutils literal"><span class="pre">GT</span></tt> genotype tag&#8217;s allele delimiter</div>
<div class="line-block">
<div class="line">e.g., <tt class="docutils literal"><span class="pre">A/G</span></tt> means an unphased genotype. Value is <strong>FALSE</strong>.</div>
<div class="line">e.g., <tt class="docutils literal"><span class="pre">A|G</span></tt> means a phased genotype. Value is <strong>TRUE</strong>.</div>
</div>
</div>
</td>
</tr>
<tr class="row-odd"><td>gt_depths</td>
<td>BLOB</td>
<td><div class="first last line-block">
<div class="line">A compressed binary vector of the depth of aligned sequence observed for each sample</div>
<div class="line">- Extracted from the VCF <tt class="docutils literal"><span class="pre">DP</span></tt> genotype tag.</div>
</div>
</td>
</tr>
<tr class="row-even"><td>gt_ref_depths</td>
<td>BLOB</td>
<td><div class="first last line-block">
<div class="line">A compressed binary vector of the depth of reference alleles observed for each sample</div>
<div class="line">- Extracted from the VCF <tt class="docutils literal"><span class="pre">AD</span></tt> genotype tag.</div>
</div>
</td>
</tr>
<tr class="row-odd"><td>gt_alt_depths</td>
<td>BLOB</td>
<td><div class="first last line-block">
<div class="line">A compressed binary vector of the depth of alternate alleles observed for each sample</div>
<div class="line">- Extracted from the VCF <tt class="docutils literal"><span class="pre">AD</span></tt> genotype tag.</div>
</div>
</td>
</tr>
<tr class="row-even"><td>gt_quals</td>
<td>BLOB</td>
<td><div class="first last line-block">
<div class="line">A compressed binary vector of the genotype quality (PHRED scale) estimates for each sample</div>
<div class="line">- Extracted from the VCF <tt class="docutils literal"><span class="pre">GQ</span></tt> genotype tag.</div>
</div>
</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="gene-information">
<h3>Gene information<a class="headerlink" href="#gene-information" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="17%" />
<col width="6%" />
<col width="77%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>gene</td>
<td>STRING</td>
<td>Corresponding gene name of the highly affected transcript</td>
</tr>
<tr class="row-odd"><td>transcript</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">The variant transcript that was most severely affected</div>
<div class="line">(for two equally affected transcripts,the protein_coding
biotype is prioritized (snpEff/VEP)</div>
</div>
</td>
</tr>
<tr class="row-even"><td>is_exonic</td>
<td>BOOL</td>
<td>Does the variant affect an exon for &gt;= 1transcript?</td>
</tr>
<tr class="row-odd"><td>is_coding</td>
<td>BOOL</td>
<td>Does the variant fall in a coding region (excl. 3&#8217; &amp; 5&#8217; UTRs) for &gt;= 1 transcript?</td>
</tr>
<tr class="row-even"><td>is_lof</td>
<td>BOOL</td>
<td>Based on the value of the impact col, is the variant LOF for &gt;= transcript?</td>
</tr>
<tr class="row-odd"><td>exon</td>
<td>STRING</td>
<td>Exon information for the severely affected transcript</td>
</tr>
<tr class="row-even"><td>codon_change</td>
<td>STRING</td>
<td>What is the codon change?</td>
</tr>
<tr class="row-odd"><td>aa_change</td>
<td>STRING</td>
<td>What is the amino acid change (for an snp)?</td>
</tr>
<tr class="row-even"><td>aa_length</td>
<td>STRING</td>
<td>The length of CDS in terms of number of amino acids (<tt class="docutils literal"><span class="pre">only</span> <span class="pre">snpEff</span></tt>)</td>
</tr>
<tr class="row-odd"><td>biotype</td>
<td>STRING</td>
<td>The &#8216;type&#8217; of the severely affected transcript (e.g.protein-coding, pseudogene, rRNA etc.) (<tt class="docutils literal"><span class="pre">only</span> <span class="pre">snpEff</span></tt>)</td>
</tr>
<tr class="row-even"><td>impact</td>
<td>STRING</td>
<td>The consequence of the most severely affected transcript</td>
</tr>
<tr class="row-odd"><td>impact_so</td>
<td>STRING</td>
<td>The Sequence ontology term for the most severe consequence</td>
</tr>
<tr class="row-even"><td>impact_severity</td>
<td>STRING</td>
<td>Severity of the highest order observed for the variant</td>
</tr>
<tr class="row-odd"><td>polyphen_pred</td>
<td>STRING</td>
<td>Polyphen predictions for the snps for the severely affected transcript (<tt class="docutils literal"><span class="pre">only</span> <span class="pre">VEP</span></tt>)</td>
</tr>
<tr class="row-even"><td>polyphen_score</td>
<td>FLOAT</td>
<td>Polyphen scores for the severely affected transcript (<tt class="docutils literal"><span class="pre">only</span> <span class="pre">VEP</span></tt>)</td>
</tr>
<tr class="row-odd"><td>sift_pred</td>
<td>STRING</td>
<td>SIFT predictions for the snp&#8217;s for the most severely affected transcript (<tt class="docutils literal"><span class="pre">only</span> <span class="pre">VEP</span></tt>)</td>
</tr>
<tr class="row-even"><td>sift_score</td>
<td>FLOAT</td>
<td>SIFT scores for the predictions (<tt class="docutils literal"><span class="pre">only</span> <span class="pre">VEP</span></tt>)</td>
</tr>
<tr class="row-odd"><td>pfam_domain</td>
<td>STRING</td>
<td>Pfam protein domain that the variant affects</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="optional-vcf-info-fields">
<h3>Optional VCF INFO fields<a class="headerlink" href="#optional-vcf-info-fields" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="6%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>anc_allele</td>
<td>STRING</td>
<td>The reported ancestral allele if there is one.</td>
</tr>
<tr class="row-odd"><td>rms_bq</td>
<td>FLOAT</td>
<td>The RMS base quality at this position.</td>
</tr>
<tr class="row-even"><td>cigar</td>
<td>STRING</td>
<td>CIGAR string describing how to align an alternate allele to the reference allele.</td>
</tr>
<tr class="row-odd"><td>depth</td>
<td>INTEGER</td>
<td>The number of aligned sequence reads that led to this variant call</td>
</tr>
<tr class="row-even"><td>strand_bias</td>
<td>FLOAT</td>
<td>Strand bias at the variant position</td>
</tr>
<tr class="row-odd"><td>rms_map_qual</td>
<td>FLOAT</td>
<td>RMS mapping quality, a measure of variance of quality scores</td>
</tr>
<tr class="row-even"><td>in_hom_run</td>
<td>INTEGER</td>
<td>Homopolymer runs for the variant allele</td>
</tr>
<tr class="row-odd"><td>num_mapq_zero</td>
<td>INTEGER</td>
<td>Total counts of reads with mapping quality equal to zero</td>
</tr>
<tr class="row-even"><td>num_alleles</td>
<td>INTEGER</td>
<td>Total number of alleles in called genotypes</td>
</tr>
<tr class="row-odd"><td>num_reads_w_dels</td>
<td>FLOAT</td>
<td>Fraction of reads with spanning deletions</td>
</tr>
<tr class="row-even"><td>haplotype_score</td>
<td>FLOAT</td>
<td>Consistency of the site with two segregating haplotypes</td>
</tr>
<tr class="row-odd"><td>qual_depth</td>
<td>FLOAT</td>
<td>Variant confidence or quality by depth</td>
</tr>
<tr class="row-even"><td>allele_count</td>
<td>INTEGER</td>
<td>Allele counts in genotypes</td>
</tr>
<tr class="row-odd"><td>allele_bal</td>
<td>FLOAT</td>
<td>Allele balance for hets</td>
</tr>
<tr class="row-even"><td>info</td>
<td>BLOB</td>
<td>Stores the <tt class="docutils literal"><span class="pre">INFO</span></tt> field of the VCF</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="population-information">
<h3>Population information<a class="headerlink" href="#population-information" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="18%" />
<col width="7%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>in_dbsnp</td>
<td>BOOL</td>
<td><div class="first last line-block">
<div class="line">Is this variant found in dbSnp (build 137)?</div>
<div class="line">0 : Absence of the variant in dbsnp</div>
<div class="line">1 : Presence of the variant in dbsnp</div>
</div>
</td>
</tr>
<tr class="row-odd"><td>rs_ids</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">A comma-separated list of rs ids for variants present in dbsnp</div>
</div>
</td>
</tr>
<tr class="row-even"><td>in_hm2</td>
<td>BOOL</td>
<td>Whether the variant was part of HapMap2.</td>
</tr>
<tr class="row-odd"><td>in_hm3</td>
<td>BOOL</td>
<td>Whether the variant was part of HapMap3.</td>
</tr>
<tr class="row-even"><td>in_esp</td>
<td>BOOL</td>
<td>Presence/absence of the variant in the ESP project data</td>
</tr>
<tr class="row-odd"><td>in_1kg</td>
<td>BOOL</td>
<td>Presence/absence of the variant in the 1000 genome project data (phase 3)</td>
</tr>
<tr class="row-even"><td>aaf_esp_ea</td>
<td>FLOAT</td>
<td>Minor Allele Frequency of the variant for European Americans in the ESP project</td>
</tr>
<tr class="row-odd"><td>aaf_esp_aa</td>
<td>FLOAT</td>
<td>Minor Allele Frequency of the variant for African Americans in the ESP project</td>
</tr>
<tr class="row-even"><td>aaf_esp_all</td>
<td>FLOAT</td>
<td>Minor Allele Frequency of the variant w.r.t both groups in the ESP project</td>
</tr>
<tr class="row-odd"><td>aaf_1kg_amr</td>
<td>FLOAT</td>
<td>Allele frequency of the variant in AMR population based on AC/AN (1000g project, phase 3)</td>
</tr>
<tr class="row-even"><td>aaf_1kg_eas</td>
<td>FLOAT</td>
<td>Allele frequency of the variant in EAS population based on AC/AN (1000g project, phase 3)</td>
</tr>
<tr class="row-odd"><td>aaf_1kg_sas</td>
<td>FLOAT</td>
<td>Allele frequency of the variant in SAS population based on AC/AN (1000g project, phase 3)</td>
</tr>
<tr class="row-even"><td>aaf_1kg_afr</td>
<td>FLOAT</td>
<td>Allele frequency of the variant in AFR population based on AC/AN (1000g project, phase 3)</td>
</tr>
<tr class="row-odd"><td>aaf_1kg_eur</td>
<td>FLOAT</td>
<td>Allele frequency of the variant in EUR population based on AC/AN (1000g project, phase 3)</td>
</tr>
<tr class="row-even"><td>aaf_1kg_all</td>
<td>FLOAT</td>
<td>Global allele frequency (based on AC/AN) (1000g project - phase 3)</td>
</tr>
<tr class="row-odd"><td>in_exac</td>
<td>BOOL</td>
<td>Presence/absence of the variant in ExAC (Exome Aggregation Consortium) data (Broad)</td>
</tr>
<tr class="row-even"><td>aaf_exac_all</td>
<td>FLOAT</td>
<td>Raw allele frequency (population independent) of the variant based on ExAC exomes (AF)</td>
</tr>
<tr class="row-odd"><td>aaf_adj_exac_all</td>
<td>FLOAT</td>
<td>Adjusted allele frequency (population independent) of the variant based on ExAC (Adj_AC/Adj_AN)</td>
</tr>
<tr class="row-even"><td>aaf_adj_exac_afr</td>
<td>FLOAT</td>
<td>Adjusted allele frequency of the variant for AFR population in ExAC (AC_AFR/AN_AFR)</td>
</tr>
<tr class="row-odd"><td>aaf_adj_exac_amr</td>
<td>FLOAT</td>
<td>Adjusted allele frequency of the variant for AMR population in ExAC (AC_AMR/AN_AMR)</td>
</tr>
<tr class="row-even"><td>aaf_adj_exac_eas</td>
<td>FLOAT</td>
<td>Adjusted allele frequency of the variant for EAS population in ExAC (AC_EAS/AN_EAS)</td>
</tr>
<tr class="row-odd"><td>aaf_adj_exac_fin</td>
<td>FLOAT</td>
<td>Adjusted allele frequency of the variant for FIN population in ExAC (AC_FIN/AN_FIN)</td>
</tr>
<tr class="row-even"><td>aaf_adj_exac_nfe</td>
<td>FLOAT</td>
<td>Adjusted allele frequency of the variant for NFE population in ExAC (AC_NFE/AN_NFE)</td>
</tr>
<tr class="row-odd"><td>aaf_adj_exac_oth</td>
<td>FLOAT</td>
<td>Adjusted allele frequency of the variant for OTH population in ExAC (AC_OTH/AN_OTH)</td>
</tr>
<tr class="row-even"><td>aaf_adj_exac_sas</td>
<td>FLOAT</td>
<td>Adjusted allele frequency of the variant for SAS population in ExAC (AC_SAS/AN_SAS)</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="disease-phenotype-info-from-clinvar">
<h3>Disease phenotype info (from ClinVar).<a class="headerlink" href="#disease-phenotype-info-from-clinvar" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="6%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>in_omim</td>
<td>BOOL</td>
<td><div class="first last line-block">
<div class="line">0 : Absence of the variant in OMIM database</div>
<div class="line">1 : Presence of the variant in OMIM database</div>
</div>
</td>
</tr>
<tr class="row-odd"><td>clinvar_causal_allele</td>
<td>STRING</td>
<td>The allele(s) that are associated or causal for the disease.</td>
</tr>
<tr class="row-even"><td>clinvar_sig</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">The clinical significance scores for each</div>
<div class="line">of the variant according to ClinVar:</div>
<div class="line"><em>unknown</em>, <em>untested</em>, <em>non-pathogenic</em></div>
<div class="line"><em>probable-non-pathogenic</em>, <em>probable-pathogenic</em></div>
<div class="line"><em>pathogenic</em>, <em>drug-response</em>, <em>histocompatibility</em></div>
<div class="line"><em>other</em></div>
</div>
</td>
</tr>
<tr class="row-odd"><td>clinvar_disease_name</td>
<td>STRING</td>
<td>The name of the disease to which the variant is relevant</td>
</tr>
<tr class="row-even"><td>clinvar_dbsource</td>
<td>STRING</td>
<td>Variant Clinical Channel IDs</td>
</tr>
<tr class="row-odd"><td>clinvar_dbsource_id</td>
<td>STRING</td>
<td>The record id in the above database</td>
</tr>
<tr class="row-even"><td>clinvar_origin</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">The type of variant.</div>
<div class="line">Any of:</div>
<div class="line"><em>unknown</em>, <em>germline</em>, <em>somatic</em>,</div>
<div class="line"><em>inherited</em>, <em>paternal</em>, <em>maternal</em>,</div>
<div class="line"><em>de-novo</em>, <em>biparental</em>, <em>uniparental</em>,</div>
<div class="line"><em>not-tested</em>, <em>tested-inconclusive</em>,</div>
<div class="line"><em>other</em></div>
</div>
</td>
</tr>
<tr class="row-odd"><td>clinvar_dsdb</td>
<td>STRING</td>
<td>Variant disease database name</td>
</tr>
<tr class="row-even"><td>clinvar_dsdbid</td>
<td>STRING</td>
<td>Variant disease database ID</td>
</tr>
<tr class="row-odd"><td>clinvar_disease_acc</td>
<td>STRING</td>
<td>Variant Accession and Versions</td>
</tr>
<tr class="row-even"><td>clinvar_in_locus_spec_db</td>
<td>BOOL</td>
<td>Submitted from a locus-specific database?</td>
</tr>
<tr class="row-odd"><td>clinvar_on_diag_assay</td>
<td>BOOL</td>
<td>Variation is interrogated in a clinical diagnostic assay?</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="structural-variation-columns">
<h3>Structural variation columns<a class="headerlink" href="#structural-variation-columns" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="6%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>sv_cipos_start_left</td>
<td>INTEGER</td>
<td>The leftmost position of the leftmost SV breakpoint confidence interval.</td>
</tr>
<tr class="row-odd"><td>sv_cipos_end_left</td>
<td>INTEGER</td>
<td>The rightmost position of the leftmost SV breakpoint confidence interval.</td>
</tr>
<tr class="row-even"><td>sv_cipos_start_right</td>
<td>INTEGER</td>
<td>The leftmost position of the rightmost SV breakpoint confidence interval.</td>
</tr>
<tr class="row-odd"><td>sv_cipos_end_right</td>
<td>INTEGER</td>
<td>The rightmost position of the rightmost SV breakpoint confidence interval.</td>
</tr>
<tr class="row-even"><td>sv_length</td>
<td>INTEGER</td>
<td>The length of the structural variant in base pairs.</td>
</tr>
<tr class="row-odd"><td>sv_is_precise</td>
<td>BOOL</td>
<td>Is the structural variant precise (i.e., to 1-bp resolution)?</td>
</tr>
<tr class="row-even"><td>sv_tool</td>
<td>STRING</td>
<td>The name of the SV discovery tool used to find the SV.</td>
</tr>
<tr class="row-odd"><td>sv_evidence_type</td>
<td>STRING</td>
<td>What type of alignment evidence supports the SV?</td>
</tr>
<tr class="row-even"><td>sv_event_id</td>
<td>STRING</td>
<td>A unique identifier for the SV.</td>
</tr>
<tr class="row-odd"><td>sv_mate_id</td>
<td>STRING</td>
<td>The ID for the &#8220;other end&#8221; of the SV.</td>
</tr>
<tr class="row-even"><td>sv_strand</td>
<td>STRING</td>
<td>The orientations of the SV breakpoint(s).</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="genome-annotations">
<h3>Genome annotations<a class="headerlink" href="#genome-annotations" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="5%" />
<col width="81%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>exome_chip</td>
<td>BOOL</td>
<td>Whether an SNP is on the Illumina HumanExome Chip</td>
</tr>
<tr class="row-odd"><td>cyto_band</td>
<td>STRING</td>
<td>Chromosomal cytobands that a variant overlaps</td>
</tr>
<tr class="row-even"><td>rmsk</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">A comma-separated list of RepeatMasker annotations that the variant overlaps.</div>
<div class="line">Each hit is of the form: <tt class="docutils literal"><span class="pre">name_class_family</span></tt></div>
</div>
</td>
</tr>
<tr class="row-odd"><td>in_cpg_island</td>
<td>BOOL</td>
<td><div class="first last line-block">
<div class="line">Does the variant overlap a CpG island?.</div>
<div class="line">Based on UCSC: Regulation &gt; CpG Islands &gt; cpgIslandExt</div>
</div>
</td>
</tr>
<tr class="row-even"><td>in_segdup</td>
<td>BOOL</td>
<td><div class="first last line-block">
<div class="line">Does the variant overlap a segmental duplication?.</div>
<div class="line">Based on UCSC: Variation&amp;Repeats &gt; Segmental Dups &gt; genomicSuperDups track</div>
</div>
</td>
</tr>
<tr class="row-odd"><td>is_conserved</td>
<td>BOOL</td>
<td><div class="first last line-block">
<div class="line">Does the variant overlap a conserved region?</div>
<div class="line">Based on the 29-way mammalian conservation study</div>
</div>
</td>
</tr>
<tr class="row-even"><td>gerp_bp_score</td>
<td>FLOAT</td>
<td><div class="first last line-block">
<div class="line">GERP conservation score.</div>
<div class="line">Only populated if the <tt class="docutils literal"><span class="pre">--load-gerp-bp</span></tt> option is used when loading.</div>
<div class="line">Higher scores reflect greater conservation. <strong>At base-pair resolution</strong>.</div>
<div class="line">Details: <a class="reference external" href="http://mendel.stanford.edu/SidowLab/downloads/gerp/">http://mendel.stanford.edu/SidowLab/downloads/gerp/</a></div>
</div>
</td>
</tr>
<tr class="row-odd"><td>gerp_element_pval</td>
<td>FLOAT</td>
<td><div class="first last line-block">
<div class="line">GERP elements P-val</div>
<div class="line">Lower P-values scores reflect greater conservation. <strong>Not at base-pair resolution</strong>.</div>
<div class="line">Details: <a class="reference external" href="http://mendel.stanford.edu/SidowLab/downloads/gerp/">http://mendel.stanford.edu/SidowLab/downloads/gerp/</a></div>
</div>
</td>
</tr>
<tr class="row-even"><td>recomb_rate</td>
<td>FLOAT</td>
<td><div class="first last line-block">
<div class="line">Returns the mean recombination rate at the variant site</div>
<div class="line">Based on HapMapII_GRCh37 genetic map</div>
</div>
</td>
</tr>
<tr class="row-odd"><td>cadd_raw</td>
<td>FLOAT</td>
<td><div class="first last line-block">
<div class="line">Raw <tt class="docutils literal"><span class="pre">CADD</span></tt> scores for scoring deleteriousness of SNV&#8217;s in the human genome</div>
<div class="line">Details: <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24487276">http://www.ncbi.nlm.nih.gov/pubmed/24487276</a></div>
</div>
</td>
</tr>
<tr class="row-even"><td>cadd_scaled</td>
<td>FLOAT</td>
<td><div class="first last line-block">
<div class="line">Scaled <tt class="docutils literal"><span class="pre">CADD</span></tt> scores (Phred like) for scoring deleteriousness of SNV&#8217;s</div>
<div class="line">Details: <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/24487276">http://www.ncbi.nlm.nih.gov/pubmed/24487276</a></div>
</div>
</td>
</tr>
<tr class="row-odd"><td>fitcons</td>
<td>FLOAT</td>
<td><div class="first last line-block">
<div class="line">fitCons scores estimating the probability that a point mutation</div>
<div class="line">at each position in a genome will influence fitness.</div>
<div class="line">Higher scores have more potential for interesting genomic function.</div>
<div class="line">Common ranges: 0.05-0.35 for non-coding and 0.4-0.8 for coding</div>
<div class="line">Provides integerated highly significant scores (i6-0).</div>
<div class="line"><a class="reference external" href="http://haldanessieve.org/2014/09/12/our-paper-probabilities-of-fitness-consequences-for-point-mutations-across-the-human-genome/">http://haldanessieve.org/2014/09/12/our-paper-probabilities-of-fitness-consequences-for-point-mutations-across-the-human-genome/</a></div>
<div class="line"><a class="reference external" href="http://biorxiv.org/content/early/2014/09/11/006825">http://biorxiv.org/content/early/2014/09/11/006825</a></div>
<div class="line"><a class="reference external" href="http://genome-mirror.bscb.cornell.edu/cgi-bin/hgTrackUi?hgsid=37121_t2bCeEgzaA9R4buMtIv5jbR0uLhZ&amp;g=fc1">http://genome-mirror.bscb.cornell.edu/cgi-bin/hgTrackUi?hgsid=37121_t2bCeEgzaA9R4buMtIv5jbR0uLhZ&amp;g=fc1</a></div>
</div>
</td>
</tr>
</tbody>
</table>
<p><strong>Note:</strong>
<tt class="docutils literal"><span class="pre">CADD</span></tt> scores (<a class="reference external" href="http://cadd.gs.washington.edu/">http://cadd.gs.washington.edu/</a>) are Copyright 2013 University of Washington and Hudson-Alpha Institute for Biotechnology
(all rights reserved) but are freely available for all academic, non-commercial applications. For commercial licensing information contact
Jennifer McCullar (<a class="reference external" href="mailto:mccullaj&#37;&#52;&#48;uw&#46;edu">mccullaj<span>&#64;</span>uw<span>&#46;</span>edu</a>).</p>
</div>
<div class="section" id="variant-error-assessment">
<h3>Variant error assessment<a class="headerlink" href="#variant-error-assessment" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="6%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>grc</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">Association with patch and fix regions from the Genome Reference Consortium:</div>
<div class="line"><a class="reference external" href="http://www.ncbi.nlm.nih.gov/projects/genome/assembly/grc/human/">http://www.ncbi.nlm.nih.gov/projects/genome/assembly/grc/human/</a></div>
<div class="line">Identifies potential problem regions associated with variant calls.</div>
<div class="line">Built with <cite>annotation_provenance/make-ncbi-grc-patches.py</cite></div>
</div>
</td>
</tr>
<tr class="row-odd"><td>gms_illumina</td>
<td>FLOAT</td>
<td><div class="first last line-block">
<div class="line">Genome Mappability Scores (GMS) for Illumina error models</div>
<div class="line">Provides low GMS scores (&lt; 25.0 in any technology) from:</div>
<div class="line"><a class="reference external" href="http://sourceforge.net/apps/mediawiki/gma-bio/index.php?title=Download_GMS">http://sourceforge.net/apps/mediawiki/gma-bio/index.php?title=Download_GMS</a></div>
<div class="line">#Download_GMS_by_Chromosome_and_Sequencing_Technology</div>
<div class="line">Input VCF for annotations prepared with:</div>
<div class="line"><a class="reference external" href="https://github.com/chapmanb/bcbio.variation/blob/master/src/bcbio/variation/utils/gms.clj">https://github.com/chapmanb/bcbio.variation/blob/master/src/bcbio/variation/utils/gms.clj</a></div>
</div>
</td>
</tr>
<tr class="row-even"><td>gms_solid</td>
<td>FLOAT</td>
<td>Genome Mappability Scores with SOLiD error models</td>
</tr>
<tr class="row-odd"><td>gms_iontorrent</td>
<td>FLOAT</td>
<td>Genome Mappability Scores with IonTorrent error models</td>
</tr>
<tr class="row-even"><td>in_cse</td>
<td>BOOL</td>
<td><div class="first last line-block">
<div class="line">Is a variant in an error prone genomic position,</div>
<div class="line">using CSE: Context-Specific Sequencing Errors</div>
<div class="line"><a class="reference external" href="https://code.google.com/p/discovering-cse/">https://code.google.com/p/discovering-cse/</a></div>
<div class="line"><a class="reference external" href="http://www.biomedcentral.com/1471-2105/14/S5/S1">http://www.biomedcentral.com/1471-2105/14/S5/S1</a></div>
</div>
</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="encode-information">
<h3>ENCODE information<a class="headerlink" href="#encode-information" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="18%" />
<col width="6%" />
<col width="76%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>encode_tfbs</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">Comma-separated list of transcription factors that were</div>
<div class="line">observed by ENCODE to bind DNA in this region.  Each hit in the list is constructed</div>
<div class="line">as TF_CELLCOUNT, where:</div>
<div class="line-block">
<div class="line"><em>TF</em> is the transcription factor name</div>
<div class="line"><em>CELLCOUNT</em> is the number of cells tested that had nonzero signals.</div>
</div>
<div class="line">Provenance: wgEncodeRegTfbsClusteredV2 UCSC table</div>
</div>
</td>
</tr>
<tr class="row-odd"><td>encode_dnaseI_cell_count</td>
<td>INTEGER</td>
<td><div class="first last line-block">
<div class="line">Count of cell types that were observed to have DnaseI hypersensitivity.</div>
</div>
</td>
</tr>
<tr class="row-even"><td>encode_dnaseI_cell_list</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">Comma separated list of cell types that were observed to have DnaseI hypersensitivity.</div>
<div class="line">Provenance: Thurman, et al, <em>Nature</em>, 489, pp. 75-82, 5 Sep. 2012</div>
</div>
</td>
</tr>
<tr class="row-odd"><td>encode_consensus_gm12878</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">ENCODE consensus segmentation prediction for GM12878.</div>
<div class="line"><br /></div>
<div class="line">CTCF: CTCF-enriched element</div>
<div class="line">E:    Predicted enhancer</div>
<div class="line">PF:   Predicted promoter flanking region</div>
<div class="line">R:    Predicted repressed or low-activity region</div>
<div class="line">TSS:  Predicted promoter region including TSS</div>
<div class="line">T:    Predicted transcribed region</div>
<div class="line">WE:   Predicted weak enhancer or open chromatin cis-regulatory element
| unknown: This region of the genome had no functional prediction.</div>
</div>
</td>
</tr>
<tr class="row-even"><td>encode_consensus_h1hesc</td>
<td>STRING</td>
<td>ENCODE consensus segmentation prediction for h1HESC.  See encode_consseg_gm12878 for details.</td>
</tr>
<tr class="row-odd"><td>encode_consensus_helas3</td>
<td>STRING</td>
<td>ENCODE consensus segmentation prediction for Helas3.  See encode_consseg_gm12878 for details.</td>
</tr>
<tr class="row-even"><td>encode_consensus_hepg2</td>
<td>STRING</td>
<td>ENCODE consensus segmentation prediction for HEPG2.   See encode_consseg_gm12878 for details.</td>
</tr>
<tr class="row-odd"><td>encode_consensus_huvec</td>
<td>STRING</td>
<td>ENCODE consensus segmentation prediction for HuVEC.   See encode_consseg_gm12878 for details.</td>
</tr>
<tr class="row-even"><td>encode_consensus_k562</td>
<td>STRING</td>
<td>ENCODE consensus segmentation prediction for k562.    See encode_consseg_gm12878 for details.</td>
</tr>
<tr class="row-odd"><td>vista_enhancers</td>
<td>STRING</td>
<td>Experimentally validated human enhancers from VISTA (<a class="reference external" href="http://enhancer.lbl.gov/frnt_page_n.shtml">http://enhancer.lbl.gov/frnt_page_n.shtml</a>)</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="cancer-related-columns">
<h3>Cancer related columns<a class="headerlink" href="#cancer-related-columns" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="6%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
<th class="head">&nbsp;</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>is_somatic</td>
<td>BOOL</td>
<td><div class="first last line-block">
<div class="line">Whether the variant is somatically acquired.</div>
</div>
</td>
</tr>
<tr class="row-odd"><td>cosmic_ids</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">A list of known COSMIC ids for this variant.</div>
<div class="line">See: <a class="reference external" href="http://cancer.sanger.ac.uk/cancergenome/projects/cosmic/">http://cancer.sanger.ac.uk/cancergenome/projects/cosmic/</a></div>
</div>
</td>
</tr>
</tbody>
</table>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>
<div class="section" id="the-variant-impacts-table">
<h2>The <tt class="docutils literal"><span class="pre">variant_impacts</span></tt> table<a class="headerlink" href="#the-variant-impacts-table" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="7%" />
<col width="79%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">column_name</th>
<th class="head">type</th>
<th class="head">notes</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>variant_id</td>
<td>INTEGER</td>
<td>PRIMARY_KEY (Foreign key to <cite>variants</cite> table)</td>
</tr>
<tr class="row-odd"><td>anno_id</td>
<td>INTEGER</td>
<td>PRIMARY_KEY (Based on variant transcripts)</td>
</tr>
<tr class="row-even"><td>gene</td>
<td>STRING</td>
<td>The gene affected by the variant.</td>
</tr>
<tr class="row-odd"><td>transcript</td>
<td>STRING</td>
<td>The transcript affected by the variant.</td>
</tr>
<tr class="row-even"><td>is_exonic</td>
<td>BOOL</td>
<td>Does the variant affect an exon for this transcript?</td>
</tr>
<tr class="row-odd"><td>is_coding</td>
<td>BOOL</td>
<td>Does the variant fall in a coding region (excludes 3&#8217; &amp; 5&#8217; UTR&#8217;s of exons)?</td>
</tr>
<tr class="row-even"><td>is_lof</td>
<td>BOOL</td>
<td>Based on the value of the impact col, is the variant LOF?</td>
</tr>
<tr class="row-odd"><td>exon</td>
<td>STRING</td>
<td>Exon information for the variants that are exonic</td>
</tr>
<tr class="row-even"><td>codon_change</td>
<td>STRING</td>
<td>What is the codon change?</td>
</tr>
<tr class="row-odd"><td>aa_change</td>
<td>STRING</td>
<td>What is the amino acid change?</td>
</tr>
<tr class="row-even"><td>aa_length</td>
<td>STRING</td>
<td>The length of CDS in terms of number of amino acids (<tt class="docutils literal"><span class="pre">snpEff</span> <span class="pre">only</span></tt>)</td>
</tr>
<tr class="row-odd"><td>biotype</td>
<td>STRING</td>
<td>The type of transcript (e.g.protein-coding, pseudogene, rRNA etc.) (<tt class="docutils literal"><span class="pre">SnpEff</span> <span class="pre">only</span></tt>)</td>
</tr>
<tr class="row-even"><td>impact</td>
<td>STRING</td>
<td>Impacts due to variation (ref.impact category)</td>
</tr>
<tr class="row-odd"><td>impact_so</td>
<td>STRING</td>
<td>The sequence ontology term for the impact</td>
</tr>
<tr class="row-even"><td>impact_severity</td>
<td>STRING</td>
<td>Severity of the impact based on the impact column value (ref.impact category)</td>
</tr>
<tr class="row-odd"><td>polyphen_pred</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">Impact of the SNP as given by PolyPhen (<tt class="docutils literal"><span class="pre">VEP</span> <span class="pre">only</span></tt>)</div>
<div class="line">benign, possibly_damaging, probably_damaging, unknown</div>
</div>
</td>
</tr>
<tr class="row-even"><td>polyphen_scores</td>
<td>FLOAT</td>
<td>Polyphen score reflecting severity (higher the impact, <em>higher</em> the score) (<tt class="docutils literal"><span class="pre">VEP</span> <span class="pre">only</span></tt>)</td>
</tr>
<tr class="row-odd"><td>sift_pred</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">Impact of the SNP as given by SIFT (<tt class="docutils literal"><span class="pre">VEP</span> <span class="pre">only</span></tt>)</div>
<div class="line">neutral, deleterious</div>
</div>
</td>
</tr>
<tr class="row-even"><td>sift_scores</td>
<td>FLOAT</td>
<td>SIFT prob. scores reflecting severity (Higher the impact, <em>lower</em> the score) (<tt class="docutils literal"><span class="pre">VEP</span> <span class="pre">only</span></tt>)</td>
</tr>
</tbody>
</table>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="details-of-the-impact-and-impact-severity-columns">
<h2>Details of the <tt class="docutils literal"><span class="pre">impact</span></tt> and <tt class="docutils literal"><span class="pre">impact_severity</span></tt> columns<a class="headerlink" href="#details-of-the-impact-and-impact-severity-columns" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="36%" />
<col width="45%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">impact severity</th>
<th class="head">impacts</th>
<th class="head">SO_impacts</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>HIGH</td>
<td><ul class="first last simple">
<li>exon_deleted</li>
<li>frame_shift</li>
<li>splice_acceptor</li>
<li>splice_donor</li>
<li>start_loss</li>
<li>stop_gain</li>
<li>stop_loss</li>
<li>non_synonymous_start</li>
<li>transcript_codon_change</li>
<li>rare_amino_acid</li>
<li>chrom_large_del</li>
</ul>
</td>
<td><ul class="first last simple">
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001572">exon_loss_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001589">frameshift_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001574">splice_acceptor_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001575">splice_donor_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002012">start_lost</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001587">stop_gained</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001578">stop_lost</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001582">initiator_codon_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001582">initiator_codon_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002008">rare_amino_acid_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:1000029">chromosomal_deletion</a></li>
</ul>
</td>
</tr>
<tr class="row-odd"><td>MED</td>
<td><ul class="first last simple">
<li>non_syn_coding</li>
<li>inframe_codon_gain</li>
<li>inframe_codon_loss</li>
<li>inframe_codon_change</li>
<li>codon_change_del</li>
<li>codon_change_ins</li>
<li>UTR_5_del</li>
<li>UTR_3_del</li>
<li>splice_region</li>
<li>mature_miRNA</li>
<li>regulatory_region</li>
<li>TF_binding_site</li>
<li>regulatory_region_ablation</li>
<li>regulatory_region_amplification</li>
<li>TFBS_ablation</li>
<li>TFBS_amplification</li>
</ul>
</td>
<td><ul class="first last simple">
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001583">missense_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001821">inframe_insertion</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001822">inframe_deletion</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001580">coding_sequence_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001826">disruptive_inframe_deletion</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001824">disruptive_inframe_insertion</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002013">5_prime_UTR_truncation</a> + <a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001572">exon_loss_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002015">3_prime_UTR_truncation</a> + <a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001572">exon_loss_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001630">splice_region_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001620">mature_miRNA_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001566">regulatory_region_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001782">TF_binding_site_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001894">regulatory_region_ablation</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001891">regulatory_region_amplification</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001895">TFBS_ablation</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001892">TFBS_amplification</a></li>
</ul>
</td>
</tr>
<tr class="row-even"><td>LOW</td>
<td><ul class="first last simple">
<li>synonymous_stop</li>
<li>synonymous_coding</li>
<li>UTR_5_prime</li>
<li>UTR_3_prime</li>
<li>intron</li>
<li>CDS</li>
<li>upstream</li>
<li>downstream</li>
<li>intergenic</li>
<li>intragenic</li>
<li>gene</li>
<li>transcript</li>
<li>exon</li>
<li>start_gain</li>
<li>synonymous_start</li>
<li>intron_conserved</li>
<li>nc_transcript</li>
<li>NMD_transcript</li>
<li>incomplete_terminal_codon</li>
<li>nc_exon</li>
<li>transcript_ablation</li>
<li>transcript_amplification</li>
<li>feature elongation</li>
<li>feature truncation</li>
</ul>
</td>
<td><ul class="first last simple">
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001567">stop_retained_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001819">synonymous_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001623">5_prime_UTR_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001624">3_prime_UTR_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001627">intron_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001580">coding_sequence_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001631">upstream_gene_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001632">downstream_gene_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001628">intergenic_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002011">intragenic_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001564">gene_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001576">transcript_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001791">exon_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001988">5_prime_UTR_premature_start_codon_gain_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002019">start_retained_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0002018">conserved_intron_variant</a></li>
<li><a class="reference external" href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001619">nc_transcript_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001621">NMD_transcript_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001626">incomplete_terminal_codon_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001792">non_coding_exon_variant</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001893">transcript_ablation</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001889">transcript_amplification</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001907">feature_elongation</a></li>
<li><a class="reference external" href="http://sequenceontology.org/browser/current_svn/term/SO:0001906">feature_truncation</a></li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="the-samples-table">
<h2>The <tt class="docutils literal"><span class="pre">samples</span></tt> table<a class="headerlink" href="#the-samples-table" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="13%" />
<col width="10%" />
<col width="76%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">column name</th>
<th class="head">type</th>
<th class="head">notes</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>sample_id</td>
<td>INTEGER</td>
<td>PRIMARY_KEY</td>
</tr>
<tr class="row-odd"><td>name</td>
<td>STRING</td>
<td>Sample names</td>
</tr>
<tr class="row-even"><td>family_id</td>
<td>INTEGER</td>
<td>Family ids for the samples [User defined, default: NULL]</td>
</tr>
<tr class="row-odd"><td>paternal_id</td>
<td>INTEGER</td>
<td>Paternal id for the samples [User defined, default: NULL]</td>
</tr>
<tr class="row-even"><td>maternal_id</td>
<td>INTEGER</td>
<td>Maternal id for the samples [User defined, default: NULL]</td>
</tr>
<tr class="row-odd"><td>sex</td>
<td>STRING</td>
<td>Sex of the sample [User defined, default: NULL]</td>
</tr>
<tr class="row-even"><td>phenotype</td>
<td>STRING</td>
<td>The associated sample phenotype [User defined, default: NULL]</td>
</tr>
<tr class="row-odd"><td>ethnicity</td>
<td>STRING</td>
<td>The ethnic group to which the sample belongs [User defined, default: NULL]</td>
</tr>
</tbody>
</table>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="the-resources-table">
<h2>The <tt class="docutils literal"><span class="pre">resources</span></tt> table<a class="headerlink" href="#the-resources-table" title="Permalink to this headline">¶</a></h2>
<p>Establishes provenance of annotation resources used to create a Gemini database.</p>
<table border="1" class="docutils">
<colgroup>
<col width="18%" />
<col width="14%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">column name</th>
<th class="head">type</th>
<th class="head">notes</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>name</td>
<td>STRING</td>
<td>Name of the annotation type</td>
</tr>
<tr class="row-odd"><td>resource</td>
<td>STRING</td>
<td>Filename of the resource, with version information</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="the-version-table">
<h2>The <tt class="docutils literal"><span class="pre">version</span></tt> table<a class="headerlink" href="#the-version-table" title="Permalink to this headline">¶</a></h2>
<p>Establishes which version of <tt class="docutils literal"><span class="pre">gemini</span></tt> was used to create a database.</p>
<table border="1" class="docutils">
<colgroup>
<col width="18%" />
<col width="14%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">column name</th>
<th class="head">type</th>
<th class="head">notes</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>version</td>
<td>STRING</td>
<td>What version of gemini was used to create the DB.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="the-gene-detailed-table">
<h2>The <tt class="docutils literal"><span class="pre">gene_detailed</span></tt> table<a class="headerlink" href="#the-gene-detailed-table" title="Permalink to this headline">¶</a></h2>
<p>Built on version 75 of Ensembl genes</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="7%" />
<col width="79%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">column_name</th>
<th class="head">type</th>
<th class="head">notes</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>uid</td>
<td>INTEGER</td>
<td>PRIMARY_KEY (unique identifier for each entry in the table)</td>
</tr>
<tr class="row-odd"><td>chrom</td>
<td>STRING</td>
<td>The chromosome on which the gene resides</td>
</tr>
<tr class="row-even"><td>gene</td>
<td>STRING</td>
<td>The gene name</td>
</tr>
<tr class="row-odd"><td>is_hgnc</td>
<td>BOOL</td>
<td>Flag for gene column: 0 for non HGNC symbol and 1 for HGNC symbol = TRUE</td>
</tr>
<tr class="row-even"><td>ensembl_gene_id</td>
<td>STRING</td>
<td>The ensembl gene id for the gene</td>
</tr>
<tr class="row-odd"><td>transcript</td>
<td>STRING</td>
<td>The ensembl transcript id for the gene</td>
</tr>
<tr class="row-even"><td>biotype</td>
<td>STRING</td>
<td>The biotype (e.g protein coding) of the transcript</td>
</tr>
<tr class="row-odd"><td>transcript_status</td>
<td>STRING</td>
<td>The status of the transcript (e.g. KNOWN, PUTATIVE etc.)</td>
</tr>
<tr class="row-even"><td>ccds_id</td>
<td>STRING</td>
<td>The consensus coding sequence transcript identifier</td>
</tr>
<tr class="row-odd"><td>hgnc_id</td>
<td>STRING</td>
<td>The HGNC identifier for the gene if HGNC symbol is TRUE</td>
</tr>
<tr class="row-even"><td>entrez_id</td>
<td>STRING</td>
<td>The entrez gene identifier for the gene</td>
</tr>
<tr class="row-odd"><td>cds_length</td>
<td>STRING</td>
<td>The length of CDS in bases</td>
</tr>
<tr class="row-even"><td>protein_length</td>
<td>STRING</td>
<td>The length of the transcript as the number of amino acids</td>
</tr>
<tr class="row-odd"><td>transcript_start</td>
<td>STRING</td>
<td>The start position of the transcript in bases</td>
</tr>
<tr class="row-even"><td>transcript_end</td>
<td>STRING</td>
<td>The end position of the transcript in bases</td>
</tr>
<tr class="row-odd"><td>strand</td>
<td>STRING</td>
<td>The strand of DNA where the gene resides</td>
</tr>
<tr class="row-even"><td>synonym</td>
<td>STRING</td>
<td>Other gene names (previous or synonyms) for the gene</td>
</tr>
<tr class="row-odd"><td>rvis_pct</td>
<td>FLOAT</td>
<td>The RVIS percentile values for the gene</td>
</tr>
<tr class="row-even"><td>mam_phenotype_id</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">High level mammalian phenotype ID applied to mouse phenotype descriptions</div>
<div class="line">in the MGI database at <a class="reference external" href="http://www.informatics.jax.org/">http://www.informatics.jax.org/</a>. Data taken from
<a class="reference external" href="ftp://ftp.informatics.jax.org/pub/reports/HMD_HumanPhenotype.rpt">ftp://ftp.informatics.jax.org/pub/reports/HMD_HumanPhenotype.rpt</a></div>
</div>
</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="the-gene-summary-table">
<h2>The <tt class="docutils literal"><span class="pre">gene_summary</span></tt> table<a class="headerlink" href="#the-gene-summary-table" title="Permalink to this headline">¶</a></h2>
<p>Built on version 75 of Ensembl genes</p>
<table border="1" class="docutils">
<colgroup>
<col width="17%" />
<col width="6%" />
<col width="77%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">column_name</th>
<th class="head">type</th>
<th class="head">notes</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>uid</td>
<td>INTEGER</td>
<td>PRIMARY_KEY (unique identifier for each entry in the table)</td>
</tr>
<tr class="row-odd"><td>chrom</td>
<td>STRING</td>
<td>The chromosome on which the gene resides</td>
</tr>
<tr class="row-even"><td>gene</td>
<td>STRING</td>
<td>The gene name</td>
</tr>
<tr class="row-odd"><td>is_hgnc</td>
<td>BOOL</td>
<td>Flag for gene column: 0 for non HGNC symbol and 1 for HGNC symbol = TRUE</td>
</tr>
<tr class="row-even"><td>ensembl_gene_id</td>
<td>STRING</td>
<td>The ensembl gene id for the gene</td>
</tr>
<tr class="row-odd"><td>hgnc_id</td>
<td>STRING</td>
<td>The HGNC identifier for the gene if HGNC symbol is TRUE</td>
</tr>
<tr class="row-even"><td>transcript_min_start</td>
<td>STRING</td>
<td>The minimum start position of all transcripts for the gene</td>
</tr>
<tr class="row-odd"><td>transcript_max_end</td>
<td>STRING</td>
<td>The maximum end position of all transcripts for the gene</td>
</tr>
<tr class="row-even"><td>strand</td>
<td>STRING</td>
<td>The strand of DNA where the gene resides</td>
</tr>
<tr class="row-odd"><td>synonym</td>
<td>STRING</td>
<td>Other gene names (previous or synonyms) for the gene</td>
</tr>
<tr class="row-even"><td>rvis_pct</td>
<td>FLOAT</td>
<td>The RVIS percentile values for the gene</td>
</tr>
<tr class="row-odd"><td>mam_phenotype_id</td>
<td>STRING</td>
<td><div class="first last line-block">
<div class="line">High level mammalian phenotype ID applied to mouse phenotype descriptions</div>
<div class="line">in the MGI database at <a class="reference external" href="http://www.informatics.jax.org/">http://www.informatics.jax.org/</a>. Data taken from
<a class="reference external" href="ftp://ftp.informatics.jax.org/pub/reports/HMD_HumanPhenotype.rpt">ftp://ftp.informatics.jax.org/pub/reports/HMD_HumanPhenotype.rpt</a></div>
</div>
</td>
</tr>
<tr class="row-even"><td>in_cosmic_census</td>
<td>BOOL</td>
<td>Are mutations in the gene implicated in cancer by the cancer gene census?</td>
</tr>
</tbody>
</table>
</div>
</div>




  

<div id="disqus_thread"></div>
 <script type="text/javascript">
     /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
     var disqus_shortname = 'geminidocs'; // required: replace example with your forum shortname

     /* * * DON'T EDIT BELOW THIS LINE * * */
     (function() {
         var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
         dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
         (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
     })();
 </script>
 <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
 <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>






          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/gemini.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">The Gemini database schema</a><ul>
<li><a class="reference internal" href="#the-variants-table">The <tt class="docutils literal"><span class="pre">variants</span></tt> table</a><ul>
<li><a class="reference internal" href="#core-vcf-fields">Core VCF fields</a></li>
<li><a class="reference internal" href="#variant-and-popgen-info">Variant and PopGen info</a></li>
<li><a class="reference internal" href="#genotype-information">Genotype information</a></li>
<li><a class="reference internal" href="#gene-information">Gene information</a></li>
<li><a class="reference internal" href="#optional-vcf-info-fields">Optional VCF INFO fields</a></li>
<li><a class="reference internal" href="#population-information">Population information</a></li>
<li><a class="reference internal" href="#disease-phenotype-info-from-clinvar">Disease phenotype info (from ClinVar).</a></li>
<li><a class="reference internal" href="#structural-variation-columns">Structural variation columns</a></li>
<li><a class="reference internal" href="#genome-annotations">Genome annotations</a></li>
<li><a class="reference internal" href="#variant-error-assessment">Variant error assessment</a></li>
<li><a class="reference internal" href="#encode-information">ENCODE information</a></li>
<li><a class="reference internal" href="#cancer-related-columns">Cancer related columns</a></li>
</ul>
</li>
<li><a class="reference internal" href="#the-variant-impacts-table">The <tt class="docutils literal"><span class="pre">variant_impacts</span></tt> table</a></li>
<li><a class="reference internal" href="#details-of-the-impact-and-impact-severity-columns">Details of the <tt class="docutils literal"><span class="pre">impact</span></tt> and <tt class="docutils literal"><span class="pre">impact_severity</span></tt> columns</a></li>
<li><a class="reference internal" href="#the-samples-table">The <tt class="docutils literal"><span class="pre">samples</span></tt> table</a></li>
<li><a class="reference internal" href="#the-resources-table">The <tt class="docutils literal"><span class="pre">resources</span></tt> table</a></li>
<li><a class="reference internal" href="#the-version-table">The <tt class="docutils literal"><span class="pre">version</span></tt> table</a></li>
<li><a class="reference internal" href="#the-gene-detailed-table">The <tt class="docutils literal"><span class="pre">gene_detailed</span></tt> table</a></li>
<li><a class="reference internal" href="#the-gene-summary-table">The <tt class="docutils literal"><span class="pre">gene_summary</span></tt> table</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="browser.html"
                        title="previous chapter">The GEMINI browser interface</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="api.html"
                        title="next chapter">Using the GEMINI API</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/database_schema.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="api.html" title="Using the GEMINI API"
             >next</a> |</li>
        <li class="right" >
          <a href="browser.html" title="The GEMINI browser interface"
             >previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2012,2013.
      Last updated on Mar 19, 2015.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>

<!-- 
<div id="editor-trap">
      <h3>Edit and improve this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/arq5x/gemini/blob/master/docs/content/database_schema.rst">
          The Gemini database schema
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/arq5x/gemini/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>
 -->
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', UA-39150563-1]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>


  </body>
</html>