<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Quick start</title>
    
    <link rel="stylesheet" href="../_static/labibi.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.12.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/gemini.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="Annotation with snpEff or VEP" href="functional_annotation.html" />
    <link rel="prev" title="Installation" href="installation.html" />


<script src="../_static/labibi.js"></script>



<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', '']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>


  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="functional_annotation.html" title="Annotation with snpEff or VEP"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  
  <div class="section" id="quick-start">
<h1>Quick start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h1>
<p><cite>gemini</cite> is designed to allow researchers to explore genetic variation contained
in a <a class="reference external" href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">VCF</a> file.
The basic workflow for working with <cite>gemini</cite> is outlined below.</p>
<div class="section" id="importing-vcf-files-into-gemini">
<h2>Importing VCF files into gemini.<a class="headerlink" href="#importing-vcf-files-into-gemini" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>we now recommend splitting variants with multiple alternates and left-aligning,
and trimming all variants before loading into gemini.</p>
<p class="last">See <a class="reference internal" href="preprocessing.html#preprocess"><em>Step 1. split, left-align, and trim variants</em></a> for a detailed explanation.</p>
</div>
<p>Before we can use GEMINI to explore genetic variation, we must first <tt class="docutils literal"><span class="pre">load</span></tt> our
VCF file into the GEMINI database framework.  We expect you to have first
annotated the functional consequence of each variant in your VCF using either
VEP or snpEff (Note that v3.0+ of snpEff is required to track the amino acid
length of each impacted transcript). Logically,the loading step is done with
the <tt class="docutils literal"><span class="pre">gemini</span> <span class="pre">load</span></tt> command.  Below are two examples based on a VCF file that
we creatively name my.vcf.  The first example assumes that the VCF has been
pre-annotated with VEP and the second assumes snpEff.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># VEP-annotated VCF</span>
<span class="nv">$ </span>gemini load -v my.vcf -t VEP my.db

<span class="c"># snpEff-annotated VCF</span>
<span class="nv">$ </span>gemini load -v my.vcf -t snpEff my.db
</pre></div>
</div>
<p>Assuming you have a valid VCF file produced by standard variation discovery
programs (e.g., GATK, FreeBayes, etc.), one loads the VCF into the gemini
framework with the <strong>load</strong> submodule:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini load -v my.vcf my.db
</pre></div>
</div>
<p>In this step, <cite>gemini</cite> reads and loads the my.vcf file into a SQLite database
named my.db, whose structure is described <a class="reference external" href="http://gemini.readthedocs.org/en/latest/content/database_schema.html">here</a>.
While loading the database, <cite>gemini</cite> computes many additional population genetics
statistics that support downstream analyses. It also stores the genotypes for
each sample at each variant in an efficient data structure that minimizes the
database size.</p>
<p>Loading is by far the slowest aspect of GEMINI.  Using multiple CPUs can
greatly speed up this process.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini load -v my.vcf --cores 8 my.db
</pre></div>
</div>
</div>
<div class="section" id="querying-the-gemini-database">
<h2>Querying the <cite>gemini</cite> database.<a class="headerlink" href="#querying-the-gemini-database" title="Permalink to this headline">¶</a></h2>
<p>If you are familiar with SQL, <tt class="docutils literal"><span class="pre">gemini</span></tt> allows you to directly query the database
in search of interesting variants via the <cite>-q</cite> option.
For example, here is a query to identify all novel, loss-of-function variants
in your database:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select * from variants where is_lof = 1 and in_dbsnp = 0&quot;</span> my.db
</pre></div>
</div>
<p>Or, we can ask for all variants that substantially deviate from
Hardy-Weinberg equilibrium:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select * from variants where hwe &lt; 0.01&quot;</span> my.db
</pre></div>
</div>
</div>
</div>




  

<div id="disqus_thread"></div>
 <script type="text/javascript">
     /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
     var disqus_shortname = 'geminidocs'; // required: replace example with your forum shortname

     /* * * DON'T EDIT BELOW THIS LINE * * */
     (function() {
         var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
         dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
         (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
     })();
 </script>
 <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
 <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>






          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/gemini.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Quick start</a><ul>
<li><a class="reference internal" href="#importing-vcf-files-into-gemini">Importing VCF files into gemini.</a></li>
<li><a class="reference internal" href="#querying-the-gemini-database">Querying the <cite>gemini</cite> database.</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="installation.html"
                        title="previous chapter">Installation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="functional_annotation.html"
                        title="next chapter">Annotation with snpEff or VEP</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/quick_start.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="functional_annotation.html" title="Annotation with snpEff or VEP"
             >next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             >previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2012,2013.
      Last updated on Mar 19, 2015.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>

<!-- 
<div id="editor-trap">
      <h3>Edit and improve this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/arq5x/gemini/blob/master/docs/content/quick_start.rst">
          Quick start
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/arq5x/gemini/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>
 -->
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', UA-39150563-1]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>


  </body>
</html>