<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>SamSifter Manual &mdash; SamSifter 0.5.0 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.5.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="SamSifter 0.5.0 documentation" href="index.html" />
    <link rel="up" title="SamSifter Help" href="help.html" />
    <link rel="next" title="SamSifter API documentation" href="api.html" />
    <link rel="prev" title="SamSifter Help" href="help.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="api.html" title="SamSifter API documentation"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="help.html" title="SamSifter Help"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">SamSifter 0.5.0 documentation</a> &raquo;</li>
          <li><a href="help.html" accesskey="U">SamSifter Help</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="samsifter-manual">
<span id="manual"></span><h1><a class="toc-backref" href="#id4">SamSifter Manual</a><a class="headerlink" href="#samsifter-manual" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#samsifter-manual" id="id4">SamSifter Manual</a><ul>
<li><a class="reference internal" href="#detailed-installation-instructions" id="id5">Detailed Installation Instructions</a><ul>
<li><a class="reference internal" href="#installing-python3-and-pyqt4" id="id6">Installing <tt class="docutils literal"><span class="pre">Python3</span></tt> and <tt class="docutils literal"><span class="pre">PyQt4</span></tt></a><ul>
<li><a class="reference internal" href="#debian-8-jessie-and-newer-or-ubuntu-14-04-and-newer" id="id7">Debian 8 (jessie) and newer or Ubuntu 14.04 and newer</a></li>
<li><a class="reference internal" href="#older-debian-or-ubuntu-systems" id="id8">Older Debian or Ubuntu systems</a></li>
<li><a class="reference internal" href="#mac-os-x-using-homebrew" id="id9">Mac OS X using Homebrew</a></li>
<li><a class="reference internal" href="#windows-32-bit" id="id10">Windows (32 bit)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#installing-samsifter" id="id11">Installing SamSifter</a></li>
<li><a class="reference internal" href="#uninstalling-samsifter" id="id12">Uninstalling SamSifter</a></li>
</ul>
</li>
<li><a class="reference internal" href="#creating-and-editing-workflows" id="id13">Creating and Editing Workflows</a><ul>
<li><a class="reference internal" href="#selecting-an-input-file" id="id14">Selecting an Input File</a></li>
<li><a class="reference internal" href="#adding-and-removing-a-tool-or-filter" id="id15">Adding and Removing a Tool or Filter</a></li>
<li><a class="reference internal" href="#rearranging-tools-and-filters" id="id16">Rearranging Tools and Filters</a></li>
<li><a class="reference internal" href="#displaying-and-editing-filter-settings" id="id17">Displaying and Editing Filter Settings</a></li>
<li><a class="reference internal" href="#selecting-an-output-file" id="id18">Selecting an Output File</a><ul>
<li><a class="reference internal" href="#optional-creating-an-rma-file-for-megan" id="id19">Optional: Creating an RMA file for MEGAN</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#executing-workflows" id="id20">Executing Workflows</a><ul>
<li><a class="reference internal" href="#running-a-workflow" id="id21">Running a Workflow</a></li>
<li><a class="reference internal" href="#log-messages" id="id22">Log Messages</a></li>
<li><a class="reference internal" href="#stopping-a-workflow" id="id23">Stopping a Workflow</a></li>
</ul>
</li>
<li><a class="reference internal" href="#loading-and-saving-workflows" id="id24">Loading and Saving Workflows</a></li>
<li><a class="reference internal" href="#export-to-bash-scripts" id="id25">Export to Bash scripts</a></li>
<li><a class="reference internal" href="#manuals-for-filters-and-tools" id="id26">Manuals for Filters and Tools</a><ul>
<li><a class="reference internal" href="#common-options" id="id27">Common Options</a></li>
<li><a class="reference internal" href="#count-reads-per-taxon" id="id28">Count reads per taxon</a></li>
<li><a class="reference internal" href="#filter-reads-by-conservation" id="id29">Filter reads by conservation</a></li>
<li><a class="reference internal" href="#filter-reads-by-list-of-qnames" id="id30">Filter reads by list of QNAMES</a></li>
<li><a class="reference internal" href="#filter-references-by-evenness-of-coverage" id="id31">Filter references by evenness of coverage</a><ul>
<li><a class="reference internal" href="#what-is-the-gini-coefficient" id="id32">What is the Gini coefficient?</a></li>
<li><a class="reference internal" href="#experimental-options" id="id33">Experimental options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#filter-references-by-identity" id="id34">Filter references by % identity</a></li>
<li><a class="reference internal" href="#filter-references-by-list-of-accessions" id="id35">Filter references by list of accessions</a></li>
<li><a class="reference internal" href="#filter-references-by-pmd-score" id="id36">Filter references by PMD score</a></li>
<li><a class="reference internal" href="#filter-taxa-by-list-of-taxon-ids" id="id37">Filter taxa by list of taxon IDs</a></li>
<li><a class="reference internal" href="#filter-taxa-by-pmd-score" id="id38">Filter taxa by PMD score</a></li>
<li><a class="reference internal" href="#gnu-gzip-compress-file" id="id39">[GNU gzip] Compress file</a></li>
<li><a class="reference internal" href="#gnu-gzip-decompress-file" id="id40">[GNU gzip] Decompress file</a></li>
<li><a class="reference internal" href="#pmdtools-calculate-pmds" id="id41">[PMDtools] Calculate PMDS</a></li>
<li><a class="reference internal" href="#pmdtools-filter-reads-by-identity" id="id42">[PMDtools] Filter reads by % identity</a></li>
<li><a class="reference internal" href="#pmdtools-filter-reads-by-pmds" id="id43">[PMDtools] Filter reads by PMDS</a></li>
<li><a class="reference internal" href="#samtools-convert-bam-to-sam" id="id44">[SAMtools] Convert BAM to SAM</a></li>
<li><a class="reference internal" href="#samtools-convert-sam-to-bam" id="id45">[SAMtools] Convert SAM to BAM</a></li>
<li><a class="reference internal" href="#samtools-remove-duplicates" id="id46">[SAMtools] Remove duplicates</a></li>
<li><a class="reference internal" href="#samtools-sort-by-coordinates" id="id47">[SAMtools] Sort by coordinates</a></li>
<li><a class="reference internal" href="#samtools-sort-by-query-names" id="id48">[SAMtools] Sort by query names</a></li>
<li><a class="reference internal" href="#betterrmdup-remove-duplicates" id="id49">[BetterRMDup] Remove duplicates</a></li>
</ul>
</li>
<li><a class="reference internal" href="#statistical-summaries" id="id50">Statistical Summaries</a><ul>
<li><a class="reference internal" href="#compiling-results-per-analysed-file" id="id51">Compiling Results Per Analysed File</a></li>
<li><a class="reference internal" href="#comparison-of-multiple-files" id="id52">Comparison of Multiple Files</a></li>
<li><a class="reference internal" href="#enriching-summaries-with-external-databases" id="id53">Enriching Summaries with External Databases</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="detailed-installation-instructions">
<span id="installation"></span><h2><a class="toc-backref" href="#id5">Detailed Installation Instructions</a><a class="headerlink" href="#detailed-installation-instructions" title="Permalink to this headline">¶</a></h2>
<p>SamSifter is written in Python3, so it runs on Windows, Mac OS X and GNU/Linux
as long as you have installed a working Python3 interpreter.
It requires the additional packages <tt class="docutils literal"><span class="pre">PyQt4</span></tt> to display the GUI, <tt class="docutils literal"><span class="pre">numpy</span></tt>
for vectorized calculations, <tt class="docutils literal"><span class="pre">pandas</span></tt> for statistical summaries, and
<tt class="docutils literal"><span class="pre">matplotlib</span></tt> to plot
optional coverage and read lengths distributions. The Python setup tools normally
take care of these requirements for you, however at time of writing the package
<tt class="docutils literal"><span class="pre">PyQt4</span></tt> is not available in the
<a class="reference external" href="http://https://pypi.python.org/pypi">PyPI repositories</a> so you have to
install it using your operating system&#8217;s package management.</p>
<p>If you already have a working Python3 installation with all required packages
you can skip the following section.</p>
<div class="section" id="installing-python3-and-pyqt4">
<h3><a class="toc-backref" href="#id6">Installing <tt class="docutils literal"><span class="pre">Python3</span></tt> and <tt class="docutils literal"><span class="pre">PyQt4</span></tt></a><a class="headerlink" href="#installing-python3-and-pyqt4" title="Permalink to this headline">¶</a></h3>
<p>Below you can find tested installation instructions for Debian-based
<strong>GNU/Linux</strong> distributions. The package names for other Linux distributions
should be very similar though. The <strong>Windows</strong> installation has only been tested
on a 32-bit Windows XP installation (currently the only test system available).
The installation instructions for <strong>Mac OS X</strong> are likely incomplete as they
could not be tested at all.</p>
<div class="section" id="debian-8-jessie-and-newer-or-ubuntu-14-04-and-newer">
<h4><a class="toc-backref" href="#id7">Debian 8 (jessie) and newer or Ubuntu 14.04 and newer</a><a class="headerlink" href="#debian-8-jessie-and-newer-or-ubuntu-14-04-and-newer" title="Permalink to this headline">¶</a></h4>
<p>The preparation of the SamSifter installation on Debian 8 (jessie)
and newer or Ubuntu 14.04 and newer is very simple as all packages are already
available in the standard repositories:</p>
<div class="highlight-python"><div class="highlight"><pre>sudo aptitude install python3 python3-dev python3-setuptools python3-nose python3-pyqt4 python3-numpy python3-matplotlib python3-pandas
</pre></div>
</div>
<p>You can now proceed with the actual SamSifter installation.</p>
</div>
<div class="section" id="older-debian-or-ubuntu-systems">
<h4><a class="toc-backref" href="#id8">Older Debian or Ubuntu systems</a><a class="headerlink" href="#older-debian-or-ubuntu-systems" title="Permalink to this headline">¶</a></h4>
<p>Older Debian/Ubuntu systems provide only a rudimentary Python3 environment
lacking the <tt class="docutils literal"><span class="pre">matplotlib</span></tt> package. We have to use the Python tools
<tt class="docutils literal"><span class="pre">easy_install</span></tt> and <tt class="docutils literal"><span class="pre">pip</span></tt> to obtain it. The following steps have been tested
successfully on Ubuntu 12.04:</p>
<div class="highlight-python"><div class="highlight"><pre># install Python3 system (as far as possible)
sudo aptitude install python3 python3-dev python3-setuptools python3-nose python3-pyqt4 python3-numpy python3-tornado libfreetype6-dev
# install pip for Python3
sudo easy_install3 -U distribute
sudo easy_install3 pip
# matplotlib is not yet available, thus we install it from PyPI
sudo pip3 install matplotlib==1.3.1
# same for pandas
sudo pip3 install pandas==0.14.1
</pre></div>
</div>
<p>You can now proceed with the actual SamSifter installation.</p>
</div>
<div class="section" id="mac-os-x-using-homebrew">
<h4><a class="toc-backref" href="#id9">Mac OS X using Homebrew</a><a class="headerlink" href="#mac-os-x-using-homebrew" title="Permalink to this headline">¶</a></h4>
<p>Using <a class="reference external" href="http://brew.sh">Homebrew</a>, the &#8220;missing package manager for OS X&#8221;,
seems to be the easiest way to obtain an up to date <tt class="docutils literal"><span class="pre">Python3</span></tt> system including
<tt class="docutils literal"><span class="pre">PyQt4</span></tt> on a Mac. Install <tt class="docutils literal"><span class="pre">Homebrew</span></tt> by pasting the following command into a
Terminal prompt:</p>
<div class="highlight-python"><div class="highlight"><pre>ruby -e &quot;$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)&quot;
</pre></div>
</div>
<p>and follow the on-screen instructions. Once you have checked that the system is
functional with</p>
<div class="highlight-python"><div class="highlight"><pre>brew doctor
</pre></div>
</div>
<p>you can install these packages with the following commands (current versions at
time of release according to <a class="reference external" href="http://braumeister.org/search">Braumeister</a>):</p>
<div class="highlight-python"><div class="highlight"><pre>brew install python3        # current version: 3.4.2
brew install pyqt           # current version: 4.11.3
brew install pkg-config     # current version: 0.28
</pre></div>
</div>
<p>Open a fresh Terminal window, confirm that it uses the newly installed Python3
interpreter at <cite>/usr/local/bin/python</cite> by running</p>
<div class="highlight-python"><div class="highlight"><pre>which python
</pre></div>
</div>
<p>and continue to install additional Python packages with the built-in package
management:</p>
<div class="highlight-python"><div class="highlight"><pre># install pip
easy_install -U distribute
easy_install pip
# install setuptools, numpy and matplotlib
pip install setuptools
pip install numpy
pip install matplotlib==1.3.1
pip install pandas==0.14.1
</pre></div>
</div>
<p>You can now proceed with the actual SamSifter installation.</p>
</div>
<div class="section" id="windows-32-bit">
<h4><a class="toc-backref" href="#id10">Windows (32 bit)</a><a class="headerlink" href="#windows-32-bit" title="Permalink to this headline">¶</a></h4>
<p>For any Windows system the use of a packaged Python 3.4 distribution like <tt class="docutils literal"><span class="pre">Anaconda</span></tt>
from <a class="reference external" href="http://continuum.io/downloads#all">http://continuum.io/downloads#all</a> is recommended. However, if you&#8217;d rather
install the individual packages by yourself you can follow these steps.
The following instructions have been tested successfully on Windows XP (32 bit):</p>
<ol class="arabic simple">
<li>Download and install Python from <a class="reference external" href="https://www.python.org/downloads/windows/">https://www.python.org/downloads/windows/</a>.
The recommended version is
<a class="reference external" href="https://www.python.org/ftp/python/3.4.2/python-3.4.2.msi">Python 3.4.2</a>.
During installation, make sure to include <tt class="docutils literal"><span class="pre">pip</span></tt> in the installation and
check the option to automatically add <tt class="docutils literal"><span class="pre">python.exe</span></tt> to your <tt class="docutils literal"><span class="pre">PATH</span></tt>.</li>
<li>Download and install <tt class="docutils literal"><span class="pre">PyQt4</span></tt> from
<a class="reference external" href="http://www.riverbankcomputing.com/software/pyqt/download">http://www.riverbankcomputing.com/software/pyqt/download</a>.
The recommended version is
<a class="reference external" href="http://sourceforge.net/projects/pyqt/files/PyQt4/PyQt-4.11.3/PyQt4-4.11.3-gpl-Py3.4-Qt4.8.6-x32.exe">PyQt 4.11.3 for Python 3.4 and Qt 4.8.6</a>.</li>
<li>Download and install <tt class="docutils literal"><span class="pre">numpy</span></tt> from <a class="reference external" href="http://www.numpy.org/">http://www.numpy.org/</a>. The recommended
version is
<a class="reference external" href="http://downloads.sourceforge.net/project/numpy/NumPy/1.9.1/numpy-1.9.1-win32-superpack-python3.4.exe">numpy 1.9.1 for Python 3.4</a>.</li>
<li>Download and install <tt class="docutils literal"><span class="pre">matplotlib</span></tt> from <a class="reference external" href="http://www.matplotlib.org/">http://www.matplotlib.org/</a>. The
recommended version is <a class="reference external" href="http://downloads.sourceforge.net/project/matplotlib/matplotlib/matplotlib-1.4.0/matplotlib-1.4.0.win32-py3.4.exe">matplotlib 1.4.0 for Python 3.4</a>.</li>
<li>Download and install <tt class="docutils literal"><span class="pre">pandas</span></tt> from
<a class="reference external" href="http://www.lfd.uci.edu/~gohlke/pythonlibs/#pandas">http://www.lfd.uci.edu/~gohlke/pythonlibs/#pandas</a>. The recommended version is
<tt class="docutils literal"><span class="pre">pandas</span></tt> 0.15.2 for Python 3.4.</li>
</ol>
<p>You can now proceed with the actual SamSifter installation.</p>
</div>
</div>
<div class="section" id="installing-samsifter">
<h3><a class="toc-backref" href="#id11">Installing SamSifter</a><a class="headerlink" href="#installing-samsifter" title="Permalink to this headline">¶</a></h3>
<p>Once you have a working Python3 environment use an administrator account to
download and install the current SamSifter package on your system with the
command:</p>
<div class="highlight-python"><div class="highlight"><pre>pip3 install SamSifter
</pre></div>
</div>
<p>Previous versions will be uninstalled automatically. The installation can be
tested with</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">samsifter</span> <span class="o">--</span><span class="n">help</span>
</pre></div>
</div>
<p>If everything went fine you should see the following help text:</p>
<div class="highlight-python"><div class="highlight"><pre>usage: samsifter [-h] [-v] [-d]

SamSifter helps you create filter workflows for next-generation sequencing
data. It is primarily used to process SAM files generated by MALT prior to
metagenomic analysis in MEGAN.

optional arguments:
  -h, --help     show this help message and exit
  -v, --verbose  print additional information to stderr
  -d, --debug    show debug options in menu
</pre></div>
</div>
<p>Starting the program without any arguments will display the GUI and let you
edit your first workflow.</p>
</div>
<div class="section" id="uninstalling-samsifter">
<h3><a class="toc-backref" href="#id12">Uninstalling SamSifter</a><a class="headerlink" href="#uninstalling-samsifter" title="Permalink to this headline">¶</a></h3>
<p>To get rid of SamSifter simply execute the following command as administrator:</p>
<div class="highlight-python"><div class="highlight"><pre>pip3 uninstall SamSifter
</pre></div>
</div>
<p>The Python utility <tt class="docutils literal"><span class="pre">pip3</span></tt> will list all currently installed
available versions for selective removal. Proceed similarly with any required
packages (e.g. <tt class="docutils literal"><span class="pre">matplotlib</span></tt>) if you don&#8217;t need them anymore.</p>
</div>
</div>
<div class="section" id="creating-and-editing-workflows">
<span id="edit-workflows"></span><h2><a class="toc-backref" href="#id13">Creating and Editing Workflows</a><a class="headerlink" href="#creating-and-editing-workflows" title="Permalink to this headline">¶</a></h2>
<p>Workflows are displayed and edited in the central part of the SamSifter window:</p>
<a class="reference external image-reference" href="_images/example_workflow.png"><img alt="The workflow editor with an example workflow to filter SAM files for ancient human pathogens." class="align-center" src="_images/example_workflow.png" style="width: 232.0px; height: 249.5px;" /></a>
<p>A workflow consists of three parts:</p>
<ul class="simple">
<li>An input file in SAM or BAM format (optionally zipped) which can be
specified in the upper part of the main window.</li>
<li>A list of one or more tools and filters that are executed in a specific order.
They are shown in the central part of the main window and can be freely
rearranged, added and removed.</li>
<li>An output file in SAM or BAM format (again optionally zipped to save space),
which similar to the input file can be specified in the lower part of the main
window. This part also contains options to enable the generation of RMA files
for MEGAN.</li>
</ul>
<p>When working with SamSifter for the first time all of these fields and lists
will be empty and waiting for you. To get back to this pristine state simply
click the <strong>[New]</strong> button to create a new empty workflow. Any changes to your
current workflow will be discarded unless you follow the prompt to save them.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#save-workflows"><em>Loading and Saving Workflows</em></a></p>
</div>
<p>We&#8217;ll now take a closer look at each of these three parts.</p>
<div class="section" id="selecting-an-input-file">
<h3><a class="toc-backref" href="#id14">Selecting an Input File</a><a class="headerlink" href="#selecting-an-input-file" title="Permalink to this headline">¶</a></h3>
<p>SamSifter supports input files in the SAM or BAM format, both compressed or
uncompressed. Some of the filters included in SamSifter may have additional
requirements for the sort order of the reads or optional annotations like
headers and MD tags. The input file can either be determined by typing the full
path into the text field or by clicking the <strong>[Open...]</strong> button and selecting
the file using a standard file open dialog.</p>
<p>SamSifter automatically checks if the file a) exists, b) is readable and c) has
the correct file type. The text field containing the file path will be
highlighted in orange to show you if any of these conditions are not fulfilled.
Hovering the mouse cursor over the text field shows a tooltip with hints how
to resolve the error.</p>
<a class="reference external image-reference" href="_images/input_error.png"><img alt="An incomplete filename causes input errors that are shown by orange highlighting and tooltips." class="align-center" src="_images/input_error.png" style="width: 261.0px; height: 55.5px;" /></a>
</div>
<div class="section" id="adding-and-removing-a-tool-or-filter">
<h3><a class="toc-backref" href="#id15">Adding and Removing a Tool or Filter</a><a class="headerlink" href="#adding-and-removing-a-tool-or-filter" title="Permalink to this headline">¶</a></h3>
<p>SamSifter includes a variety of filters and tools that can be used to process
your input file. You can add one to the workflow by selecting it from either the
<strong>[Edit] &gt; [Add Filter...]</strong> menu or double-clicking an entry in the
<strong>Filters and Tools</strong> dock.</p>
<a class="reference external image-reference" href="_images/filters_dock.png"><img alt="The Filters and Tools dock lists all available filters and tools by category. Press F2 to show it or activate it in the **[View]** menu." class="align-center" src="_images/filters_dock.png" style="width: 150.0px; height: 249.0px;" /></a>
<p>You can toggle the dock on and off by pressing F2 or activating it in the
<strong>[View]</strong> menu.</p>
<p>The new entry is added to the currently selected row of your workflow or
appended to the end if no row is selected. To remove the entry click
<strong>[Edit] &gt; [Delete]</strong> or press the Delete key.</p>
</div>
<div class="section" id="rearranging-tools-and-filters">
<h3><a class="toc-backref" href="#id16">Rearranging Tools and Filters</a><a class="headerlink" href="#rearranging-tools-and-filters" title="Permalink to this headline">¶</a></h3>
<p>The order of steps in the workflow can be freely rearranged by selecting any
entry and moving it up or down. To move an entry click <strong>[Edit] &gt; [Move up]</strong> /
<strong>[Edit] &gt; [Move down]</strong>, click the corresponding toolbar buttons, or use the
keyboard shortcuts Ctrl+Up and Ctrl+Down.</p>
<p>SamSifter validates the entire workflow after each change. Individual steps
in the workflow may have specific requirements for the type of input they accept
and/or the type of output they produce. If any of these requirements are not
fulfilled the color of the corresponding entry will change to orange and a
message will appear in the <strong>Messages</strong> dock with hints on how to resolve these
errors.</p>
<a class="reference external image-reference" href="_images/messages_dock.png"><img alt="The Messages dock shows error messages and hints on how to resolve them. Press F3 to show it or activate it in the **[View]** menu." class="align-center" src="_images/messages_dock.png" style="width: 255.0px; height: 106.0px;" /></a>
<p>You can toggle the dock on and off by pressing F3 or activating it in the
<strong>[View]</strong> menu.</p>
</div>
<div class="section" id="displaying-and-editing-filter-settings">
<h3><a class="toc-backref" href="#id17">Displaying and Editing Filter Settings</a><a class="headerlink" href="#displaying-and-editing-filter-settings" title="Permalink to this headline">¶</a></h3>
<p>Most tools and filters in your workflow have settings that are required for the
tool to function properly as well as optional settings that can tweak the
functionality. The <strong>Filter Settings</strong> dock serves to show these settings along
with a description of the tool and a summary of its input and output
requirements.</p>
<a class="reference external image-reference" href="_images/settings_dock.png"><img alt="The Filter Settings dock shows error messages and hints on how to resolve them. Press F4 to show it, or activate it in the **[View]** menu, or double-click any workflow entry." class="align-center" src="_images/settings_dock.png" style="width: 150.0px; height: 142.0px;" /></a>
<p>You can toggle the dock on and off by pressing F4, activating it in the
<strong>[View]</strong> menu or simply double-clicking the workflow step you are interested
in. The individual settings have tooltips that provide more information on their
meaning and effect.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">The available options depend on the tool or filter. Please see
<a class="reference internal" href="#filter-manuals"><em>Manuals for Filters and Tools</em></a> for details.</p>
</div>
</div>
<div class="section" id="selecting-an-output-file">
<h3><a class="toc-backref" href="#id18">Selecting an Output File</a><a class="headerlink" href="#selecting-an-output-file" title="Permalink to this headline">¶</a></h3>
<p>SamSifter supports output files in the SAM and BAM formats, both compressed and
uncompressed. The output file can either be determined by typing the full
path into the text field or by clicking the <strong>[Save as...]</strong> button and selecting
the new filename using a standard file save dialog.</p>
<p>Similar to the input field, the program automatically checks if the target
directory a) exists, b) is readable and c) if the new file has the correct file
type. The text field containing the file path will be highlighted in orange to
show you if any of these conditions are not fulfilled. Hovering the mouse cursor
over the text field shows a tooltip with hints how to resolve the error.</p>
<a class="reference external image-reference" href="_images/output_error.png"><img alt="A non-existing directory causes output errors that are shown by orange highlighting and tooltips." class="align-center" src="_images/output_error.png" style="width: 384.0px; height: 55.5px;" /></a>
<div class="section" id="optional-creating-an-rma-file-for-megan">
<h4><a class="toc-backref" href="#id19">Optional: Creating an RMA file for MEGAN</a><a class="headerlink" href="#optional-creating-an-rma-file-for-megan" title="Permalink to this headline">¶</a></h4>
<p>SamSifter can automatically create a RMA file from SAM output files if a recent
version of MEGAN (version 5.8.3 or newer) including the SAM2RMA tool is
installed on your computer. The tool can be configured by clicking the
<strong>[Settings...]</strong> button.</p>
<a class="reference external image-reference" href="_images/sam2rma_dialog.png"><img alt="The SAM2RMA dialog is used to change the settings for RMA file creation." class="align-center" src="_images/sam2rma_dialog.png" style="width: 254.0px; height: 127.0px;" /></a>
<p>For details on the effect of the individual setting see the MEGAN manual or the
SAM2RMA help function. Clicking the <strong>[Restore defaults]</strong> button will revert
your changes to the default settings.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The default settings here are recommended by the Krause Lab and may
differ from the MEGAN default settings!</p>
</div>
<p>If the path to the SAM2RMA binary is highlighted orange you may have to update
the installation path or change the file permissions to make the program
executable.</p>
</div>
</div>
</div>
<div class="section" id="executing-workflows">
<span id="execute-workflows"></span><h2><a class="toc-backref" href="#id20">Executing Workflows</a><a class="headerlink" href="#executing-workflows" title="Permalink to this headline">¶</a></h2>
<p>Once a workflow is properly set up with input file, a list of tools or filters
to run and an output file you can run the entire workflow on your computer.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#bash-export"><em>Export to Bash scripts</em></a> if you&#8217;d like to process more than one file at a
time or run the workflow using the command line of another
computer.</p>
</div>
<div class="section" id="running-a-workflow">
<h3><a class="toc-backref" href="#id21">Running a Workflow</a><a class="headerlink" href="#running-a-workflow" title="Permalink to this headline">¶</a></h3>
<p>To run a workflow simply click the <strong>[Run]</strong> button in the toolbar or choose
<strong>[Run] &gt; [Run]</strong> from the menu. This button may be disabled in two situations:</p>
<ul class="simple">
<li>if your workflow has errors that need to be addressed -&gt; take a closer look at
entries with orange highlighting to resolve the errors, then try again.</li>
<li>if another workflow is currently being processed -&gt; wait for it to finish or
click <strong>[Stop]</strong> to stop it, then try again.</li>
</ul>
</div>
<div class="section" id="log-messages">
<h3><a class="toc-backref" href="#id22">Log Messages</a><a class="headerlink" href="#log-messages" title="Permalink to this headline">¶</a></h3>
<p>Once the workflow is being executed the <strong>[Messages]</strong> dock will pop up and show
a bunch of messages:</p>
<a class="reference external image-reference" href="_images/workflow_progress.png"><img alt="A workflow produces messages to inform about warnings and errors." class="align-center" src="_images/workflow_progress.png" style="width: 339.5px; height: 207.0px;" /></a>
<p>Depending on your workflow and settings there will be</p>
<ul class="simple">
<li>Status messages in black, eg. the current status and process ID.</li>
<li>Info messages in green; they&#8217;re enabled by default but can be disabled for
each step by disabling the individual <tt class="docutils literal"><span class="pre">verbose</span></tt> options if you don&#8217;t like
the noise.</li>
<li>Warnings in yellow; they&#8217;re usually harmless and SamSifter will continue to
process the workflow but you may want to review these warnings to see if they
affect the expected results.</li>
<li>Errors in red; SamSifter will attempt to abort the processing immediately if
it encounters an error.</li>
</ul>
</div>
<div class="section" id="stopping-a-workflow">
<h3><a class="toc-backref" href="#id23">Stopping a Workflow</a><a class="headerlink" href="#stopping-a-workflow" title="Permalink to this headline">¶</a></h3>
<p>If you notice that you made a mistake or that the processing doesn&#8217;t work as
expected you can abort the run by pressing the <strong>[Stop]</strong> button at any time.
Depending on your workflow this may leave temporary files in your work directory
that should be removed manually to not affect the next run.</p>
</div>
</div>
<div class="section" id="loading-and-saving-workflows">
<span id="save-workflows"></span><h2><a class="toc-backref" href="#id24">Loading and Saving Workflows</a><a class="headerlink" href="#loading-and-saving-workflows" title="Permalink to this headline">¶</a></h2>
<p>SamSifter uses a simple XML file type with the extension <tt class="docutils literal"><span class="pre">ssx</span></tt> (SamSifter XML)
that is human- and machine-readable can be opened and edited in any text editor.</p>
<p>Use the <strong>[Open]</strong> button or the corresponding entry <strong>[File] &gt; [Open]</strong> in the
menu to open a workflow file located on your computer or network.</p>
<p>To save your workflow to a SSX file you can click the <strong>[Save]</strong> button in the
toolbar or choose the corresponding entry <strong>[File] &gt; [Save]</strong> in the menu.
Similarly, use <strong>[File] &gt; [Save as...]</strong> to save a copy of your current workflow
to a new file with different filename.</p>
<p>The <strong>[File]</strong> menu lists up to 10 recently used files for quicker access.</p>
</div>
<div class="section" id="export-to-bash-scripts">
<span id="bash-export"></span><h2><a class="toc-backref" href="#id25">Export to Bash scripts</a><a class="headerlink" href="#export-to-bash-scripts" title="Permalink to this headline">¶</a></h2>
<p>If processing one SAM file at a time is not enough anymore and you want to move
to large-scale processing of many files with an established workflow, maybe even
on bigger hardware with more CPU cores and RAM, then exporting your workflow
to a Bash script may be the next step.</p>
<p>All tools and filters included in SamSifter can also be executed as stand-alone
scripts on the commandline interface (CLI). To create a Bash script that can
call these commands similar to the graphical user interface (the SamSifter GUI)
simply click the <strong>[Export]</strong> button or select it from the
<strong>[Run] &gt; [Export to Bash]</strong> menu. A dialog with several options will pop up,
allowing you to choose between three distinct types of processing modes and
giving you three addtional settings for Bash options (eg. to print all executed
commands, to shorten long file paths to just the filename and/or to stop on
errors):</p>
<a class="reference external image-reference" href="_images/bash_dialog.png"><img alt="The Bash Export dialog offering three processing modes and three additional settings." class="align-center" src="_images/bash_dialog.png" style="width: 294.0px; height: 213.0px;" /></a>
<p>The available processing modes are:</p>
<ol class="arabic simple">
<li><strong>Single Mode</strong> processes only the specified input file and produces the
specified output file similar to running the workflow in the GUI. This
mode should be used with unmodified filenames unless all input files are
located in the same directory as the exported bash script.</li>
<li><strong>Sequential Mode</strong> processes a list of arbitrary input files one file
after another and saves the output to files renamed with the filename
extension <tt class="docutils literal"><span class="pre">sifted</span></tt>. This mode should be used with shortened filenames
unless all of the required list files (CSV) are available at identical
paths from all machines that this script is deployed to.</li>
<li><strong>Parallel Mode</strong> speeds up the process significantly by distributing jobs
across all available CPU cores and running them in parallel. This requires
the installation of GNU <cite>parallel</cite>. Similar to sequential mode above it
should be used with shortened filenames unless all of the required list
files (CSV) are available at identical paths from all machines that this
script is deployed to.</li>
</ol>
<p>Clicking the <strong>[Save]</strong> button lets you choose the name and location of the new
Bash script. Of course you can open this script in any text editor and modify it
to fit your needs. To execute the script you need to make it executable first:</p>
<div class="highlight-python"><div class="highlight"><pre>chmod +x example.sh
</pre></div>
</div>
<p>then run it with the command</p>
<div class="highlight-python"><div class="highlight"><pre>./example.sh one.sam two.sam three.sam
</pre></div>
</div>
<p>to process three different input files with the same example workflow. The number
of input files is unlimited and they can be located anywhere on your computer or
the network. The usual Bash wildcards like <tt class="docutils literal"><span class="pre">*</span></tt> and <tt class="docutils literal"><span class="pre">~</span></tt> can be used:</p>
<div class="highlight-python"><div class="highlight"><pre>./example.sh ~/test_dir/*.sam
</pre></div>
</div>
<p>would process all SAM files located in a directory called <cite>test_dir</cite> in the
user&#8217;s home directory. If you
have chosen the option to shorten filepaths to filenames only you need to keep
any additional files like taxon lists or other CSVs in the same directory as the
Bash script.</p>
<p>Running a Bash script that uses GNU <cite>parallel</cite> will show a summary of currently
running jobs, the number of available CPU cores, the average time per job as
well as an update on the estimated time until completion of all jobs (only
available once the first jobs have finished).</p>
</div>
<div class="section" id="manuals-for-filters-and-tools">
<span id="filter-manuals"></span><h2><a class="toc-backref" href="#id26">Manuals for Filters and Tools</a><a class="headerlink" href="#manuals-for-filters-and-tools" title="Permalink to this headline">¶</a></h2>
<div class="section" id="common-options">
<span id="id1"></span><h3><a class="toc-backref" href="#id27">Common Options</a><a class="headerlink" href="#common-options" title="Permalink to this headline">¶</a></h3>
<p>SamSifter includes a suite of built-in filters as well as wrappers for external
tools which are characterized by the name of the underlying tool in square
brackets (eg. [SAMtools] or [PMDtools]). All of the built-in filters and also
some of the wrappers share a set of general options:</p>
<ul class="simple">
<li><strong>filter direction</strong> has two possible settings: one <em>keeps</em> all entries in the
analyzed file that match the filter requirements, the other <em>discards</em> these
entries. By default all filters use the discard setting, but in case of eg.
a pathogen screening with a small list of interesting taxon IDs it may be
useful to invert this in order to <em>keep</em> all reads that map to a pathogen of
interest.</li>
<li><strong>verbose</strong> is activated by default and provides additional information on the
filter process. It can be safely deactivated, resulting in less noise in the
messages dock and shorter log files on the command-line interface.</li>
<li><strong>debug</strong> is deactivated by default and prints information to the command-line
which may help with troubleshooting and debugging the program. It is only
useful to software developers.</li>
</ul>
</div>
<div class="section" id="count-reads-per-taxon">
<span id="count-taxon-reads"></span><h3><a class="toc-backref" href="#id28">Count reads per taxon</a><a class="headerlink" href="#count-reads-per-taxon" title="Permalink to this headline">¶</a></h3>
<p>This analysis step will not alter the input file but count the number of reads
assigned to each taxon ID. The results are stored to a temporary statistics file
as comma-separated values (CSV) with taxon IDs in column <cite>taxon_id</cite> and the
read counts in column <cite>read_count</cite>.</p>
<p>Up to 1000 of these steps can be included in a workflow to compare read counts
between arbitrary filtering steps. The temporary files as well as the
corresponding read count columns will be enumerated from <cite>000</cite> to <cite>999</cite> and
compiled into one large spreadsheet during post-processing.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#statistics"><em>Statistical Summaries</em></a></p>
</div>
</div>
<div class="section" id="filter-reads-by-conservation">
<span id="filter-read-conservation"></span><h3><a class="toc-backref" href="#id29">Filter reads by conservation</a><a class="headerlink" href="#filter-reads-by-conservation" title="Permalink to this headline">¶</a></h3>
<p>This filter identifies reads mapping to the references of multiple taxa with similar
alignment scores. It is useful when dealing with reads mapping to highly
conserved or orthologous genes that are shared across multiple taxa. The filter
requires you to specify the permitted deviation from the maximal alignment score
(default is 5) and the maximum number of taxa that a read can be mapped to
within this range of alignment scores before it gets excluded (default is 10).</p>
</div>
<div class="section" id="filter-reads-by-list-of-qnames">
<span id="filter-read-list"></span><h3><a class="toc-backref" href="#id30">Filter reads by list of QNAMES</a><a class="headerlink" href="#filter-reads-by-list-of-qnames" title="Permalink to this headline">¶</a></h3>
<p>This filter is used to remove specific reads from a SAM file by using a list of
unique read identifiers (so-called QNAMES or query names). The list should be
formatted as tab-separated CSV file and list the query names in the first
column without headers. Any additional columns will be ignored.</p>
</div>
<div class="section" id="filter-references-by-evenness-of-coverage">
<span id="filter-ref-coverage"></span><h3><a class="toc-backref" href="#id31">Filter references by evenness of coverage</a><a class="headerlink" href="#filter-references-by-evenness-of-coverage" title="Permalink to this headline">¶</a></h3>
<p>This filter analyses the distribution of read bases to reference bases to
calculate the Gini coefficient for each reference, then filters all
references exceeding the given maximimal Gini threshold (default 0.90) to achieve
a higher prevalence of evenly covered references.</p>
<p>The filter can be further restricted to only consider references with average
coverage falling within a certain range. By default the range is set from 0x
minimal to 9999x maximal coverage, hence the thresholds have no effect. However,
because the Gini coefficient is a measure of <em>relative</em> inequality it may have
less significance for references with low minimum coverage, which can be
effectively excluded from filtering by setting a corresponding minimum coverage.</p>
<p>By default all reference bases are considered, but the calculation of coverage
and read length distributions can optionally be restricted to only the covered
bases (coverage &gt;= 1).</p>
<div class="section" id="what-is-the-gini-coefficient">
<h4><a class="toc-backref" href="#id32">What is the Gini coefficient?</a><a class="headerlink" href="#what-is-the-gini-coefficient" title="Permalink to this headline">¶</a></h4>
<p>The Gini coefficient is often used as a metric for the relative inequality of
income distribution across the population, however it can also be applied to
describe the heterogenity of the distribution of read bases to reference bases
(= the coverage distribution). This makes it a useful tool when dealing with
references that show distinct peaks in their coverage plot due to large clusters
of reads that map to few conserved areas while leaving the larger part of the
reference sequence uncovered.</p>
<p>Given the normed cumulative coverage distribution of all covered reference bases
(= a Lorenz curve) the Gini coefficient describes the ratio of the area A between
the Lorenz curve and the diagonal (describing a perfectly even distribution of
read bases) over the total area B underneath the diagonal. Thus, a completely even
distribution of read bases across the entire reference would have a minimal Gini
coefficient of 0.0 while a completely uneven distribution with all read bases
mapped to just one reference base would have a maximal Gini coefficient of 1.0.</p>
</div>
<div class="section" id="experimental-options">
<h4><a class="toc-backref" href="#id33">Experimental options</a><a class="headerlink" href="#experimental-options" title="Permalink to this headline">¶</a></h4>
<p>This filter can also be used to analyze the read length distribution per
reference (disabled by default) and to plot coverage and read length
distributions (also disabled by default).</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Activating the plotting of these distributions for a large input
dataset can create I/O problems due to the large amounts of PNG
files generated. It will also decrease the performance of this
filter considerably and should only be used to troubleshoot filter
parameters for small subsets of the data.</p>
</div>
</div>
</div>
<div class="section" id="filter-references-by-identity">
<span id="filter-ref-identity"></span><h3><a class="toc-backref" href="#id34">Filter references by % identity</a><a class="headerlink" href="#filter-references-by-identity" title="Permalink to this headline">¶</a></h3>
<p>This is a two-step filter that first determines the number of reads mapping to a
reference with at least 90% identity, then identifies those references with at
least 50% reads passing the first threshold. Both filter settings can be changed
individually and the thresholds can also be set as upper instead of lower
limits. By default all references fulfilling the filter criteria are discarded.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The % identity calculation is based on information in the CIGAR
string and the optional MD tag and will fail if the SAM file contains
no header.</p>
</div>
</div>
<div class="section" id="filter-references-by-list-of-accessions">
<span id="filter-ref-list"></span><h3><a class="toc-backref" href="#id35">Filter references by list of accessions</a><a class="headerlink" href="#filter-references-by-list-of-accessions" title="Permalink to this headline">¶</a></h3>
<p>This filter is used to remove specific reads from a SAM file by using a list of
unique accessions. The list should be formatted as tab-separated CSV file and
list the accessions in the first column without headers. Any additional columns
will be ignored. By default all reads mapping to an accession in the list will
be discarded.</p>
</div>
<div class="section" id="filter-references-by-pmd-score">
<span id="filter-ref-pmds"></span><h3><a class="toc-backref" href="#id36">Filter references by PMD score</a><a class="headerlink" href="#filter-references-by-pmd-score" title="Permalink to this headline">¶</a></h3>
<p>This is a two-step filter that calculates the post-mortem degradation score for
each read and in first step determines the number of ancient reads (PMDS at
least 3) mapping to a reference, then identifies those references with at least
50% reads passing the first threshold. Both filter settings can be changed
individually. By default all references fulfilling the filter criteria are
discarded.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The PMD score calculation is based on information in the CIGAR
string and the optional MD tag and will fail if the SAM file contains
no header.</p>
</div>
</div>
<div class="section" id="filter-taxa-by-list-of-taxon-ids">
<span id="filter-taxon-list"></span><h3><a class="toc-backref" href="#id37">Filter taxa by list of taxon IDs</a><a class="headerlink" href="#filter-taxa-by-list-of-taxon-ids" title="Permalink to this headline">¶</a></h3>
<p>This filter is used to remove specific taxa from a SAM file by using a list of
unique NCBI taxon IDs. The list should be formatted as tab-separated CSV file and
list the taxon IDs in the first column without headers. Any additional columns
will be ignored. By default all reads mapping to a taxon in the list will
be discarded.</p>
</div>
<div class="section" id="filter-taxa-by-pmd-score">
<span id="filter-taxon-pmds"></span><h3><a class="toc-backref" href="#id38">Filter taxa by PMD score</a><a class="headerlink" href="#filter-taxa-by-pmd-score" title="Permalink to this headline">¶</a></h3>
<p>This is a two-step filter that calculates the post-mortem degradation score for
each read and in first step determines the number of ancient reads (PMDS at
least 3) mapping to a taxon, then identifies those taxa with at least
50% reads passing the first threshold. Both filter settings can be changed
individually. By default all references fulfilling the filter criteria are
discarded.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The PMD score calculation is based on information in the CIGAR
string and the optional MD tag and will fail if the SAM file contains
no header.</p>
</div>
</div>
<div class="section" id="gnu-gzip-compress-file">
<span id="compress"></span><h3><a class="toc-backref" href="#id39">[GNU gzip] Compress file</a><a class="headerlink" href="#gnu-gzip-compress-file" title="Permalink to this headline">¶</a></h3>
<p>This wrapper for GNU <cite>gzip</cite> compresses any uncompressed file. It should be
used only once towards the end of a workflow as frequent compression and
decompression between steps can slow processing down.</p>
<p>This step will simply execute the command</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gzip</span> <span class="o">-</span><span class="n">c</span>
</pre></div>
</div>
<p>and has no additional options.</p>
</div>
<div class="section" id="gnu-gzip-decompress-file">
<span id="decompress"></span><h3><a class="toc-backref" href="#id40">[GNU gzip] Decompress file</a><a class="headerlink" href="#gnu-gzip-decompress-file" title="Permalink to this headline">¶</a></h3>
<p>A wrapper for GNU <cite>gzip</cite> that decompresses any zipped file. It should be used
only once towards the start of a workflow as frequent compression and
decompression between steps can slow processing down.</p>
<p>This step simply runs the command</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gzip</span> <span class="o">-</span><span class="n">d</span>
</pre></div>
</div>
<p>and has no additional options.</p>
</div>
<div class="section" id="pmdtools-calculate-pmds">
<span id="calculate-pmds"></span><h3><a class="toc-backref" href="#id41">[PMDtools] Calculate PMDS</a><a class="headerlink" href="#pmdtools-calculate-pmds" title="Permalink to this headline">¶</a></h3>
<p>This wrapper for a modified version of PMDtools does not have any filter
functionality but simply calculates the post-mortem degradation score for each
read and writes it into a custom DS tag by running the command</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">pmdtools_mod</span> <span class="o">--</span><span class="n">dry</span> <span class="o">--</span><span class="n">writesamfield</span> <span class="o">--</span><span class="n">header</span>
</pre></div>
</div>
<p>Following PMDtools steps in the workflow will utilize these tags and skip the
calculation to save time.</p>
<p>PMDtools can only calculate post-mortem degradation scores for SAM files which
contain a complete header with CIGAR strings and optional MD tags because it
has to reconstruct the reference sequence from these fields. The tool can
optionally write a brief statistical summary to the Messages dock when done,
however this is disabled by default.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference external" href="http://www.pnas.org/content/111/6/2229.abstract">Skoglund et al. (2014)</a>
for details on the calculation of post-mortem degradation scores
and the original version of PMDtools.</p>
</div>
</div>
<div class="section" id="pmdtools-filter-reads-by-identity">
<span id="filter-read-identity"></span><h3><a class="toc-backref" href="#id42">[PMDtools] Filter reads by % identity</a><a class="headerlink" href="#pmdtools-filter-reads-by-identity" title="Permalink to this headline">¶</a></h3>
<p>This filter is a wrapper for a modified version of PMDtools that targets reads
with insufficient identity to their respective reference sequence. Any reads with
identity values below the set minimum threshold (default is 0.95 = 95%) will be
discarded.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The calculation of identity according to PMDtools ignores indels,
possibly deaminated bases as well as unknown bases (Ns). In order to
achieve compatibility with MALT and other tools that include all of
these in their calculation the three corresponding options have to be
activated.</p>
</div>
<p>PMDtools can only calculate identity values for SAM files which
contain a complete header with CIGAR strings and optional MD tags because it
has to reconstruct the reference sequence from these fields. The tool can
optionally write a brief statistical summary to the Messages dock when done,
however this is disabled by default.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference external" href="http://www.pnas.org/content/111/6/2229.abstract">Skoglund et al. (2014)</a>
for the original version of PMDtools.</p>
</div>
</div>
<div class="section" id="pmdtools-filter-reads-by-pmds">
<span id="filter-read-pmds"></span><h3><a class="toc-backref" href="#id43">[PMDtools] Filter reads by PMDS</a><a class="headerlink" href="#pmdtools-filter-reads-by-pmds" title="Permalink to this headline">¶</a></h3>
<p>This filter uses a modified version of PMDtools to calculate the post-mortem
degradation score (PMDS) for each read and discards all reads with scores below
the set threshold.</p>
<p>PMDtools can only calculate post-mortem degradation scores
for SAM files which contain a complete header with CIGAR strings and optional MD
tags because it has to reconstruct the reference sequence from these fields.
The tool can optionally write a brief statistical summary to the Messages dock
when done, however this is disabled by default.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference external" href="http://www.pnas.org/content/111/6/2229.abstract">Skoglund et al. (2014)</a>
for details on the calculation of post-mortem degradation scores
and the original version of PMDtools.</p>
</div>
</div>
<div class="section" id="samtools-convert-bam-to-sam">
<span id="bam-2-sam"></span><h3><a class="toc-backref" href="#id44">[SAMtools] Convert BAM to SAM</a><a class="headerlink" href="#samtools-convert-bam-to-sam" title="Permalink to this headline">¶</a></h3>
<p>This is a simple wrapper for SAMtools <cite>view</cite> which converts the binary BAM
format to the text-based SAM format by running the command</p>
<div class="highlight-python"><div class="highlight"><pre>samtools view -
</pre></div>
</div>
<p>The option to also print the SAM header (<cite>-h</cite> on the command-line interface) is
enabled by default, as the information in those header lines is crucial for most
of SamSifter&#8217;s filters.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a></p>
</div>
</div>
<div class="section" id="samtools-convert-sam-to-bam">
<span id="sam-2-bam"></span><h3><a class="toc-backref" href="#id45">[SAMtools] Convert SAM to BAM</a><a class="headerlink" href="#samtools-convert-sam-to-bam" title="Permalink to this headline">¶</a></h3>
<p>This is a simple wrapper for SAMtools <cite>view</cite> which converts the text-based SAM
format to the binary BAM format by running the command</p>
<div class="highlight-python"><div class="highlight"><pre>samtools view -S -b -
</pre></div>
</div>
<p>The output is compressed by default but compression can be disabled to speed up
the following steps. This is only recommended when all following steps are part
of the SAMtools suite as well.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a></p>
</div>
</div>
<div class="section" id="samtools-remove-duplicates">
<span id="remove-duplicates"></span><h3><a class="toc-backref" href="#id46">[SAMtools] Remove duplicates</a><a class="headerlink" href="#samtools-remove-duplicates" title="Permalink to this headline">¶</a></h3>
<p>This wrapper for SAMtools <cite>rmdup</cite> removes duplicate reads with identical start
coordinates from BAM files by running the command</p>
<div class="highlight-python"><div class="highlight"><pre>samtools rmdup - -
</pre></div>
</div>
<p>By default single-end reads are included in
duplicate removal. Paired-end reads can optionally be treated as single-end
reads.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a></p>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#better-remove-duplicates"><em>[BetterRMDup] Remove duplicates</em></a> for the improved capabilities of
Alexander Peltzer&#8217;s <cite>BetterRMDup</cite> tool</p>
</div>
</div>
<div class="section" id="samtools-sort-by-coordinates">
<span id="sort-by-coordinates"></span><h3><a class="toc-backref" href="#id47">[SAMtools] Sort by coordinates</a><a class="headerlink" href="#samtools-sort-by-coordinates" title="Permalink to this headline">¶</a></h3>
<p>This wrapper for SAMtools <cite>sort</cite> sorts entries in a BAM file by their
coordinates by running the command</p>
<div class="highlight-python"><div class="highlight"><pre>samtools sort - tmp -o
</pre></div>
</div>
<p>and has no additional options. Sorting uses a temporary file with prefix <tt class="docutils literal"><span class="pre">tmp</span></tt>
that will briefly occupy additional storage space in the current work directory
if the whole alignment does not fit into memory.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a></p>
</div>
</div>
<div class="section" id="samtools-sort-by-query-names">
<span id="sort-by-names"></span><h3><a class="toc-backref" href="#id48">[SAMtools] Sort by query names</a><a class="headerlink" href="#samtools-sort-by-query-names" title="Permalink to this headline">¶</a></h3>
<p>This wrapper for SAMtools <cite>sort</cite> sorts entries in a BAM file by their
query names (unique read identifiers) by running the command</p>
<div class="highlight-python"><div class="highlight"><pre>samtools sort -n - tmp -o
</pre></div>
</div>
<p>and has no additional options. Sorting uses a temporary file with prefix <tt class="docutils literal"><span class="pre">tmp</span></tt>
that will briefly occupy additional storage space in the current work directory
if the whole alignment does not fit into memory.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference external" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a></p>
</div>
</div>
<div class="section" id="betterrmdup-remove-duplicates">
<span id="better-remove-duplicates"></span><h3><a class="toc-backref" href="#id49">[BetterRMDup] Remove duplicates</a><a class="headerlink" href="#betterrmdup-remove-duplicates" title="Permalink to this headline">¶</a></h3>
<p>This wrapper for Alexander Peltzer&#8217;s <cite>BetterRMDup</cite> removes duplicate reads from
BAM files and improves upon the SAMtools <cite>rmdup</cite> tool by considering both the
start <em>and</em> the stop coordinates. It runs the command</p>
<div class="highlight-python"><div class="highlight"><pre>betterrmdup -
</pre></div>
</div>
<p>which itself is a wrapper for</p>
<div class="highlight-python"><div class="highlight"><pre>java -jar BetterRMDup.jar -
</pre></div>
</div>
<p>The tool currently provides no further options.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference internal" href="#remove-duplicates"><em>[SAMtools] Remove duplicates</em></a> for the original SAMtools <cite>rmdup</cite></p>
</div>
</div>
</div>
<div class="section" id="statistical-summaries">
<span id="statistics"></span><h2><a class="toc-backref" href="#id50">Statistical Summaries</a><a class="headerlink" href="#statistical-summaries" title="Permalink to this headline">¶</a></h2>
<p>Summaries of read counts before or after any step in a workflow can easily be
created by including one or more analysis steps (eg. <a class="reference internal" href="#count-taxon-reads"><em>Count reads per taxon</em></a>)
in the workflow.</p>
<div class="section" id="compiling-results-per-analysed-file">
<h3><a class="toc-backref" href="#id51">Compiling Results Per Analysed File</a><a class="headerlink" href="#compiling-results-per-analysed-file" title="Permalink to this headline">¶</a></h3>
<p>Each of these steps will create a temporary CSV file that is
removed once the data have been automatically compiled into a spreadsheet at the
end of a workflow. The tool responsible for this compilation is called
<tt class="docutils literal"><span class="pre">compile_stats</span></tt>. By default it is included in the post-processing of every
SamSifter workflow. It produces one CSV file for each
analysed SAM file with tab-separated columns: the first column contains the
taxon IDs, the following columns the read counts - one for each analysis step
that was included in the workflow. If no temporary files are found it does
nothing.</p>
<p>The statistics can be used to validate filter settings or their order and
evaluate the impact on size and composition of the filtered SAM file. One
example: If you are in doubt which threshold for a filter setting is going to
give the expected result you can create a cascade of filter steps, each with
slightly stricter thresholds than the previous and each followed by an analysis
step. By running the workflow once and looking at the resulting spreadsheet you
can then pinpoint the one filter threshold at which only the required entries
are left in the file. Your final workflow will then contain only one filter step
with this exact threshold and may even leave out the analysis step to save time.</p>
</div>
<div class="section" id="comparison-of-multiple-files">
<h3><a class="toc-backref" href="#id52">Comparison of Multiple Files</a><a class="headerlink" href="#comparison-of-multiple-files" title="Permalink to this headline">¶</a></h3>
<p>When you repeat a workflow with different SAM files or use a Bash script to
process an entire batch of them (see <a class="reference internal" href="#bash-export"><em>Export to Bash scripts</em></a>), you can use the tool
<tt class="docutils literal"><span class="pre">summarize_stats</span></tt> to combine all those individual statistics spreadsheets into
one summary file. All Bash scripts exported from SamSifter workflows that
provide batch processing functionalities already execute this tool by default.
The resulting spreadsheet will again contain taxon IDs in the first
column and then a bunch of columns with the final read counts for each analysed
SAM file. It can be very useful to quickly screen larger collections for
interesting data and compare the results between individual files.</p>
</div>
<div class="section" id="enriching-summaries-with-external-databases">
<h3><a class="toc-backref" href="#id53">Enriching Summaries with External Databases</a><a class="headerlink" href="#enriching-summaries-with-external-databases" title="Permalink to this headline">¶</a></h3>
<p>Often a cryptic taxon ID is not sufficient to identify interesting entries in a
filtered SAM file. One example is additional information like species names in
English and Latin, host organisms or associated diseases that may be useful for
a filter workflow that screens metagenomic samples for human pathogens.</p>
<p>The tool <tt class="docutils literal"><span class="pre">enrich_summary</span></tt> combines SamSifter&#8217;s statistical summaries with
any CSV file or database dump containing this additional
information. It is included in SamSifter, but is <strong>not executed by default</strong> as
the required data and settings will differ from use case to use case.</p>
<p>One potential source for such a database dump is the
<a class="reference external" href="https://img.jgi.doe.gov/cgi-bin/m/main.cgi?section=FindGenomes&amp;page=genomeSearch">IMG/M database</a>,
but in
theory all text dumps from arbitrary databases are supported. <tt class="docutils literal"><span class="pre">enrich_summary</span></tt>
expects the database dump in a tab-delimited CSV format with
column headers. By default it searches for a column named <cite>NCBI_ID</cite> and
assumes that this column contains NCBI taxon IDs that match to those used in
the first column of the SamSifter summary file. Of course this index column can
be changed with a command-line option: <tt class="docutils literal"><span class="pre">--index</span> <span class="pre">my_taxon_id</span></tt> will look for a
column called <cite>my_taxon_id</cite> instead. By default the tool sorts the new, enriched
spreadsheet using a column called <cite>Genome_Name</cite>, but this sort field can also
be modified using the <tt class="docutils literal"><span class="pre">--sort</span> <span class="pre">my_species_name</span></tt> command-line option to sort by
<cite>my_species_name</cite> instead. The advantage of sorting by species name is the
clustering of taxonomically related entries in consecutive spreadsheet rows,
making interesting taxa, families or genera a lot easier to detect by visual
inspection.</p>
<p>A complete example related to the aforementioned screen of metagenomic samples
for human pathogens is the following command:</p>
<div class="highlight-python"><div class="highlight"><pre>enrich_summary --input samsifter.summary.csv --database imgm_database.txt --index &#39;NCBI Taxon ID&#39; --sort &#39;Species&#39; &gt; samsifter.summary.enriched.csv
</pre></div>
</div>
<p>Here the <tt class="docutils literal"><span class="pre">--input</span></tt> SamSifter summary file is enriched with data from the
<tt class="docutils literal"><span class="pre">--database</span></tt> IMG/M using the <tt class="docutils literal"><span class="pre">--index</span></tt> column called <cite>NCBI taxon ID</cite> and
sorted by <tt class="docutils literal"><span class="pre">--sort</span></tt> column <cite>Species</cite>. The results are written to a new CSV file.
If you have any doubts
about the spelling or meaning of any of these options you can always use the
built-in help function by calling this tool (and any of the other tools included
with SamSifter) with the option <tt class="docutils literal"><span class="pre">-h</span></tt> or <tt class="docutils literal"><span class="pre">--help</span></tt> to show the help text with
useful explanations.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference external" href="https://img.jgi.doe.gov/cgi-bin/m/main.cgi">https://img.jgi.doe.gov/cgi-bin/m/main.cgi</a> for the IMG/M database
on Integrated Microbial Genomes with Microbiome Samples, which
provides rich but uncurated annotations for microbial sequencing
data. The integrated Genome Search functionality allows
downloads of customized database exports.</p>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">SamSifter Manual</a><ul>
<li><a class="reference internal" href="#detailed-installation-instructions">Detailed Installation Instructions</a><ul>
<li><a class="reference internal" href="#installing-python3-and-pyqt4">Installing <tt class="docutils literal"><span class="pre">Python3</span></tt> and <tt class="docutils literal"><span class="pre">PyQt4</span></tt></a><ul>
<li><a class="reference internal" href="#debian-8-jessie-and-newer-or-ubuntu-14-04-and-newer">Debian 8 (jessie) and newer or Ubuntu 14.04 and newer</a></li>
<li><a class="reference internal" href="#older-debian-or-ubuntu-systems">Older Debian or Ubuntu systems</a></li>
<li><a class="reference internal" href="#mac-os-x-using-homebrew">Mac OS X using Homebrew</a></li>
<li><a class="reference internal" href="#windows-32-bit">Windows (32 bit)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#installing-samsifter">Installing SamSifter</a></li>
<li><a class="reference internal" href="#uninstalling-samsifter">Uninstalling SamSifter</a></li>
</ul>
</li>
<li><a class="reference internal" href="#creating-and-editing-workflows">Creating and Editing Workflows</a><ul>
<li><a class="reference internal" href="#selecting-an-input-file">Selecting an Input File</a></li>
<li><a class="reference internal" href="#adding-and-removing-a-tool-or-filter">Adding and Removing a Tool or Filter</a></li>
<li><a class="reference internal" href="#rearranging-tools-and-filters">Rearranging Tools and Filters</a></li>
<li><a class="reference internal" href="#displaying-and-editing-filter-settings">Displaying and Editing Filter Settings</a></li>
<li><a class="reference internal" href="#selecting-an-output-file">Selecting an Output File</a><ul>
<li><a class="reference internal" href="#optional-creating-an-rma-file-for-megan">Optional: Creating an RMA file for MEGAN</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#executing-workflows">Executing Workflows</a><ul>
<li><a class="reference internal" href="#running-a-workflow">Running a Workflow</a></li>
<li><a class="reference internal" href="#log-messages">Log Messages</a></li>
<li><a class="reference internal" href="#stopping-a-workflow">Stopping a Workflow</a></li>
</ul>
</li>
<li><a class="reference internal" href="#loading-and-saving-workflows">Loading and Saving Workflows</a></li>
<li><a class="reference internal" href="#export-to-bash-scripts">Export to Bash scripts</a></li>
<li><a class="reference internal" href="#manuals-for-filters-and-tools">Manuals for Filters and Tools</a><ul>
<li><a class="reference internal" href="#common-options">Common Options</a></li>
<li><a class="reference internal" href="#count-reads-per-taxon">Count reads per taxon</a></li>
<li><a class="reference internal" href="#filter-reads-by-conservation">Filter reads by conservation</a></li>
<li><a class="reference internal" href="#filter-reads-by-list-of-qnames">Filter reads by list of QNAMES</a></li>
<li><a class="reference internal" href="#filter-references-by-evenness-of-coverage">Filter references by evenness of coverage</a><ul>
<li><a class="reference internal" href="#what-is-the-gini-coefficient">What is the Gini coefficient?</a></li>
<li><a class="reference internal" href="#experimental-options">Experimental options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#filter-references-by-identity">Filter references by % identity</a></li>
<li><a class="reference internal" href="#filter-references-by-list-of-accessions">Filter references by list of accessions</a></li>
<li><a class="reference internal" href="#filter-references-by-pmd-score">Filter references by PMD score</a></li>
<li><a class="reference internal" href="#filter-taxa-by-list-of-taxon-ids">Filter taxa by list of taxon IDs</a></li>
<li><a class="reference internal" href="#filter-taxa-by-pmd-score">Filter taxa by PMD score</a></li>
<li><a class="reference internal" href="#gnu-gzip-compress-file">[GNU gzip] Compress file</a></li>
<li><a class="reference internal" href="#gnu-gzip-decompress-file">[GNU gzip] Decompress file</a></li>
<li><a class="reference internal" href="#pmdtools-calculate-pmds">[PMDtools] Calculate PMDS</a></li>
<li><a class="reference internal" href="#pmdtools-filter-reads-by-identity">[PMDtools] Filter reads by % identity</a></li>
<li><a class="reference internal" href="#pmdtools-filter-reads-by-pmds">[PMDtools] Filter reads by PMDS</a></li>
<li><a class="reference internal" href="#samtools-convert-bam-to-sam">[SAMtools] Convert BAM to SAM</a></li>
<li><a class="reference internal" href="#samtools-convert-sam-to-bam">[SAMtools] Convert SAM to BAM</a></li>
<li><a class="reference internal" href="#samtools-remove-duplicates">[SAMtools] Remove duplicates</a></li>
<li><a class="reference internal" href="#samtools-sort-by-coordinates">[SAMtools] Sort by coordinates</a></li>
<li><a class="reference internal" href="#samtools-sort-by-query-names">[SAMtools] Sort by query names</a></li>
<li><a class="reference internal" href="#betterrmdup-remove-duplicates">[BetterRMDup] Remove duplicates</a></li>
</ul>
</li>
<li><a class="reference internal" href="#statistical-summaries">Statistical Summaries</a><ul>
<li><a class="reference internal" href="#compiling-results-per-analysed-file">Compiling Results Per Analysed File</a></li>
<li><a class="reference internal" href="#comparison-of-multiple-files">Comparison of Multiple Files</a></li>
<li><a class="reference internal" href="#enriching-summaries-with-external-databases">Enriching Summaries with External Databases</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="help.html"
                        title="previous chapter">SamSifter Help</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="api.html"
                        title="next chapter">SamSifter API documentation</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/manual.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="api.html" title="SamSifter API documentation"
             >next</a> |</li>
        <li class="right" >
          <a href="help.html" title="SamSifter Help"
             >previous</a> |</li>
        <li><a href="index.html">SamSifter 0.5.0 documentation</a> &raquo;</li>
          <li><a href="help.html" >SamSifter Help</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2015, Florian Aldehoff.
      Last updated on Mar 01, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
  </body>
</html>