<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Using the GEMINI API</title>
    
    <link rel="stylesheet" href="../_static/labibi.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.12.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/gemini.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="Acknowledgements" href="acknowledgements.html" />
    <link rel="prev" title="The Gemini database schema" href="database_schema.html" />


<script src="../_static/labibi.js"></script>



<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', '']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>


  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="acknowledgements.html" title="Acknowledgements"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="database_schema.html" title="The Gemini database schema"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  
  <div class="section" id="module-gemini">
<span id="using-the-gemini-api"></span><h1>Using the GEMINI API<a class="headerlink" href="#module-gemini" title="Permalink to this headline">¶</a></h1>
<div class="section" id="the-geminiquery-class">
<h2>The GeminiQuery class<a class="headerlink" href="#the-geminiquery-class" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="gemini.GeminiQuery">
<em class="property">class </em><tt class="descclassname">gemini.</tt><tt class="descname">GeminiQuery</tt><big>(</big><em>db</em>, <em>include_gt_cols=False</em>, <em>out_format=&lt;gemini.GeminiQuery.DefaultRowFormat object at 0x10c573d10&gt;</em><big>)</big><a class="headerlink" href="#gemini.GeminiQuery" title="Permalink to this definition">¶</a></dt>
<dd><p>An interface to submit queries to an existing Gemini database
and iterate over the results of the query.</p>
<p>We create a GeminiQuery object by specifying database to which to
connect:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">gemini</span> <span class="kn">import</span> <span class="n">GeminiQuery</span>
<span class="n">gq</span> <span class="o">=</span> <span class="n">GeminiQuery</span><span class="p">(</span><span class="s">&quot;my.db&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can then issue a query against the database and iterate through
the results by using the <tt class="docutils literal"><span class="pre">run()</span></tt> method:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gq</span><span class="p">:</span>
    <span class="k">print</span> <span class="n">row</span>
</pre></div>
</div>
<p>Instead of printing the entire row, one access print specific columns:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gq</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="s">&quot;select chrom, start, end from variants&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gq</span><span class="p">:</span>
    <span class="k">print</span> <span class="n">row</span><span class="p">[</span><span class="s">&#39;chrom&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Also, all of the underlying numpy genotype arrays are
always available:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gq</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="s">&quot;select chrom, start, end from variants&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gq</span><span class="p">:</span>
    <span class="n">gts</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">gts</span>
    <span class="k">print</span> <span class="n">row</span><span class="p">[</span><span class="s">&#39;chrom&#39;</span><span class="p">],</span> <span class="n">gts</span>
    <span class="c"># yields &quot;chr1&quot; [&#39;A/G&#39; &#39;G/G&#39; ... &#39;A/G&#39;]</span>
</pre></div>
</div>
<p>The <tt class="docutils literal"><span class="pre">run()</span></tt> methods also accepts genotype filter:</p>
<div class="highlight-python"><div class="highlight"><pre>query = &quot;select chrom, start, end&quot; from variants&quot;
gt_filter = &quot;gt_types.NA20814 == HET&quot;
gq.run(query)
for row in gq:
    print row
</pre></div>
</div>
<p>Lastly, one can use the <tt class="docutils literal"><span class="pre">sample_to_idx</span></tt> and <tt class="docutils literal"><span class="pre">idx_to_sample</span></tt>
dictionaries to gain access to sample-level genotype information
either by sample name or by sample index:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># grab dict mapping sample to genotype array indices</span>
<span class="n">smp2idx</span> <span class="o">=</span> <span class="n">gq</span><span class="o">.</span><span class="n">sample_to_idx</span>

<span class="n">query</span>  <span class="o">=</span> <span class="s">&quot;select chrom, start, end from variants&quot;</span>
<span class="n">gt_filter</span>  <span class="o">=</span> <span class="s">&quot;gt_types.NA20814 == HET&quot;</span>
<span class="n">gq</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">query</span><span class="p">,</span> <span class="n">gt_filter</span><span class="p">)</span>

<span class="c"># print a header listing the selected columns</span>
<span class="k">print</span> <span class="n">gq</span><span class="o">.</span><span class="n">header</span>
<span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gq</span><span class="p">:</span>
    <span class="c"># access a NUMPY array of the sample genotypes.</span>
    <span class="n">gts</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s">&#39;gts&#39;</span><span class="p">]</span>
    <span class="c"># use the smp2idx dict to access sample genotypes</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">smp2idx</span><span class="p">[</span><span class="s">&#39;NA20814&#39;</span><span class="p">]</span>
    <span class="k">print</span> <span class="n">row</span><span class="p">,</span> <span class="n">gts</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="gemini.GeminiQuery.header">
<tt class="descname">header</tt><a class="headerlink" href="#gemini.GeminiQuery.header" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a header describing the columns that
were selected in the query issued to a GeminiQuery object.</p>
</dd></dl>

<dl class="attribute">
<dt id="gemini.GeminiQuery.index2sample">
<tt class="descname">index2sample</tt><a class="headerlink" href="#gemini.GeminiQuery.index2sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary mapping sample names to
genotype array offsets:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gq</span> <span class="o">=</span> <span class="n">GeminiQuery</span><span class="p">(</span><span class="s">&quot;my.db&quot;</span><span class="p">)</span>
<span class="n">i2s</span> <span class="o">=</span> <span class="n">gq</span><span class="o">.</span><span class="n">index2sample</span>

<span class="k">print</span> <span class="n">i2s</span><span class="p">[</span><span class="mi">1088</span><span class="p">]</span>
<span class="c"># yields &quot;NA20814&quot;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="gemini.GeminiQuery.run">
<tt class="descname">run</tt><big>(</big><em>query</em>, <em>gt_filter=None</em>, <em>show_variant_samples=False</em>, <em>variant_samples_delim='</em>, <em>'</em>, <em>predicates=None</em>, <em>needs_genotypes=False</em>, <em>needs_genes=False</em>, <em>show_families=False</em>, <em>subjects=None</em><big>)</big><a class="headerlink" href="#gemini.GeminiQuery.run" title="Permalink to this definition">¶</a></dt>
<dd><p>Execute a query against a Gemini database. The user may
specify:</p>
<blockquote>
<div><ol class="arabic simple">
<li>(reqd.) an SQL <cite>query</cite>.</li>
<li>(opt.) a genotype filter.</li>
</ol>
</div></blockquote>
</dd></dl>

<dl class="attribute">
<dt id="gemini.GeminiQuery.sample2index">
<tt class="descname">sample2index</tt><a class="headerlink" href="#gemini.GeminiQuery.sample2index" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary mapping sample names to
genotype array offsets:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">gq</span> <span class="o">=</span> <span class="n">GeminiQuery</span><span class="p">(</span><span class="s">&quot;my.db&quot;</span><span class="p">)</span>
<span class="n">s2i</span> <span class="o">=</span> <span class="n">gq</span><span class="o">.</span><span class="n">sample2index</span>

<span class="k">print</span> <span class="n">s2i</span><span class="p">[</span><span class="s">&#39;NA20814&#39;</span><span class="p">]</span>
<span class="c"># yields 1088</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<div class="section" id="extracting-the-vcf-info-tags-with-gemini-api">
<h3>Extracting the VCF INFO tags with GEMINI API<a class="headerlink" href="#extracting-the-vcf-info-tags-with-gemini-api" title="Permalink to this headline">¶</a></h3>
<p>The GEMINI API is useful to extract the individual tags within the INFO field of a VCF (stored as a
compressed dictionary in the variants table). This would be of particular interest to those who want
to add custom annotations to their VCF and still be able to access the individual tags programmatically.
Here is an example where we try to extract the dbNSFP fields from the &#8216;INFO&#8217; tag of a VCF, using the API.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c">#!/usr/bin/env python</span>
import sys
from gemini import GeminiQuery

<span class="nv">database</span> <span class="o">=</span> sys.argv<span class="o">[</span>1<span class="o">]</span>
<span class="nv">gq</span> <span class="o">=</span> GeminiQuery<span class="o">(</span>database<span class="o">)</span>
<span class="nv">query</span> <span class="o">=</span> <span class="s2">&quot;SELECT variant_id, chrom, start, end, ref, alt, info \</span>
<span class="s2">         FROM variants&quot;</span>

gq.run<span class="o">(</span>query<span class="o">)</span>

<span class="k">for </span>row in gq:
    try:
        print <span class="s2">&quot;\t&quot;</span>.join<span class="o">([</span>str<span class="o">(</span>row<span class="o">[</span><span class="s1">&#39;chrom&#39;</span><span class="o">])</span>, str<span class="o">(</span>row<span class="o">[</span><span class="s1">&#39;start&#39;</span><span class="o">])</span>, str<span class="o">(</span>row<span class="o">[</span><span class="s1">&#39;end&#39;</span><span class="o">])</span>,
                      str<span class="o">(</span>row<span class="o">[</span><span class="s1">&#39;ref&#39;</span><span class="o">])</span>, str<span class="o">(</span>row<span class="o">[</span><span class="s1">&#39;alt&#39;</span><span class="o">])</span>, str<span class="o">(</span>row.info<span class="o">[</span><span class="s1">&#39;dbNSFP_SIFT_pred&#39;</span><span class="o">])])</span>
    except KeyError:
        pass

<span class="c"># yields</span>
chr1    906272  906273  C       T       P|D|P
chr1    906273  906274  C       A       D|D|D
chr1    906276  906277  T       C       D|D|D
chr1    906297  906298  G       T       B|B|B
chr1    1959074 1959075 A       C       D
chr1    1959698 1959699 G       A       B
chr1    1961452 1961453 C       T       P
chr1    2337953 2337954 C       T       D
</pre></div>
</div>
</div>
</div>
</div>




  

<div id="disqus_thread"></div>
 <script type="text/javascript">
     /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
     var disqus_shortname = 'geminidocs'; // required: replace example with your forum shortname

     /* * * DON'T EDIT BELOW THIS LINE * * */
     (function() {
         var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
         dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
         (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
     })();
 </script>
 <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
 <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>






          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/gemini.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Using the GEMINI API</a><ul>
<li><a class="reference internal" href="#the-geminiquery-class">The GeminiQuery class</a><ul>
<li><a class="reference internal" href="#extracting-the-vcf-info-tags-with-gemini-api">Extracting the VCF INFO tags with GEMINI API</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="database_schema.html"
                        title="previous chapter">The Gemini database schema</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="acknowledgements.html"
                        title="next chapter">Acknowledgements</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/api.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="acknowledgements.html" title="Acknowledgements"
             >next</a> |</li>
        <li class="right" >
          <a href="database_schema.html" title="The Gemini database schema"
             >previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2012,2013.
      Last updated on Mar 19, 2015.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>

<!-- 
<div id="editor-trap">
      <h3>Edit and improve this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/arq5x/gemini/blob/master/docs/content/api.rst">
          Using the GEMINI API
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/arq5x/gemini/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>
 -->
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', UA-39150563-1]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>


  </body>
</html>