<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Release History</title>
    
    <link rel="stylesheet" href="../_static/labibi.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.12.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/gemini.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="F.A.Q." href="faq.html" />
    <link rel="prev" title="&lt;no title&gt;" href="caveats.html" />


<script src="../_static/labibi.js"></script>



<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', '']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>


  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="faq.html" title="F.A.Q."
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="caveats.html" title="&lt;no title&gt;"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  
  <div class="section" id="release-history">
<h1>Release History<a class="headerlink" href="#release-history" title="Permalink to this headline">¶</a></h1>
<div class="section" id="id1">
<h2>0.12.1<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Support for input VCF files containing variants with multiple alternate alleles. Thanks to Brent Pedersen.</li>
<li>Updated, decomposed, and normalized the ExAC, Clinvar, Cosmic, dbSNP, and ESP annotation files to properly support variants with multiple alternate alleles.</li>
<li>Updated the logic for the clinvar significance column to retain all documented singificances.</li>
<li>Support for VCF annotation files in the <cite>annotate</cite> tool.</li>
<li>Improved the speed of loading by 10-15%. Thanks to Brent Pedersen.</li>
<li>Added <cite>&#8211;only-affected</cite> and <cite>&#8211;min-kindreds</cite> options to the compound heterozygotes tool.</li>
<li>Added a <cite>&#8211;format vcf</cite> option to the <cite>query</cite> tool to output query results in VCF format.</li>
<li>Added the <cite>&#8211;families</cite> option to the <cite>auto_*</cite>, <cite>de_novo</cite>, and <cite>comp_hets</cite> tools. Thanks to Mark Cowley and Tony Roscioli.</li>
<li>Added the <cite>&#8211;only-affected</cite> option to the <cite>de_novo</cite> tool.</li>
<li>Allow the <cite>&#8211;sample-filter</cite> to work with <cite>&#8211;format TPED</cite>. Thanks to Rory Kirchner.</li>
<li>Add <cite>&#8211;format sampledetail</cite> option that provides a melted/tidy/flattened version of samples along with <cite>&#8211;showsample</cite> and includes information from samples table. Thanks to Brad Chapman.</li>
<li>Add &#8216;not&#8217; option to &#8211;in filtering. Thanks to Rory Kirchner.</li>
<li>Fixed a bug in the <cite>de_novo</cite> tool that prevented proper function when families have affected and unaffected children. Thanks to Andrew Oler.</li>
<li>Fixed a bug in cyvcf that falsely treated &#8216;.|.&#8217; genotypes as homozygous alternate.  Thanks to Xiao Xu.</li>
<li>GEMINI now checks for and warns of old grabix index files. Thanks to Andrew Oler and Brent Pedersen.</li>
<li>Fixed a bug that added newlines at the end of tab delimited PED files.  Thanks to Brad Chapman.</li>
</ol>
</div>
<div class="section" id="id2">
<h2>0.11.0<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Integration of ExAC annotations (v0.2): <a class="reference external" href="http://exac.broadinstitute.org/">http://exac.broadinstitute.org/</a></li>
<li>New tools for cancer genome analysis. Many thanks to fantastic work from Colby Chiang.</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><cite>gemini set_somatic</cite></li>
<li><cite>gemini actionable_mutations</cite></li>
<li><cite>gemini fusions</cite></li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="3">
<li>Improved support for structural variants. New columns include:</li>
</ol>
<blockquote>
<div><ul class="simple">
<li><cite>sv_cipos_start_left</cite></li>
<li><cite>sv_cipos_end_left</cite></li>
<li><cite>sv_cipos_start_right</cite></li>
<li><cite>sv_cipos_end_right</cite></li>
<li><cite>sv_length</cite></li>
<li><cite>sv_is_precise</cite></li>
<li><cite>sv_tool</cite></li>
<li><cite>sv_evidence_type</cite></li>
<li><cite>sv_event_id</cite></li>
<li><cite>sv_mate_id</cite></li>
<li><cite>sv_strand</cite></li>
</ul>
</div></blockquote>
<ol class="arabic simple" start="4">
<li>Updated the 1000 Genomes annotations to the Phase variant set.</li>
<li>Added <cite>clinvar_causal_allele</cite> column.</li>
<li>Fixed a bug in grabix that caused occasional duplicate and missed variants.</li>
</ol>
</div>
<div class="section" id="id3">
<h2>0.10.1<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Add <cite>fitCons &lt;http://biorxiv.org/content/early/2014/09/11/006825&gt;</cite> scores as
an additional measure of potential function in variants of interest, supplementing
existing CADD and dbNSFP approaches.</li>
<li>Updated Clinvar, COSMIC, and dbSNP to their latest versions.</li>
</ol>
</div>
<div class="section" id="id4">
<h2>0.10.0<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Provide an <tt class="docutils literal"><span class="pre">--annotation-dir</span></tt> argument that specifies the path the
annotation databases, to overwrite configured data inputs. Thanks to Björn Grüning,</li>
<li>Support reproducible versioned installs of GEMINI with Python
dependencies. Enables Galaxy integration. Thanks to Björn Grüning,</li>
</ol>
</div>
<div class="section" id="id5">
<h2>0.8.0<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Support arbitrary annotation supplied to VEP, which translate into queryable
columns in the main variant table.</li>
<li>Improve the power of the genotype filter wildcard functionality.</li>
</ol>
</div>
<div class="section" id="id6">
<h2>0.7.1<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Suppress openpyxl/pandas warnings (thanks to &#64;chapmanb)</li>
<li>Fix unit tests to account for cases where a user has not downloaded the CADD or GERP annotation files.
Thanks to Xialoin Zhu and Daniel Swensson for reporting this and to Uma Paila for correcting it.</li>
</ol>
</div>
<div class="section" id="id7">
<h2>0.7.0<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Added support for CADD scores via new <tt class="docutils literal"><span class="pre">cadd_raw</span></tt> and <tt class="docutils literal"><span class="pre">cadd_scaled</span></tt> columns.</li>
<li>Added support for genotype wildcards in query select statements. E.g., <tt class="docutils literal"><span class="pre">SELECT</span> <span class="pre">chrom,</span> <span class="pre">start,</span> <span class="pre">end</span> <span class="pre">(gts).(phenotype==2)</span> <span class="pre">FROM</span> <span class="pre">variants</span></tt>. See <a class="reference external" href="http://gemini.readthedocs.org/en/latest/content/querying.html#selecting-sample-genotypes-based-on-wildcards">http://gemini.readthedocs.org/en/latest/content/querying.html#selecting-sample-genotypes-based-on-wildcards</a>.</li>
<li>Added support for genotype wildcards in the &#8211;gt-filter. E.g., <tt class="docutils literal"><span class="pre">--gt-filter</span> <span class="pre">&quot;(gt_types).(phenotype==2).(==HET)</span></tt>. See <a class="reference external" href="http://gemini.readthedocs.org/en/latest/content/querying.html#gt-filter-wildcard-filtering-on-genotype-columns">http://gemini.readthedocs.org/en/latest/content/querying.html#gt-filter-wildcard-filtering-on-genotype-columns</a>.</li>
<li>Added support for the VCF INFO field both in the API and as a column that can be SELECT&#8217;ed.</li>
<li>Upgraded to the latest version of ClinVar.</li>
<li>Standardized impacts to use Sequence Ontology (SO) terms.</li>
<li>Automatically add indexes to custom, user-supplied annotation columns.</li>
<li>Improvements to the installation script.</li>
<li>Fixed bugs in the handling of ClinVar UTF8 encoded strings.</li>
<li>Upgraded the <tt class="docutils literal"><span class="pre">gene_summary</span></tt> and <tt class="docutils literal"><span class="pre">gene_detailed</span></tt> tables to version 75 of Ensembl.</li>
<li>Added support for the MPI Mouse Phenotype database via the <tt class="docutils literal"><span class="pre">mam_phenotype_id</span></tt> column in the <tt class="docutils literal"><span class="pre">gene_summary</span></tt> table.</li>
<li>Enhanced security.</li>
<li>Corrected the ESP allele frequencies to be based report _alternate_ allele frequency instead of _minor_ allele frequency.</li>
<li>VEP version support updated (73-75) Support for aa length and bio type in VEP.</li>
<li>The <cite>lof_sieve</cite> tool support has been extended to VEP annotations.</li>
<li>Added the <tt class="docutils literal"><span class="pre">ccds_id</span></tt> and <tt class="docutils literal"><span class="pre">entrez_id</span></tt> columns to the <tt class="docutils literal"><span class="pre">gene_detailed</span></tt> table.</li>
</ol>
</div>
<div class="section" id="id8">
<h2>0.6.6<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Added COSMIC mutation information via new cosmic_ids column.</li>
</ol>
</div>
<div class="section" id="jan-03">
<h2>0.6.4 (2014-Jan-03)<a class="headerlink" href="#jan-03" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>New annotation: experimentally validated human enhancers from VISTA.</li>
<li>Installation improvements to enable isolated installations inside of virtual
machines and containers without data. Allow data-only upgrades as part of
<tt class="docutils literal"><span class="pre">update</span></tt> process.</li>
<li>Fix for gemini query error when <tt class="docutils literal"><span class="pre">--header</span></tt> specified (#241).</li>
</ol>
</div>
<div class="section" id="dec-10">
<h2>0.6.3.2 (2013-Dec-10)<a class="headerlink" href="#dec-10" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Fixed a bug that caused <tt class="docutils literal"><span class="pre">--gt-filter</span></tt> to no be enforced from <tt class="docutils literal"><span class="pre">query</span></tt> tool unless a GT* column was selected.</li>
<li>Support for ref and alt allele depths provided by FreeBayes.</li>
</ol>
</div>
<div class="section" id="nov-19">
<h2>0.6.3.1 (2013-Nov-19)<a class="headerlink" href="#nov-19" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Fixed undetected bug preventing the <tt class="docutils literal"><span class="pre">comp_hets</span></tt> tool from functioning.</li>
<li>Added unit tests for the <tt class="docutils literal"><span class="pre">comp_hets</span></tt> tool.</li>
</ol>
</div>
<div class="section" id="nov-7">
<h2>0.6.3 (2013-Nov-7)<a class="headerlink" href="#nov-7" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Addition permutation testing to the c-alpha test via the <tt class="docutils literal"><span class="pre">--permutations</span></tt>
option.</li>
<li>Addition of the <tt class="docutils literal"><span class="pre">--passonly</span></tt> option during loading to filter out all
variants with a filter flag set.</li>
<li>Fixed bug with parallel loading using the extended sample table format.</li>
<li>SLURM support added.</li>
<li>Refactor of loading options to remove explosion of xxx-queue options. Now
load using <tt class="docutils literal"><span class="pre">--scheduler</span></tt> on <tt class="docutils literal"><span class="pre">--queue</span></tt>.</li>
<li>Refactor of Sample class to handle the expanded samples table.</li>
<li>Addition of <tt class="docutils literal"><span class="pre">--carrier-summary-by-phenotype</span></tt> for summarizing the counts of
carriers and non-carriers stratified by the given sample phenotype column.</li>
<li>Added a <tt class="docutils literal"><span class="pre">--nonsynonymous</span></tt> option to the C-alpha test.</li>
<li>Added <tt class="docutils literal"><span class="pre">gemini</span> <span class="pre">amend</span></tt> to edit an existing database. For now only handles updating
the samples table.</li>
<li>Fixed a bug that prevented variants that overlapped with multiple 1000G variants
from having AAF info extracted from 1000G annotations.  This is now corrected such
that multiple overlaps with 1000G variants are tolerated, yet the logic ensures
that the AAF info is extracted for the correct variant.</li>
<li>Fixed installation issues for the GEMINI browser.</li>
<li>Added <tt class="docutils literal"><span class="pre">--show-families</span></tt> option to gemini query.</li>
</ol>
</div>
<div class="section" id="oct-7">
<h2>0.6.2 (2013-Oct-7)<a class="headerlink" href="#oct-7" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Moved <cite>&#8211;tped</cite> and <cite>&#8211;json</cite> options into the more generic <cite>&#8211;format</cite> option.</li>
<li>Fixed bug in handling missing phenotypes in the sample table.</li>
<li>Fixed <cite>&#8211;tped</cite> output formatting error.</li>
<li>API change: GeminiQuery.run takes an optional list of predicates that a row
must pass to be returned.</li>
<li><cite>&#8211;sample-filter</cite> option added to allow for restricting variants to samples
that pass the given sample query.</li>
<li>ethnicity removed as a default PED field.</li>
<li>PED file format extended to allow for extra columns to be added to the samples table under the column named in the header.</li>
<li>The autosomal_recessive and autosomal_dominant tools now warn, but allow for variants to be detected in the absence of known parent/child relationships.</li>
</ol>
</div>
<div class="section" id="sep-09">
<h2>0.6.1 (2013-Sep-09)<a class="headerlink" href="#sep-09" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Corrected bug in de_novo tool that was undetected in 0.6.0.  Unit tests have been added to head this off in the future. Thanks to <strong>Jessica Chong</strong></li>
<li>Added the <cite>-d</cite> option (minimum sequence depth allowed for a genotype) to the <cite>autosmal_recessive</cite> and <cite>autosmal_dominant</cite> tools.</li>
<li>New <cite>&#8211;tped</cite> option in the <cite>query</cite> tool for reporting variants in TPED format. Thanks to <strong>Rory Kirchner</strong>.</li>
<li>New <cite>&#8211;tfam</cite> option in the <cite>dump</cite> tool for reporting sample infor in TFAM format. Thanks to <strong>Rory Kirchner</strong>.</li>
</ol>
</div>
<div class="section" id="sep-02">
<h2>0.6.0 (2013-Sep-02)<a class="headerlink" href="#sep-02" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Add the <tt class="docutils literal"><span class="pre">--min-kindreds</span></tt> option to the <tt class="docutils literal"><span class="pre">autosomal_recessive</span></tt> and <tt class="docutils literal"><span class="pre">autosomal_dominant</span></tt> tools to restrict candidate variants/genes to those affecting at least <tt class="docutils literal"><span class="pre">--min-kindreds</span></tt>. Thanks to <strong>Jessica Chong</strong></li>
<li>Addition of a new <tt class="docutils literal"><span class="pre">burden</span></tt> tool for gene or region based burden tests.  First release supports the C-alpha test.  Thanks to <strong>Rory Kirchner</strong>.</li>
<li>Use of Continuum Analytics Anaconda python package for the automated installer. Thanks to <strong>Brad Chapman</strong>.</li>
<li>Enhancements to the <tt class="docutils literal"><span class="pre">annotate</span></tt> tool allowing one to create new database columns from values in custom BED+ annotation files.  Thanks to <strong>Jessica Chong</strong> and <strong>Graham Ritchie</strong>.</li>
<li>Addition of the <tt class="docutils literal"><span class="pre">--column</span></tt>, <tt class="docutils literal"><span class="pre">--filter</span></tt>, and <tt class="docutils literal"><span class="pre">--json</span></tt> options to the <tt class="docutils literal"><span class="pre">region</span></tt> tool.</li>
<li>Improvements to unit tests.</li>
<li>Allow alternate sample delimiters in the <tt class="docutils literal"><span class="pre">query</span></tt> tool via the <tt class="docutils literal"><span class="pre">--sample-delim</span></tt> option.  Thanks to <strong>Jessica Chong</strong>.</li>
<li>Provide a REST-like interface to the gemini browser.  In support of future visualization tools.</li>
<li>Allow the <tt class="docutils literal"><span class="pre">query</span></tt> tool to report results in JSON format via the <tt class="docutils literal"><span class="pre">--json</span></tt> option.</li>
<li>Various minor improvements and bug fixes.</li>
</ol>
</div>
<div class="section" id="b-2013-jul-23">
<h2>0.5.0b (2013-Jul-23)<a class="headerlink" href="#b-2013-jul-23" title="Permalink to this headline">¶</a></h2>
<ol class="arabic">
<li><p class="first">Tolerate either -9 or 0 for unknown parent or affected status in PED files.</p>
</li>
<li><p class="first">Refine the rules for inheritance and parental affected status for autosomal dominant inheritance models.</p>
</li>
<li><p class="first">The <tt class="docutils literal"><span class="pre">autosomal_dominant</span></tt>, <tt class="docutils literal"><span class="pre">autosomal_recessive</span></tt>, and <tt class="docutils literal"><span class="pre">de_novo</span></tt> mutation tools have received the following improvements.</p>
<blockquote>
<div><ul class="simple">
<li>improved speed (especially when there are multiple families)</li>
<li>by default, all columns in the variant table are reported and no conditions are placed on the returned variants.  That is, as long as the variant meets
the inheritance model, it will be reported.</li>
<li>the addition of a <tt class="docutils literal"><span class="pre">--columns</span></tt> option allowing one to override the above default behavior and report a subset of columns.</li>
<li>the addition of a <tt class="docutils literal"><span class="pre">--filter</span></tt> option allowing one to override the above default behavior and filter reported variants based on specific criteria.</li>
</ul>
</div></blockquote>
</li>
</ol>
<p>4. The default minimum aligned sequencing depth for each variant reported by
the <tt class="docutils literal"><span class="pre">de_novo</span></tt> tool is 0.  Greater stringency can be applied with the <tt class="docutils literal"><span class="pre">-d</span></tt>
option.</p>
</div>
<div class="section" id="b-2013-jun-12">
<h2>0.4.0b (2013-Jun-12)<a class="headerlink" href="#b-2013-jun-12" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Added new <tt class="docutils literal"><span class="pre">gt_ref_depths</span></tt>, <tt class="docutils literal"><span class="pre">gt_alt_depths</span></tt>, and <tt class="docutils literal"><span class="pre">gt_quals</span></tt> columns.</li>
<li>Added a new <tt class="docutils literal"><span class="pre">--show-samples</span></tt> option to the <tt class="docutils literal"><span class="pre">query</span></tt> module to display samples with alternate allele genotypes.</li>
<li>Improvements and bug fixes for installation.</li>
</ol>
</div>
<div class="section" id="b">
<h2>0.3.0b<a class="headerlink" href="#b" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li>Improved speed for adding custom annotations.</li>
<li>Added GERP conserved elements.</li>
<li>Optionally addition of GERP conservation scores at base pair resolution.</li>
<li>Move annotation files to Amazon S3.</li>
</ol>
</div>
</div>




  

<div id="disqus_thread"></div>
 <script type="text/javascript">
     /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
     var disqus_shortname = 'geminidocs'; // required: replace example with your forum shortname

     /* * * DON'T EDIT BELOW THIS LINE * * */
     (function() {
         var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
         dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
         (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
     })();
 </script>
 <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
 <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>






          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/gemini.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Release History</a><ul>
<li><a class="reference internal" href="#id1">0.12.1</a></li>
<li><a class="reference internal" href="#id2">0.11.0</a></li>
<li><a class="reference internal" href="#id3">0.10.1</a></li>
<li><a class="reference internal" href="#id4">0.10.0</a></li>
<li><a class="reference internal" href="#id5">0.8.0</a></li>
<li><a class="reference internal" href="#id6">0.7.1</a></li>
<li><a class="reference internal" href="#id7">0.7.0</a></li>
<li><a class="reference internal" href="#id8">0.6.6</a></li>
<li><a class="reference internal" href="#jan-03">0.6.4 (2014-Jan-03)</a></li>
<li><a class="reference internal" href="#dec-10">0.6.3.2 (2013-Dec-10)</a></li>
<li><a class="reference internal" href="#nov-19">0.6.3.1 (2013-Nov-19)</a></li>
<li><a class="reference internal" href="#nov-7">0.6.3 (2013-Nov-7)</a></li>
<li><a class="reference internal" href="#oct-7">0.6.2 (2013-Oct-7)</a></li>
<li><a class="reference internal" href="#sep-09">0.6.1 (2013-Sep-09)</a></li>
<li><a class="reference internal" href="#sep-02">0.6.0 (2013-Sep-02)</a></li>
<li><a class="reference internal" href="#b-2013-jul-23">0.5.0b (2013-Jul-23)</a></li>
<li><a class="reference internal" href="#b-2013-jun-12">0.4.0b (2013-Jun-12)</a></li>
<li><a class="reference internal" href="#b">0.3.0b</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="caveats.html"
                        title="previous chapter">&lt;no title&gt;</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="faq.html"
                        title="next chapter">F.A.Q.</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/history.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="faq.html" title="F.A.Q."
             >next</a> |</li>
        <li class="right" >
          <a href="caveats.html" title="&lt;no title&gt;"
             >previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2012,2013.
      Last updated on Mar 19, 2015.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>

<!-- 
<div id="editor-trap">
      <h3>Edit and improve this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/arq5x/gemini/blob/master/docs/content/history.rst">
          Release History
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/arq5x/gemini/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>
 -->
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', UA-39150563-1]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>


  </body>
</html>