diff --git a/arch/PowerPC/PPCDisassembler.c b/arch/PowerPC/PPCDisassembler.c
index 5170a3e..d6bb0b4 100644
--- a/arch/PowerPC/PPCDisassembler.c
+++ b/arch/PowerPC/PPCDisassembler.c
@@ -370,6 +370,7 @@ static DecodeStatus getInstruction(MCInst *MI,
 		memset(MI->flat_insn->detail, 0, sizeof(cs_detail));
 	}
 
+printf(">>> insn = %x\n", insn);
 	if (MI->csh->mode & CS_MODE_QPX) {
 		result = decodeInstruction_4(DecoderTableQPX32, MI, insn, Address, 4);
 		if (result != MCDisassembler_Fail)
@@ -379,6 +380,7 @@ static DecodeStatus getInstruction(MCInst *MI,
 	}
 
 	result = decodeInstruction_4(DecoderTable32, MI, insn, Address, 4);
+printf(">>> result = %x\n", result);
 	if (result != MCDisassembler_Fail) {
 		*Size = 4;
 		return result;
