<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>pmdtools package &mdash; SamSifter 0.5.3 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.5.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="SamSifter 0.5.3 documentation" href="index.html" />
    <link rel="up" title="SamSifter API documentation" href="api.html" />
    <link rel="next" title="samsifter package" href="samsifter.html" />
    <link rel="prev" title="SamSifter API documentation" href="api.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="samsifter.html" title="samsifter package"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="api.html" title="SamSifter API documentation"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">SamSifter 0.5.3 documentation</a> &raquo;</li>
          <li><a href="api.html" accesskey="U">SamSifter API documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="module-pmdtools">
<span id="pmdtools-package"></span><h1>pmdtools package<a class="headerlink" href="#module-pmdtools" title="Permalink to this headline">¶</a></h1>
<p>Separating ancient DNA from modern contamination.</p>
<p>This package includes a modified version of PMDtools. The original software
developed by Pontus Skoglund is available at
<a class="reference external" href="https://code.google.com/p/pmdtools/">https://code.google.com/p/pmdtools/</a>.</p>
<p>PMDtools implements a likelihood framework incorporating postmortem damage
(PMD), base quality scores and biological polymorphism to identify degraded DNA
sequences that are unlikely to originate from modern contamination. Using the
model, each sequence is assigned a PMD score, for which positive values
indicate support for the sequence being genuinely ancient. For details of the
method, please see the main paper in PNAS:
<a class="reference external" href="http://www.pnas.org/content/111/6/2229.abstract">http://www.pnas.org/content/111/6/2229.abstract</a></p>
<p>In addition, PMDtools also offers PMD-aware base quality score adjustment and
investigation of damage patterns.</p>
<p>PMDtools takes SAM-formatted input, and requires an MD tag with alignment
information. The MD tag is featured in the output of many aligners but can
otherwise be added e.g. using the SAMtools fillmd/calmd tool
(Li, Handsaker et al. 2009).</p>
<div class="section" id="module-pmdtools.pmdtools_mod">
<span id="pmdtools-mod-module"></span><h2>pmdtools_mod module<a class="headerlink" href="#module-pmdtools.pmdtools_mod" title="Permalink to this headline">¶</a></h2>
<p>Calculate and filter post-mortem degeneration score for NGS reads in SAM files.</p>
<p>Modified version of PMDtools based on v0.55 by Pontus Skoglund:</p>
<p>cite: P Skoglund, BH Northoff, MV Shunkov, AP Derevianko, S Paabo, J Krause, M
Jakobsson (2014) Separating endogenous ancient DNA from modern day
contamination in a Siberian Neandertal, PNAS, advance online 27 January</p>
<p>Included changes:</p>
<ul class="simple">
<li>count and show excluded reads due to unsupported CIGAR operations</li>
<li>count and show excluded reads due to % identity filtering</li>
<li>port from Python2 to Python3</li>
<li>bugfix: ambiguous combinations of CIGAR and MD with deletions lead to shifts
in reconstructed reference sequence</li>
<li>support for all CIGAR operations except padded references (untested)</li>
<li>support of IUPAC ambiguity codes</li>
<li>optional parameters to adapt % identity calculation to MALT defaults</li>
<li>refactoring into main method</li>
<li>use of ArgumentParser over deprecated OptionParser</li>
<li>improved PEP8 and pylint compliance</li>
<li>ReST documentation</li>
</ul>
<dl class="function">
<dt id="pmdtools.pmdtools_mod.L_match">
<tt class="descname">L_match</tt><big>(</big><em>position</em>, <em>model</em>, <em>qual</em>, <em>poly</em><big>)</big><a class="reference internal" href="_modules/pmdtools/pmdtools_mod.html#L_match"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pmdtools.pmdtools_mod.L_match" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate likelihood of a match under given model.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pos</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; Position of base from end of read.</li>
<li><strong>model</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; Distribution of damage probabilities.</li>
<li><strong>qual</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; PHRED score of base.</li>
<li><strong>poly</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; Probability of a true polymorphism at this base.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Probability of match under given model.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">float</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pmdtools.pmdtools_mod.L_mismatch">
<tt class="descname">L_mismatch</tt><big>(</big><em>position</em>, <em>model</em>, <em>qual</em>, <em>poly</em><big>)</big><a class="reference internal" href="_modules/pmdtools/pmdtools_mod.html#L_mismatch"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pmdtools.pmdtools_mod.L_mismatch" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate likelihood of a mismatch under given model.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pos</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; Position of base from end of read.</li>
<li><strong>model</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; Distribution of damage probabilities.</li>
<li><strong>qual</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; PHRED score of base.</li>
<li><strong>poly</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; Probability of a true polymorphism at this base.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Probability of mismatch under given model.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">float</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pmdtools.pmdtools_mod.adjust_quality">
<tt class="descname">adjust_quality</tt><big>(</big><em>position</em>, <em>model</em>, <em>qual</em><big>)</big><a class="reference internal" href="_modules/pmdtools/pmdtools_mod.html#adjust_quality"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pmdtools.pmdtools_mod.adjust_quality" title="Permalink to this definition">¶</a></dt>
<dd><p>Adjust base quality value (sequencing error probability).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>position</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; Position of base from end of read.</li>
<li><strong>model</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; Distribution of damage probabilities.</li>
<li><strong>qual</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; PHRED score of base.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Adjusted base quality value.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">float</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pmdtools.pmdtools_mod.fa_get">
<tt class="descname">fa_get</tt><big>(</big><em>ffile</em>, <em>chrom</em>, <em>fstart</em>, <em>fend</em>, <em>samtoolspath</em>, <em>verbose=False</em><big>)</big><a class="reference internal" href="_modules/pmdtools/pmdtools_mod.html#fa_get"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pmdtools.pmdtools_mod.fa_get" title="Permalink to this definition">¶</a></dt>
<dd><p>Unused method stub to get reference sequence from supplied FASTA?</p>
<p>Calls SAMtools in subprocess to create index of FASTA file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>ffile</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; Path to FASTA file.</li>
<li><strong>chrom</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; Chromosome number.</li>
<li><strong>fstart</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; Start position.</li>
<li><strong>fend</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; End position.</li>
<li><strong>samtoolspath</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; Path to samtools.</li>
<li><strong>verbose</strong> (<em>bool, optional</em>) &#8211; Enable output of executed commandline, defaults to False.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">reference sequence</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">str</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pmdtools.pmdtools_mod.geometric">
<tt class="descname">geometric</tt><big>(</big><em>pval</em>, <em>kval</em>, <em>constant</em><big>)</big><a class="reference internal" href="_modules/pmdtools/pmdtools_mod.html#geometric"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pmdtools.pmdtools_mod.geometric" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate geometrically distributed probability.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pval</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; Parameter p of geometric distribution.</li>
<li><strong>kval</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; Parameter k of geometric distribution.</li>
<li><strong>constant</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; Constant value of geometric distribution.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Probability of deamination based on geometric distribution.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">float</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pmdtools.pmdtools_mod.main">
<tt class="descname">main</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/pmdtools/pmdtools_mod.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pmdtools.pmdtools_mod.main" title="Permalink to this definition">¶</a></dt>
<dd><p>Executable to calculate post-mortem degradation scores for SAM files.</p>
<p>See <tt class="docutils literal"><span class="pre">--help</span></tt> for details on expected arguments. Takes input only on
STDIN. Logs messages to STDERR and writes processed SAM file to
STDOUT.</p>
</dd></dl>

<dl class="function">
<dt id="pmdtools.pmdtools_mod.phred2prob">
<tt class="descname">phred2prob</tt><big>(</big><em>Q</em><big>)</big><a class="reference internal" href="_modules/pmdtools/pmdtools_mod.html#phred2prob"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pmdtools.pmdtools_mod.phred2prob" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert PHRED score to probability of sequencing error.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>Q</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; PHRED score.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">The probability of a sequencing error.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">float</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pmdtools.pmdtools_mod.prob">
<tt class="descname">prob</tt><big>(</big><em>pos</em>, <em>model</em>, <em>qual</em>, <em>poly</em>, <em>match=True</em><big>)</big><a class="reference internal" href="_modules/pmdtools/pmdtools_mod.html#prob"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pmdtools.pmdtools_mod.prob" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate probability of either match or mismatch under given model.</p>
<p>Expects distance of base from end of read, a pre-computed distribution of
damage probabilities, PHRED score of base (probability of sequencing
error) and probability of true polymorphism.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pos</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; Position of base from end of read.</li>
<li><strong>model</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; Distribution of damage probabilities.</li>
<li><strong>qual</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#int" title="(in Python v2.7)"><em>int</em></a>) &#8211; PHRED score of base.</li>
<li><strong>poly</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; Probability of a true polymorphism at this base.</li>
<li><strong>match</strong> (<em>bool, optional</em>) &#8211; Switch between calculation for a match (default) or mismatch.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Probability of either match or mismatch under given model.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">float</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pmdtools.pmdtools_mod.prob2ascii">
<tt class="descname">prob2ascii</tt><big>(</big><em>P</em>, <em>offset=33</em><big>)</big><a class="reference internal" href="_modules/pmdtools/pmdtools_mod.html#prob2ascii"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pmdtools.pmdtools_mod.prob2ascii" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert probability of sequencing error to ASCII-encoded PHRED score.</p>
<p>By default uses ASCII offset 33 (Illumina standard).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>P</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; The probability of a sequencing error.</li>
<li><strong>offset</strong> (<em>int, optional</em>) &#8211; ASCII encoding offset, defaults to 33.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pmdtools.pmdtools_mod.prob2phred">
<tt class="descname">prob2phred</tt><big>(</big><em>P</em><big>)</big><a class="reference internal" href="_modules/pmdtools/pmdtools_mod.html#prob2phred"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pmdtools.pmdtools_mod.prob2phred" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert probability of sequencing error to PHRED score.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>P</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#float" title="(in Python v2.7)"><em>float</em></a>) &#8211; The probability of a sequencing error.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">PHRED score.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">float</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="pmdtools.pmdtools_mod.score_pmd">
<tt class="descname">score_pmd</tt><big>(</big><em>read</em>, <em>ref</em>, <em>quals</em>, <em>ancient_model_deam</em>, <em>modern_model_deam</em>, <em>adjustment_model_deam=None</em>, <em>polymorphism_contamination=0.001</em>, <em>polymorphism_ancient=0.001</em>, <em>adjustbaseq_all=False</em>, <em>adjustbaseq=False</em>, <em>baseq=0</em>, <em>deamination=False</em>, <em>cpg=False</em>, <em>nocpg=False</em>, <em>UDGhalf=False</em>, <em>PMDSprim=False</em><big>)</big><a class="reference internal" href="_modules/pmdtools/pmdtools_mod.html#score_pmd"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pmdtools.pmdtools_mod.score_pmd" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate post-mortem degradation score (PMDS).</p>
<p>Requires full read and reference sequence including clips, skips and gaps
as recovered from CIGAR and MD to use the correct distance from 5&#8217; or 3&#8217;
end in the PMDS calculation. However only the aligned read bases with
PHRED scores are considered for PMD scoring.</p>
<p>Optional adjustment of base qualities requires additional adjustment model.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>read</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; Full read sequence.</li>
<li><strong>ref</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#str" title="(in Python v2.7)"><em>str</em></a>) &#8211; Aligned reference sequence including clips, skips and gaps.</li>
<li><strong>quals</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; Quality scores for aligned bases.</li>
<li><strong>ancient_model_deam</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; Probability distribution for ancient deamination.</li>
<li><strong>modern_model_deam</strong> (<a class="reference external" href="http://docs.python.org/library/functions.html#list" title="(in Python v2.7)"><em>list</em></a>) &#8211; Probability distribution for modern deamination.</li>
<li><strong>adjustment_model_deam</strong> (<em>list, optional</em>) &#8211; Adjustments to the deamination probabilities, defaults to None.</li>
<li><strong>polymorphism_contamination</strong> (<em>float, optional</em>) &#8211; Probability of polymorphism due to contamination, defaults to 0.001.</li>
<li><strong>polymorphism_ancient</strong> (<em>float, optional</em>) &#8211; Probability of polymorphism due to age, defaults to 0.001.</li>
<li><strong>adjustbaseq_all</strong> (<em>bool, optional</em>) &#8211; Switch to adjust all base quality values, defaults to False.</li>
<li><strong>adjustbaseq</strong> (<em>bool, optional</em>) &#8211; Switch to adjust affected base quality values, defaults to False.</li>
<li><strong>baseq</strong> (<em>int, optional</em>) &#8211; Minimum quality value for processed bases, defaults to 0</li>
<li><strong>deamination</strong> (<em>bool, optional</em>) &#8211; Enable output of base frequencies in deaminated context, defaults to
False.</li>
<li><strong>cpg</strong> (<em>bool, optional</em>) &#8211; Only use Cs and Gs in CpG context, defaults to False.</li>
<li><strong>nocpg</strong> (<em>bool, optional</em>) &#8211; Do NOT use Cs and Gs in CpG context, defaults to False.</li>
<li><strong>UDGhalf</strong> (<em>bool, optional</em>) &#8211; Only use Cs and Gs in CpG context, the first and last base are used
regardless of dinucleotide context. Defaults to False.</li>
<li><strong>PMDSprim</strong> (<em>bool, optional</em>) &#8211; Defaults to False.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li><em>float</em> &#8211;
Likelihood of deamination under ancient DNA model.</li>
<li><em>float</em> &#8211;
Likelihood of deamination under ancient DNA model.</li>
<li><em>float</em> &#8211;
Post-mortem degradation score</li>
<li><em>list</em> &#8211;
Adjusted base qualities.</li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">pmdtools package</a><ul>
<li><a class="reference internal" href="#module-pmdtools.pmdtools_mod">pmdtools_mod module</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="api.html"
                        title="previous chapter">SamSifter API documentation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="samsifter.html"
                        title="next chapter">samsifter package</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/pmdtools.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="samsifter.html" title="samsifter package"
             >next</a> |</li>
        <li class="right" >
          <a href="api.html" title="SamSifter API documentation"
             >previous</a> |</li>
        <li><a href="index.html">SamSifter 0.5.3 documentation</a> &raquo;</li>
          <li><a href="api.html" >SamSifter API documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2015, Florian Aldehoff.
      Last updated on Mar 09, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
  </body>
</html>