<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Querying the GEMINI database</title>
    
    <link rel="stylesheet" href="../_static/labibi.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.12.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/gemini.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="Built-in analysis tools" href="tools.html" />
    <link rel="prev" title="Preprocessing and Loading a VCF file into GEMINI" href="preprocessing.html" />


<script src="../_static/labibi.js"></script>



<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', '']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>


  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tools.html" title="Built-in analysis tools"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="preprocessing.html" title="Preprocessing and Loading a VCF file into GEMINI"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  
  <div class="section" id="querying-the-gemini-database">
<h1>Querying the GEMINI database<a class="headerlink" href="#querying-the-gemini-database" title="Permalink to this headline">¶</a></h1>
<p>The real power in the <tt class="docutils literal"><span class="pre">GEMINI</span></tt> framework lies in the fact that all of your
genetic variants have been stored in a convenient database in the context of a
wealth of genome annotations that facilitate variant interpretation.  The
expressive power of SQL allows one to pose intricate questions of one&#8217;s variation
data.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If you are unfamiliar with SQL, <a class="reference external" href="http://sqlzoo.net/">sqlzoo</a> has a decent
online tutorial describing the basics.  Really all you need to learn is the
SELECT statement, and the examples below will give you a flavor of how to
compose base SQL queries against the GEMINI framework.</p>
</div>
<div class="section" id="basic-queries">
<h2>Basic queries<a class="headerlink" href="#basic-queries" title="Permalink to this headline">¶</a></h2>
<p>GEMINI has a specific tool for querying a gemini database that has been <tt class="docutils literal"><span class="pre">load``ed</span>
<span class="pre">using</span> <span class="pre">the</span> <span class="pre">``gemini</span> <span class="pre">load</span></tt> command.  That&#8217;s right, the tool is called
<tt class="docutils literal"><span class="pre">gemini</span> <span class="pre">query</span></tt>. Below are a few basic queries that give you a sense of how to
interact with the gemini database using the <tt class="docutils literal"><span class="pre">query</span></tt> tool.</p>
<ol class="arabic simple">
<li>Extract all transitions with a call rate &gt; 95%</li>
</ol>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select * from variants \</span>
<span class="s2">                      where sub_type = &#39;ts&#39; \</span>
<span class="s2">                      and call_rate &gt;= 0.95&quot;</span> my.db
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li>Extract all loss-of-function variants with an alternate allele frequency &lt; 1%:</li>
</ol>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select * from variants \</span>
<span class="s2">                      where is_lof = 1 \</span>
<span class="s2">                      and aaf &gt;= 0.01&quot;</span> my.db
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li>Extract the nucleotide diversity for each variant:</li>
</ol>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select chrom, start, end, pi from variants&quot;</span> my.db
</pre></div>
</div>
<ol class="arabic simple" start="4">
<li>Combine <tt class="docutils literal"><span class="pre">GEMINI</span></tt> with <tt class="docutils literal"><span class="pre">bedtools</span></tt> to compute nucleotide diversity estimates across 100kb windows:</li>
</ol>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select chrom, start, end, pi from variants \</span>
<span class="s2">                      order by chrom, start, end&quot;</span> my.db | <span class="se">\</span>
  bedtools map -a hg19.windows.bed -b - -c 4 -o mean
</pre></div>
</div>
</div>
<div class="section" id="selecting-sample-genotypes">
<h2>Selecting sample genotypes<a class="headerlink" href="#selecting-sample-genotypes" title="Permalink to this headline">¶</a></h2>
<p>The above examples illustrate <em>ad hoc</em> queries that do not request or filter
upon the genotypes of individual samples.  Since <tt class="docutils literal"><span class="pre">GEMINI</span></tt> stores the genotype
information for each variant in compressed arrays that are stored as BLOBs
in the database, standard SQL queries cannot directly access individual
genotypes. However, we have enhanced the SQL syntax to support such queries
with C &#8220;struct-like&#8221; access.  For example, to retrieve the alleles for a given
sample&#8217;s (in this case, sample 1094PC0009), one would add <tt class="docutils literal"><span class="pre">gts.1094PC0009</span></tt>
to the select statement.</p>
<p>Here is an example of selecting the genotype alleles for four
different samples (note the examples below use the test.snpEff.vcf.db
file that is created in the ./test directory when you run the
<cite>bash master-test.sh</cite> command as described above):</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select chrom, start, end, ref, alt, gene, \</span>
<span class="s2">                          gts.1094PC0005, \</span>
<span class="s2">                          gts.1094PC0009, \</span>
<span class="s2">                          gts.1094PC0012, \</span>
<span class="s2">                          gts.1094PC0013 \</span>
<span class="s2">                   from variants&quot;</span> test.snpEff.vcf.db

chr1        30547   30548   T       G       FAM138A ./.     ./.     ./.     ./.
chr1        30859   30860   G       C       FAM138A G/G     G/G     G/G     G/G
chr1        30866   30869   CCT     C       FAM138A CCT/CCT CCT/CCT CCT/C   CCT/CCT
chr1        30894   30895   T       C       FAM138A T/C     T/C     T/T     T/T
chr1        30922   30923   G       T       FAM138A ./.     ./.     ./.     ./.
chr1        69269   69270   A       G       OR4F5   ./.     ./.     G/G     G/G
chr1        69427   69428   T       G       OR4F5   T/T     T/T     T/T     T/T
chr1        69510   69511   A       G       OR4F5   ./.     ./.     A/G     A/G
chr1        69760   69761   A       T       OR4F5   A/A     A/T     A/A     A/A
chr1        69870   69871   G       A       OR4F5   ./.     G/G     G/G     G/G
</pre></div>
</div>
<p>You can also add a header so that you can keep track of who&#8217;s who:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select chrom, start, end, ref, alt, gene, \</span>
<span class="s2">                          gts.1094PC0005, \</span>
<span class="s2">                          gts.1094PC0009, \</span>
<span class="s2">                          gts.1094PC0012, \</span>
<span class="s2">                          gts.1094PC0013 \</span>
<span class="s2">                   from variants&quot;</span> <span class="se">\</span>
                   --header test.snpEff.vcf.db

chrom       start   end     ref     alt     gene gts.1094PC0005     gts.1094PC0009  gts.1094PC0012  gts.1094PC0013
chr1        30547   30548   T       G       FAM138A ./.     ./.     ./.     ./.
chr1        30859   30860   G       C       FAM138A G/G     G/G     G/G     G/G
chr1        30866   30869   CCT     C       FAM138A CCT/CCT CCT/CCT CCT/C   CCT/CCT
chr1        30894   30895   T       C       FAM138A T/C     T/C     T/T     T/T
chr1        30922   30923   G       T       FAM138A ./.     ./.     ./.     ./.
chr1        69269   69270   A       G       OR4F5   ./.     ./.     G/G     G/G
chr1        69427   69428   T       G       OR4F5   T/T     T/T     T/T     T/T
chr1        69510   69511   A       G       OR4F5   ./.     ./.     A/G     A/G
chr1        69760   69761   A       T       OR4F5   A/A     A/T     A/A     A/A
chr1        69870   69871   G       A       OR4F5   ./.     G/G     G/G     G/G
</pre></div>
</div>
<p>Let&#8217;s now get the genotype and the depth of aligned sequence observed for a
sample so that we can assess the confidence in the genotype:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select chrom, start, end, ref, alt, gene,</span>
<span class="s2">                      gts.1094PC0005, \</span>
<span class="s2">                      gt_depths.1094PC0005 \</span>
<span class="s2">               from variants&quot;</span> test.snpEff.vcf.db

chr1    30547   30548   T       G       FAM138A ./.     -1
chr1    30859   30860   G       C       FAM138A G/G     7
chr1    30866   30869   CCT     C       FAM138A CCT/CCT 8
chr1    30894   30895   T       C       FAM138A T/C     8
chr1    30922   30923   G       T       FAM138A ./.     -1
chr1    69269   69270   A       G       OR4F5   ./.     -1
chr1    69427   69428   T       G       OR4F5   T/T     2
chr1    69510   69511   A       G       OR4F5   ./.     -1
chr1    69760   69761   A       T       OR4F5   A/A     1
chr1    69870   69871   G       A       OR4F5   ./.     -1
</pre></div>
</div>
</div>
<div class="section" id="selecting-sample-genotypes-based-on-wildcards">
<h2>Selecting sample genotypes based on &#8220;wildcards&#8221;.<a class="headerlink" href="#selecting-sample-genotypes-based-on-wildcards" title="Permalink to this headline">¶</a></h2>
<p>The above examples demonstrate how one can select individual sample genotype
information by explicitly listing each column and sample that one wishes to see.
Obviously, this can become tedious when a project involves hundreds or thousands of samples
&#8212; if you wanted to see genotype information for the 345 of 1145 affected samples in your study,
you would have to type each and every column.sample name out. Brutal.</p>
<div class="section" id="the-wildcard">
<h3>The &#8220;*&#8221; wildcard<a class="headerlink" href="#the-wildcard" title="Permalink to this headline">¶</a></h3>
<p>To get around this, one can bulk-select sample genotype information using &#8220;wildcards&#8221;. The column and the wildcard must each be surrounded with parentheses and separated by a period. The &#8220;*&#8221; is a shortcut (wildcard) meaning &#8220;all samples&#8221;.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The syntax for SELECTing genotype columns using a wildcard is <tt class="docutils literal"><span class="pre">(COLUMN).(WILDCARD)</span></tt>.</p>
</div>
<p>For example, a shortcut to reporting the genotype for <em>all</em> samples (in this case 4) in the study, one could do the following:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header -q <span class="s2">&quot;select chrom, start, end, ref, alt, gene, (gts).(*) \</span>
<span class="s2">                            from variants&quot;</span> extended_ped.db
chrom start end ref alt gene  gts.M10475  gts.M10478  gts.M10500  gts.M128215
chr10 1142207 1142208 T C WDR37 C/C C/C C/C C/C
chr10 48003991  48003992  C T ASAH2C  T/T C/T C/T C/C
chr10 52004314  52004315  T C ASAH2 ./. ./. C/C C/C
chr10 52497528  52497529  G C ASAH2B  ./. C/C C/C ./.
chr10 126678091 126678092 G A CTBP2 G/G G/G G/G G/A
chr10 135210790 135210791 T C MTG1.1  T/T C/C C/C T/T
chr10 135336655 135336656 G A SPRN  ./. A/A ./. A/A
chr10 135369531 135369532 T C SYCE1 T/T T/C T/C T/T
chr16 72057434  72057435  C T DHODH C/T C/C C/C C/C
</pre></div>
</div>
</div>
<div class="section" id="wildcards-based-on-sample-attributes">
<h3>Wildcards based on sample attributes<a class="headerlink" href="#wildcards-based-on-sample-attributes" title="Permalink to this headline">¶</a></h3>
<p>To report the genotypes for solely those samples that are affected (phenotype == 2) with the phenotype in question, one could do the following:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header -q <span class="s2">&quot;select chrom, start, end, ref, alt, gene, (gts).(phenotype==2) \</span>
<span class="s2">                            from variants&quot;</span> extended_ped.db
chrom start end ref alt gene  gts.M10478  gts.M10500
chr10 1142207 1142208 T C WDR37 C/C C/C
chr10 48003991  48003992  C T ASAH2C  C/T C/T
chr10 52004314  52004315  T C ASAH2 ./. C/C
chr10 52497528  52497529  G C ASAH2B  C/C C/C
chr10 126678091 126678092 G A CTBP2 G/G G/G
chr10 135210790 135210791 T C MTG1.1  C/C C/C
chr10 135336655 135336656 G A SPRN  A/A ./.
chr10 135369531 135369532 T C SYCE1 T/C T/C
chr16 72057434  72057435  C T DHODH C/C C/C
</pre></div>
</div>
<p>One can add multiple wildcard criteria as well:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header -q <span class="s2">&quot;select chrom, start, end, ref, alt, gene, (gts).(phenotype==1 and hair_color==&#39;blue&#39;) \</span>
<span class="s2">                            from variants&quot;</span> extended_ped.db
chrom start end ref alt gene  gts.M128215
chr10 1142207 1142208 T C WDR37 C/C
chr10 48003991  48003992  C T ASAH2C  C/C
chr10 52004314  52004315  T C ASAH2 C/C
chr10 52497528  52497529  G C ASAH2B  ./.
chr10 126678091 126678092 G A CTBP2 G/A
chr10 135210790 135210791 T C MTG1.1  T/T
chr10 135336655 135336656 G A SPRN  A/A
chr10 135369531 135369532 T C SYCE1 T/T
chr16 72057434  72057435  C T DHODH C/C
</pre></div>
</div>
</div>
</div>
<div class="section" id="gt-filter-filtering-on-genotypes">
<h2><tt class="docutils literal"><span class="pre">--gt-filter</span></tt> Filtering on genotypes<a class="headerlink" href="#gt-filter-filtering-on-genotypes" title="Permalink to this headline">¶</a></h2>
<p>Now, we often want to focus only on variants where a given sample has a
specific genotype (e.g., looking for homozygous variants in family trios).
Unfortunately, we cannot directly do this in the SQL query, but the <cite>gemini query</cite>
tool has an option called <cite>&#8211;gt-filter</cite> that allows one to specify filters to
apply to the returned rows.  The rules followed in the <cite>&#8211;gt-filter</cite> option
follow Python syntax.</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">As you will see from the examples below, appropriate use of the &#8211;gt-filter
option will allow you to compose queries that return variants meeting
inheritance patterns that are relevant to the disease model of interest
in your study.</p>
</div>
<p>As an example, let&#8217;s only return rows where sample
1094PC0012 is heterozygous.  In order to do this, we apply a filter to the
<cite>gt_types</cite> columns for this individual:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select chrom, start, end, ref, alt, gene,</span>
<span class="s2">                      gts.1094PC0005, \</span>
<span class="s2">                      gts.1094PC0009, \</span>
<span class="s2">                      gts.1094PC0012, \</span>
<span class="s2">                      gts.1094PC0013 \</span>
<span class="s2">               from variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;gt_types.1094PC0012 == HET&quot;</span> <span class="se">\</span>
               --header <span class="se">\</span>
               test.snpEff.vcf.db

chrom   start   end     ref     alt     gene gts.1094PC0005     gts.1094PC0009  gts.1094PC0012  gts.1094PC0013
chr1    30866   30869   CCT     C       FAM138A CCT/CCT CCT/CCT CCT/C   CCT/CCT
chr1    69510   69511   A       G       OR4F5   ./.     ./.     A/G     A/G
</pre></div>
</div>
<p>Now let&#8217;s be a bit less restrictive and return variants where either sample
1094PC0012 is heterozygous or sample 1094PC0005 is homozygous for the reference
allele:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select chrom, start, end, ref, alt, gene,</span>
<span class="s2">                      gts.1094PC0005, \</span>
<span class="s2">                      gts.1094PC0009, \</span>
<span class="s2">                      gts.1094PC0012, \</span>
<span class="s2">                      gts.1094PC0013 \</span>
<span class="s2">               from variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;gt_types.1094PC0012 == HET or \</span>
<span class="s2">               gt_types.1094PC0005 == HOM_REF&quot;</span> <span class="se">\</span>
               --header <span class="se">\</span>
               test.snpEff.vcf.db

chrom   start   end     ref     alt     gene gts.1094PC0005     gts.1094PC0009  gts.1094PC0012  gts.1094PC0013
chr1    30859   30860   G       C       FAM138A G/G     G/G     G/G     G/G
chr1    30866   30869   CCT     C       FAM138A CCT/CCT CCT/CCT CCT/C   CCT/CCT
chr1    69427   69428   T       G       OR4F5   T/T     T/T     T/T     T/T
chr1    69510   69511   A       G       OR4F5   ./.     ./.     A/G     A/G
chr1    69760   69761   A       T       OR4F5   A/A     A/T     A/A     A/A
</pre></div>
</div>
<p>Eh, I changed my mind, let&#8217;s restrict the above to those variants where sample
1094PC0012 must also be heterozygous:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select chrom, start, end, ref, alt, gene,</span>
<span class="s2">                      gts.1094PC0005, \</span>
<span class="s2">                      gts.1094PC0009, \</span>
<span class="s2">                      gts.1094PC0012, \</span>
<span class="s2">                      gts.1094PC0013 \</span>
<span class="s2">               from variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;(gt_types.1094PC0012 == HET or \</span>
<span class="s2">               gt_types.1094PC0005 == HOM_REF) \</span>
<span class="s2">               and \</span>
<span class="s2">               (gt_types.1094PC0013 == HET)&quot;</span> <span class="se">\</span>
               --header <span class="se">\</span>
               test.snpEff.vcf.db

 chrom  start   end     ref     alt     gene gts.1094PC0005     gts.1094PC0009  gts.1094PC0012  gts.1094PC0013
</pre></div>
</div>
</div>
<div class="section" id="gt-filter-wildcard-filtering-on-genotype-columns">
<h2><tt class="docutils literal"><span class="pre">--gt-filter</span></tt> Wildcard filtering on genotype columns.<a class="headerlink" href="#gt-filter-wildcard-filtering-on-genotype-columns" title="Permalink to this headline">¶</a></h2>
<p>Many times, we want to be able to apply the same rule to multiple samples
without having to enter the rule over and over again for each sample. For example,
let&#8217;s imaging there are 100 samples in your study and you only want to report variants
where every sample has an observed alignment depth of at least 20 reads. Traditionally,
one would have enter each of the 100 samples from the command line as follows:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select chrom, start, end, ref, alt, gene from variants&quot;</span> <span class="se">\</span>
                     --gt-filter <span class="s2">&quot;gt_depths.sample1 &gt;= 20 and \</span>
<span class="s2">                                  gt_depths.sample2 &gt;= 20 and \</span>
<span class="s2">                                  gt_depths.sample3 &gt;= 20 and \</span>
<span class="s2">                                  ...</span>
<span class="s2">                                  gt_depths.sample100 &gt;= 20&quot;</span> <span class="se">\</span>
                     extended_ped.db
</pre></div>
</div>
<div class="section" id="the-basics">
<h3>The basics.<a class="headerlink" href="#the-basics" title="Permalink to this headline">¶</a></h3>
<p>Obviously, this is deeply painful. Thankfully, there an option allowing the use of &#8220;wildcards&#8221; to prevent this.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The syntax of the wildcard <tt class="docutils literal"><span class="pre">--gt-filters</span></tt> is <tt class="docutils literal"><span class="pre">(COLUMN).(SAMPLE_WILDCARD).(SAMPLE_WILDCARD_RULE).(RULE_ENFORCEMENT)</span></tt>.</p>
</div>
<p>For example, using wildcards, the above could be converted to:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select chrom, start, end, ref, alt, gene from variants&quot;</span> <span class="se">\</span>
                     --gt-filter <span class="s2">&quot;(gt_depths).(*).(&gt;=20).(all)&quot;</span> <span class="se">\</span>
                     extended_ped.db
</pre></div>
</div>
<p>Obviously, this makes things much simpler.</p>
</div>
<div class="section" id="the-all-operator">
<h3>The <tt class="docutils literal"><span class="pre">all</span></tt> operator<a class="headerlink" href="#the-all-operator" title="Permalink to this headline">¶</a></h3>
<p>One can also apply wildcards that select samples based on the values in specific
columns in the <tt class="docutils literal"><span class="pre">samples</span></tt> table. For example, let&#8217;s imagine we wanted to require that variants
are returned only in cases where <em>ALL</em> the affected individuals in the study (i.e., the <tt class="docutils literal"><span class="pre">phenotype</span></tt> column in the <tt class="docutils literal"><span class="pre">samples</span></tt> table is equal to <tt class="docutils literal"><span class="pre">2</span></tt>) have non-reference genotypes. We could do the following:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header <span class="se">\</span>
               -q <span class="s2">&quot;SELECT chrom, start, end, ref, alt, gene, (gts).(phenotype==2) \</span>
<span class="s2">                   FROM variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;(gt_types).(phenotype==2).(!=HOM_REF).(all)&quot;</span> <span class="se">\</span>
               extended_ped.db
chrom start end ref alt gene  gts.M10478  gts.M10500
chr10 1142207 1142208 T C WDR37 C/C C/C
chr10 48003991  48003992  C T ASAH2C  C/T C/T
chr10 52004314  52004315  T C ASAH2 ./. C/C
chr10 52497528  52497529  G C ASAH2B  C/C C/C
chr10 135210790 135210791 T C MTG1.1  C/C C/C
chr10 135336655 135336656 G A SPRN  A/A ./.
chr10 135369531 135369532 T C SYCE1 T/C T/C
</pre></div>
</div>
<p>Or perhaps we wanted to be more restrictive. We could also enforce that the affected individuals also had at least 20 aligned reads at such variant sites:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header <span class="se">\</span>
               -q <span class="s2">&quot;SELECT chrom, start, end, ref, alt, gene, \</span>
<span class="s2">                          (gts).(phenotype==2), (gt_depths).(phenotype==2) \</span>
<span class="s2">                   FROM variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;(gt_types).(phenotype==2).(!=HOM_REF).(all) \</span>
<span class="s2">                            and \</span>
<span class="s2">                            (gt_depths).(phenotype==2).(&gt;=20).(all)&quot;</span> <span class="se">\</span>
               extended_ped.db
chrom start end ref alt gene  gts.M10478  gts.M10500  gt_depths.M10478  gt_depths.M10500
chr10 1142207 1142208 T C WDR37 C/C C/C 29  24
chr10 48003991  48003992  C T ASAH2C  C/T C/T 38  44
chr10 135369531 135369532 T C SYCE1 T/C T/C 32  21
</pre></div>
</div>
</div>
<div class="section" id="the-any-operator">
<h3>The <tt class="docutils literal"><span class="pre">any</span></tt> operator<a class="headerlink" href="#the-any-operator" title="Permalink to this headline">¶</a></h3>
<p>The examples provided thus far have deomnstrated how to enforce that <em>ALL</em> of the samples meeting specific criteria meet the same genotype column restrictions. However, clearly there are cases where one would want to be less restrictive. In order to accomondate such queries, there are three other enforcement operators allowed: <tt class="docutils literal"><span class="pre">any</span></tt>, <tt class="docutils literal"><span class="pre">none</span></tt>, and <tt class="docutils literal"><span class="pre">count</span></tt>.</p>
<p>For example, perhaps we want to relax the above query a bit and only require that at least one (i.e., <tt class="docutils literal"><span class="pre">any</span></tt>) of the affected samples has depth &gt; 20:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header <span class="se">\</span>
               -q <span class="s2">&quot;SELECT chrom, start, end, ref, alt, gene, \</span>
<span class="s2">                          (gts).(phenotype==2), (gt_depths).(phenotype==2) \</span>
<span class="s2">                   FROM variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;(gt_types).(phenotype==2).(!=HOM_REF).(all) \</span>
<span class="s2">                            and \</span>
<span class="s2">                            (gt_depths).(phenotype==2).(&gt;=20).(any)&quot;</span> <span class="se">\</span>
               extended_ped.db
</pre></div>
</div>
</div>
<div class="section" id="the-none-operator">
<h3>The <tt class="docutils literal"><span class="pre">none</span></tt> operator<a class="headerlink" href="#the-none-operator" title="Permalink to this headline">¶</a></h3>
<p>Or we enforce that <tt class="docutils literal"><span class="pre">none</span></tt> of the affected samples have depth less than 10:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header <span class="se">\</span>
               -q <span class="s2">&quot;SELECT chrom, start, end, ref, alt, gene, \</span>
<span class="s2">                          (gts).(phenotype==2), (gt_depths).(phenotype==2) \</span>
<span class="s2">                   FROM variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;(gt_types).(phenotype==2).(!=HOM_REF).(all) \</span>
<span class="s2">                            and \</span>
<span class="s2">                            (gt_depths).(phenotype==2).(&lt;10).(none)&quot;</span> <span class="se">\</span>
               extended_ped.db
</pre></div>
</div>
</div>
<div class="section" id="the-count-operator">
<h3>The <tt class="docutils literal"><span class="pre">count</span></tt> operator<a class="headerlink" href="#the-count-operator" title="Permalink to this headline">¶</a></h3>
<p>Finally, we could enforce that at most 2 of all the samples in the study have depths &lt; 10:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header <span class="se">\</span>
               -q <span class="s2">&quot;SELECT chrom, start, end, ref, alt, gene, \</span>
<span class="s2">                          (gts).(phenotype==2), (gt_depths).(phenotype==2) \</span>
<span class="s2">                   FROM variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;(gt_types).(phenotype==2).(!=HOM_REF).(all) \</span>
<span class="s2">                            and \</span>
<span class="s2">                            (gt_depths).(phenotype==2).(&lt;10).(count &lt;= 2)&quot;</span> <span class="se">\</span>
               extended_ped.db
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <tt class="docutils literal"><span class="pre">count</span></tt> operator allows the following comparisons: <tt class="docutils literal"><span class="pre">&gt;</span></tt>, <tt class="docutils literal"><span class="pre">&gt;=</span></tt>, <tt class="docutils literal"><span class="pre">&lt;</span></tt>, <tt class="docutils literal"><span class="pre">&lt;=</span></tt>, <tt class="docutils literal"><span class="pre">==</span></tt>, <tt class="docutils literal"><span class="pre">&lt;&gt;</span></tt>, and <tt class="docutils literal"><span class="pre">!=</span></tt>.</p>
</div>
</div>
<div class="section" id="example-scenario">
<h3>Example scenario.<a class="headerlink" href="#example-scenario" title="Permalink to this headline">¶</a></h3>
<p>One usage of the wildcard functionality is screening for variants
that are present in affected individuals yet absent from unaffected individuals (this is obviously an idealized scenario).</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header <span class="se">\</span>
               -q <span class="s2">&quot;SELECT chrom, start, end, ref, alt, gene, (gts).(*) \</span>
<span class="s2">                   FROM variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;(gt_types).(phenotype==1).(==HOM_REF).(all) \</span>
<span class="s2">                            and \</span>
<span class="s2">                            (gt_depths).(phenotype==2).(==HOM_REF).(none)&quot;</span> <span class="se">\</span>
               extended_ped.db
</pre></div>
</div>
</div>
<div class="section" id="more-complexity">
<h3>More complexity.<a class="headerlink" href="#more-complexity" title="Permalink to this headline">¶</a></h3>
<p>One can also combine wildcard filters with &#8220;basic&#8221; genotype filters
to create more complex logic.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header <span class="se">\</span>
               -q <span class="s2">&quot;SELECT chrom, start, end, ref, alt, gene, (gts).(*) \</span>
<span class="s2">                   FROM variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;((gt_types).(phenotype==1).(==HOM_REF).(all) \</span>
<span class="s2">                              and \</span>
<span class="s2">                             (gt_depths).(phenotype==2).(==HOM_REF).(none)) \</span>
<span class="s2">                            or gt.NA12878 == HET&quot;</span> <span class="se">\</span>
               extended_ped.db
</pre></div>
</div>
</div>
<div class="section" id="other-details">
<h3>Other details.<a class="headerlink" href="#other-details" title="Permalink to this headline">¶</a></h3>
<p>The system is fairly flexible in that it allows one to wildcard-select samples based on custom columns
that have been added to the <tt class="docutils literal"><span class="pre">samples</span></tt> table based upon a custom PED file.  For example, let&#8217;s imaging our
custom PED file had an extra column defining the hair color of each sample. We coukd use that to restrict interesting variants to those where samples with blue hair were heterozygous:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;SELECT chrom, start, end, ref, alt, gene FROM variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;(gt_types).(hair_color=&#39;blue&#39;).(==HET).(all)&quot;</span> <span class="se">\</span>
               extended_ped.db
</pre></div>
</div>
<p>Or possibly, you want to stratify based on sub-population:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;SELECT chrom, start, end, ref, alt, gene FROM variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;(gt_types).(population=&#39;CEU&#39;).(==HET).(all) \</span>
<span class="s2">                             and \</span>
<span class="s2">                            (gt_types).(population=&#39;YRI&#39;).(==HOM_ALT).(any)&quot;</span> <span class="se">\</span>
               extended_ped.db
</pre></div>
</div>
<p>One can also base the wildcard on multiple criteria (in this case, brown hair and affected):</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;SELECT chrom, start, end, ref, alt, gene FROM variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;(gt_types).(hair_color==&#39;brown&#39; and phenotype==2).(!= HET).(all)&quot;</span> <span class="se">\</span>
               extended_ped.db
</pre></div>
</div>
<p>Lastly, wildcards can, of course, be combined with non-wildcard criteria:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;SELECT chrom, start, end, gene FROM variants&quot;</span> <span class="se">\</span>
               --gt-filter <span class="s2">&quot;(gt_types).(hair_color==&#39;brown&#39; and phenotype==2).(!= HET).(all) \</span>
<span class="s2">                             and \</span>
<span class="s2">                            gt_types.M128215 == HOM_REF&quot;</span> <span class="se">\</span>
               extended_ped.db
</pre></div>
</div>
<p>Hopefully this gives you a sense of what you can do with the &#8220;wildcard&#8221; genotype filter functionality.</p>
</div>
</div>
<div class="section" id="show-samples-finding-out-which-samples-have-a-variant">
<h2><tt class="docutils literal"><span class="pre">--show-samples</span></tt> Finding out which samples have a variant<a class="headerlink" href="#show-samples-finding-out-which-samples-have-a-variant" title="Permalink to this headline">¶</a></h2>
<p>While exploring your data you might hit on a set of interesting variants and want to know
which of your samples have that variant in them. You can display the samples containing
a variant with the &#8211;show-sample-variants flag:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header --show-samples -q <span class="s2">&quot;select chrom, start, end, ref, alt \</span>
<span class="s2">                                from variants where is_lof=1 limit 5&quot;</span> test.query.db

chrom   start   end     ref     alt     variant_samples HET_samples     HOM_ALT_samples
chr1    874815  874816  C       CT      1478PC0006B,1478PC0007B,1478PC0010,1478PC0013B,1478PC0022B,1478PC0023B,1478PC0025,1719PC0007,1719PC0009,1719PC0010,1719PC0022   1478PC0006B,1478PC0007B,1478PC0010,1478PC0013B,1478PC0022B,1478PC0023B,1719PC0007,1719PC0009,1719PC0010 1478PC0025,1719PC0022
chr1    1140811 1140813 TC      T       1478PC0011      1478PC0011
chr1    1219381 1219382 C       G       1719PC0012      1719PC0012
chr1    1221487 1221490 CAA     C       1478PC0004      1478PC0004
</pre></div>
</div>
<p>variant_samples is a list of all of the samples with a variant, HET_samples is the subset
of those heterozygous for the variant and HOM_ALT_samples is the subset homozygous for
the variant.</p>
</div>
<div class="section" id="show-samples-format-sampledetail-provide-a-flattened-view-of-samples">
<h2><tt class="docutils literal"><span class="pre">--show-samples</span> <span class="pre">--format</span> <span class="pre">sampledetail</span></tt> Provide a flattened view of samples<a class="headerlink" href="#show-samples-format-sampledetail-provide-a-flattened-view-of-samples" title="Permalink to this headline">¶</a></h2>
<p>If you&#8217;d like to be able to export variants plus associated sample metadata into
a downstream tool like R or pandas for additional exploration, adding the
<tt class="docutils literal"><span class="pre">--format</span> <span class="pre">sampledetail</span></tt> command flattens all found samples and attached
metadata information:</p>
<div class="highlight-bash"><div class="highlight"><pre>$ gemini query --header --format sampledetail --show-samples \
   -q &quot;select chrom, start, ref from variants where is_lof=1 limit 1&quot; test.query.db

chrom       start   ref     family_id       name    paternal_id     maternal_id     sex     phenotype
chr1        30547   T       0       1478PC0016      0       0       -9      -9
chr1        30547   T       0       1719PC0007      0       0       -9      -9
chr1        30547   T       0       1719PC0009      0       0       -9      -9&quot; &gt; exp
</pre></div>
</div>
<p>The denormalized results contain a row for each sample associated with a
variant, along with information from the sample table. This provides a way to
join sample information with a variant query.</p>
</div>
<div class="section" id="show-families-finding-out-which-families-have-a-variant">
<h2><tt class="docutils literal"><span class="pre">--show-families</span></tt> Finding out which families have a variant<a class="headerlink" href="#show-families-finding-out-which-families-have-a-variant" title="Permalink to this headline">¶</a></h2>
<p>This works exactly like <tt class="docutils literal"><span class="pre">--show-samples</span></tt> except lists all of the families with a
variant instead of the individual samples.</p>
</div>
<div class="section" id="region-restrict-a-query-to-a-specified-region">
<h2><tt class="docutils literal"><span class="pre">--region</span></tt> Restrict a query to a specified region<a class="headerlink" href="#region-restrict-a-query-to-a-specified-region" title="Permalink to this headline">¶</a></h2>
<p>If you are only interested in a specific region, you can restrict queries to
that region using the <tt class="docutils literal"><span class="pre">--region</span></tt> tool.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --region chr1:30859-30900 -q <span class="s2">&quot;select chrom, start, end, ref, alt \</span>
<span class="s2">             from variants&quot;</span>  test1.snpeff.db
chr1 30859   30860   G       C
</pre></div>
</div>
</div>
<div class="section" id="sample-filter-restrict-a-query-to-specified-samples">
<h2><tt class="docutils literal"><span class="pre">--sample-filter</span></tt> Restrict a query to specified samples<a class="headerlink" href="#sample-filter-restrict-a-query-to-specified-samples" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">--sample-filter</span></tt> option allows you to select samples that a variant
must be in by doing a SQL query on the samples table. For example if you
wanted to show the set of variants that appear in all samples with
a phenotype status of 2, you could do that query with:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --sample-filter <span class="s2">&quot;phenotype=2&quot;</span> -q <span class="s2">&quot;select gts, gt_types from variants&quot;</span> test.family.db
T/T,T/T,T/C,T/T,T/T,T/T,T/T,T/T,C/C  0,0,1,0,0,0,0,0,3       1_kid,3_kid     1_kid   3_kid
T/T,T/T,T/C,T/T,T/T,T/C,T/T,T/T,T/C  0,0,1,0,0,1,0,0,1       1_kid,2_kid,3_kid       1_kid,2_kid,3_kid
T/T,T/T,T/T,T/T,T/T,T/T,T/T,T/T,T/C  0,0,0,0,0,0,0,0,1       3_kid   3_kid
</pre></div>
</div>
<p>By default &#8211;sample-filter will show the variant if at least one sample contains the
variant. You can change this behavior by using the <tt class="docutils literal"><span class="pre">--in</span></tt> option along with
<tt class="docutils literal"><span class="pre">--sample-filter</span></tt>. <tt class="docutils literal"><span class="pre">--in</span> <span class="pre">all</span></tt> will return a variant if all samples matching
the query have the variant. <tt class="docutils literal"><span class="pre">in</span> <span class="pre">none</span></tt> will return a variant if the variant
does not appear in any of the matching samples. <tt class="docutils literal"><span class="pre">--in</span> <span class="pre">only</span></tt> will return a variant
if the variant is only in the matching samples and not in any of the non-matching
samples. <tt class="docutils literal"><span class="pre">--in</span> <span class="pre">only</span> <span class="pre">all</span></tt> will show all of the variant which are in all of the
matching samples and not in any of the non-matching samples.</p>
<p>The <tt class="docutils literal"><span class="pre">--family-wise</span></tt> flag applies the <tt class="docutils literal"><span class="pre">--sample-filter</span></tt> and <tt class="docutils literal"><span class="pre">--in</span></tt> behavior
on a family-wise basis. For example to show all variants that are only in samples
with a phenotype status of 2 in at least one family:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --family-wise --in only all --sample-filter <span class="s2">&quot;phenotype=2&quot;</span> -q <span class="s2">&quot;select gts, gt_types from variants&quot;</span> test.family.db
T/T,T/T,T/C,T/T,T/T,T/T,T/T,T/T,C/C  0,0,1,0,0,0,0,0,3       1_kid,3_kid     1_kid   3_kid
T/T,T/T,T/C,T/T,T/T,T/C,T/T,T/T,T/C  0,0,1,0,0,1,0,0,1       1_kid,2_kid,3_kid       1_kid,2_kid,3_kid
T/T,T/T,T/T,T/T,T/T,T/T,T/T,T/T,T/C  0,0,0,0,0,0,0,0,1       3_kid   3_kid
</pre></div>
</div>
<p>You can also specify that a variant passes this filter in multiple families with
the <tt class="docutils literal"><span class="pre">--min-kindreds</span></tt> option. So if you want to do the same query above, but restrict it
such that at least three families have to pass the filter:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --min-kindreds 3 --family-wise --in only all --sample-filter <span class="s2">&quot;phenotype=2&quot;</span> -q <span class="s2">&quot;select gts, gt_types from variants&quot;</span> test.family.db
T/T,T/T,T/C,T/T,T/T,T/C,T/T,T/T,T/C  0,0,1,0,0,1,0,0,1       1_kid,2_kid,3_kid       1_kid,2_kid,3_kid
</pre></div>
</div>
<p>If the PED file you loaded has extra fields in it, those will also work with the
<tt class="docutils literal"><span class="pre">--sample-filter</span></tt> option. For example if you had a <tt class="docutils literal"><span class="pre">hair_color</span></tt> extended field,
you could query on that as well as phenotype:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query  --in only all --sample-filter <span class="s2">&quot;phenotype=1 and hair_color=&#39;blue&#39;&quot;</span> -q <span class="s2">&quot;select gts, gt_types from variants&quot;</span> extended_ped.db
G/G,G/G,G/G,G/A      0,0,0,1 M128215 M128215
</pre></div>
</div>
</div>
<div class="section" id="sample-delim-changing-the-sample-list-delimiter">
<h2><tt class="docutils literal"><span class="pre">--sample-delim</span></tt> Changing the sample list delimiter<a class="headerlink" href="#sample-delim-changing-the-sample-list-delimiter" title="Permalink to this headline">¶</a></h2>
<p>One can modify the default comma delimiter used by the <tt class="docutils literal"><span class="pre">--show-samples</span></tt>
option through the use of the <tt class="docutils literal"><span class="pre">--sample-delim</span></tt> option.  For example, to use
a semi-colon instead of a comma, one would do the following:</p>
<div class="highlight-bash"><div class="highlight"><pre>  <span class="nv">$ </span>gemini query --header --show-samples --sample-delim <span class="s2">&quot;;&quot;</span> <span class="se">\</span>
                 -q <span class="s2">&quot;select chrom, start, end, ref, alt \</span>
<span class="s2">                     from variants where is_lof=1 limit 5&quot;</span> test.query.db

chrom start end ref alt variant_samples HET_samples HOM_ALT_samples
chr1  874815  874816  C CT  1478PC0006B;1478PC0007B;1478PC0010,1478PC0013B;1478PC0022B;1478PC0023B;1478PC0025;1719PC0007;1719PC0009;1719PC0010;1719PC0022 1478PC0006B;1478PC0007B;1478PC0010;1478PC0013B;1478PC0022B;1478PC0023B;1719PC0007;1719PC0009;1719PC0010 1478PC0025;1719PC0022
chr1  1140811 1140813 TC  T 1478PC0011  1478PC0011
chr1  1219381 1219382 C G 1719PC0012  1719PC0012
chr1  1221487 1221490 CAA C 1478PC0004  1478PC0004
</pre></div>
</div>
</div>
<div class="section" id="format-reporting-query-output-in-an-alternate-format">
<h2><tt class="docutils literal"><span class="pre">--format</span></tt> Reporting query output in an alternate format.<a class="headerlink" href="#format-reporting-query-output-in-an-alternate-format" title="Permalink to this headline">¶</a></h2>
<p>The results of GEMINI queries can automatically be formatted for use with
other programs using the &#8211;format command. Supported alternative
formats are JSON and TPED (Transposed PED) format.</p>
<p>Reporting query output in JSON format may enable
HTML/Javascript apps to query GEMINI and retrieve
the output in a format that is amenable to web development protocols.</p>
<p>Here is a basic query:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select chrom, start, end from variants&quot;</span> my.db | head
chr1  10067 10069
chr1  10230 10231
chr1  12782 12783
chr1  13109 13110
chr1  13115 13116
chr1  13117 13118
chr1  13272 13273
chr1  13301 13302
chr1  13416 13417
chr1  13417 13418
</pre></div>
</div>
<p>To report in JSON format, use the <tt class="docutils literal"><span class="pre">--format</span> <span class="pre">json</span></tt> option. For example:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --format json -q <span class="s2">&quot;select chrom, start, end from variants&quot;</span> my.db | head
<span class="o">{</span><span class="s2">&quot;chrom&quot;</span>: <span class="s2">&quot;chr1&quot;</span>, <span class="s2">&quot;start&quot;</span>: 10067, <span class="s2">&quot;end&quot;</span>: 10069<span class="o">}</span>
<span class="o">{</span><span class="s2">&quot;chrom&quot;</span>: <span class="s2">&quot;chr1&quot;</span>, <span class="s2">&quot;start&quot;</span>: 10230, <span class="s2">&quot;end&quot;</span>: 10231<span class="o">}</span>
<span class="o">{</span><span class="s2">&quot;chrom&quot;</span>: <span class="s2">&quot;chr1&quot;</span>, <span class="s2">&quot;start&quot;</span>: 12782, <span class="s2">&quot;end&quot;</span>: 12783<span class="o">}</span>
<span class="o">{</span><span class="s2">&quot;chrom&quot;</span>: <span class="s2">&quot;chr1&quot;</span>, <span class="s2">&quot;start&quot;</span>: 13109, <span class="s2">&quot;end&quot;</span>: 13110<span class="o">}</span>
<span class="o">{</span><span class="s2">&quot;chrom&quot;</span>: <span class="s2">&quot;chr1&quot;</span>, <span class="s2">&quot;start&quot;</span>: 13115, <span class="s2">&quot;end&quot;</span>: 13116<span class="o">}</span>
<span class="o">{</span><span class="s2">&quot;chrom&quot;</span>: <span class="s2">&quot;chr1&quot;</span>, <span class="s2">&quot;start&quot;</span>: 13117, <span class="s2">&quot;end&quot;</span>: 13118<span class="o">}</span>
<span class="o">{</span><span class="s2">&quot;chrom&quot;</span>: <span class="s2">&quot;chr1&quot;</span>, <span class="s2">&quot;start&quot;</span>: 13272, <span class="s2">&quot;end&quot;</span>: 13273<span class="o">}</span>
<span class="o">{</span><span class="s2">&quot;chrom&quot;</span>: <span class="s2">&quot;chr1&quot;</span>, <span class="s2">&quot;start&quot;</span>: 13301, <span class="s2">&quot;end&quot;</span>: 13302<span class="o">}</span>
<span class="o">{</span><span class="s2">&quot;chrom&quot;</span>: <span class="s2">&quot;chr1&quot;</span>, <span class="s2">&quot;start&quot;</span>: 13416, <span class="s2">&quot;end&quot;</span>: 13417<span class="o">}</span>
<span class="o">{</span><span class="s2">&quot;chrom&quot;</span>: <span class="s2">&quot;chr1&quot;</span>, <span class="s2">&quot;start&quot;</span>: 13417, <span class="s2">&quot;end&quot;</span>: 13418<span class="o">}</span>
</pre></div>
</div>
<p>If you would to use tools such as PLINK that use
the PED format, you can dump out a set of variants
matching any query in TPED (Transposed PED) format
by adding the <tt class="docutils literal"><span class="pre">--tped</span></tt> flag to your query:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --format tped -q <span class="s2">&quot;select * from variants where chrom=10&quot;</span> test4.snpeff.db
10 rs10794716 0 1142207 C/C C/C C/C C/C
10 rs142685947 0 48003991 T/T C/T C/T C/C
10 rs2842123 0 52004314 ./. ./. C/C C/C
10 rs4935178 0 52497528 ./. C/C C/C ./.
16 rs201947120 0 72057434 C/T C/C C/C C/C
10 rs73373169 0 126678091 G/G G/G G/G G/A
10 rs2265637 0 135210790 T/T C/C C/C T/T
10 rs6537611 0 135336655 ./. A/A ./. A/A
10 rs3747881 0 135369531 T/T T/C T/C T/T
</pre></div>
</div>
<p>You can pass &#8211;header to get a header to see which samples have which
variant. To use the TPED format you also need to generate a corresponing TFAM
file from your data as well, which you can get from the GEMINI dump tool:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini dump  --tfam test4.snpeff.db &gt; obs
None    M10475  None    None    None    None
None    M10478  None    None    None    None
None    M10500  None    None    None    None
None    M128215 None    None    None    None
</pre></div>
</div>
</div>
<div class="section" id="carrier-summary-by-phenotype-summarize-carrier-status">
<h2><tt class="docutils literal"><span class="pre">--carrier-summary-by-phenotype</span></tt> Summarize carrier status<a class="headerlink" href="#carrier-summary-by-phenotype-summarize-carrier-status" title="Permalink to this headline">¶</a></h2>
<p>For prioritizing variants sometimes it is useful to have summary counts of
the carrier status for all samples with a variant stratified across a phenotype.
<tt class="docutils literal"><span class="pre">--carrier-summary-by-phenotype</span></tt> takes a column in the samples table that you
want to summarize the carrier status of and adds a set of counts of
carrier/non-carrier status for each phenotype in the given column. For example,
to get a summary of how a set of variants segregate with affected status:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --show-samples --carrier-summary-by-phenotype affected --header -q <span class="s2">&quot;select chrom, start, ref, alt, gt_types from variants&quot;</span> extended_ped_test.db
chrom   start   ref     alt     gt_types        variant_samples HET_samples     HOM_ALT_samples unaffected_carrier      affected_carrier        unaffected_noncarrier   affected_noncarrier     unknown
chr10   1142207 T       C       3,3,3,3 M10475,M10478,M10500,M128215            M10475,M10478,M10500,M128215    2       2       0       0       0
chr10   48003991        C       T       3,1,1,0 M10475,M10478,M10500    M10478,M10500   M10475  1       2       1       0       0
chr10   52004314        T       C       2,2,3,3 M10500,M128215          M10500,M128215  1       1       0       0       2
chr10   52497528        G       C       2,3,3,2 M10478,M10500           M10478,M10500   0       2       0       0       2
chr16   72057434        C       T       1,0,0,0 M10475  M10475          1       0       1       2       0
chr10   126678091       G       A       0,0,0,1 M128215 M128215         1       0       1       2       0
chr10   135210790       T       C       0,3,3,0 M10478,M10500           M10478,M10500   0       2       2       0       0
chr10   135336655       G       A       2,3,2,3 M10478,M128215          M10478,M128215  1       1       0       0       2
chr10   135369531       T       C       0,1,1,0 M10478,M10500   M10478,M10500           0       2       2       0       0
</pre></div>
</div>
<p>Or if you have another phenotypic feature you are interested in summarizing,
like hair color:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --show-samples --carrier-summary-by-phenotype hair_color --header -q <span class="s2">&quot;select chrom, start, ref, alt, gt_types from variants&quot;</span> extended_ped.db
chrom   start   ref     alt     gt_types        variant_samples HET_samples     HOM_ALT_samples blue_carrier    brown_carrier   purple_carrier  blue_noncarrier brown_noncarrier        purple_noncarrier       unknown
chr10   1142207 T       C       3,3,3,3 M10475,M10478,M10500,M128215            M10475,M10478,M10500,M128215    1       2       1       0       0       0       0
chr10   48003991        C       T       3,1,1,0 M10475,M10478,M10500    M10478,M10500   M10475  0       2       1       1       0       0       0
chr10   52004314        T       C       2,2,3,3 M10500,M128215          M10500,M128215  1       0       1       0       0       0       2
chr10   52497528        G       C       2,3,3,2 M10478,M10500           M10478,M10500   0       1       1       0       0       0       2
chr16   72057434        C       T       1,0,0,0 M10475  M10475          0       1       0       1       1       1       0
chr10   126678091       G       A       0,0,0,1 M128215 M128215         1       0       0       0       2       1       0
chr10   135210790       T       C       0,3,3,0 M10478,M10500           M10478,M10500   0       1       1       1       1       0       0
chr10   135336655       G       A       2,3,2,3 M10478,M128215          M10478,M128215  1       1       0       0       0       0       2
chr10   135369531       T       C       0,1,1,0 M10478,M10500   M10478,M10500           0       1       1       1       1       0       0
</pre></div>
</div>
</div>
</div>
<div class="section" id="querying-the-gene-tables">
<h1>Querying the gene tables<a class="headerlink" href="#querying-the-gene-tables" title="Permalink to this headline">¶</a></h1>
<p>The gene tables viz. <tt class="docutils literal"><span class="pre">gene_detailed</span> <span class="pre">table</span></tt> and the <tt class="docutils literal"><span class="pre">gene_summary</span> <span class="pre">table</span></tt> have been built on version 73 of the ensembl genes. The column specifications are
available at <a class="reference internal" href="database_schema.html"><em>The Gemini database schema</em></a>. These tables contain gene specific information e.g. gene synonyms, RVIS percentile scores(Petrovski et.al 2013), strand specifications, cancer gene census etc. While the former is more detailed, the later lacks transcript wise information and summarizes some aspects of the former. For e.g. while the gene_detailed table lists all transcripts of a gene with their start and end co-ordinates, the gene_summary table reports only the minimum start and maximum end co-ordinates of the gene transcripts. The <tt class="docutils literal"><span class="pre">chrom</span></tt>, <tt class="docutils literal"><span class="pre">gene</span></tt> and the <tt class="docutils literal"><span class="pre">transcript</span></tt> columns of the gene tables may be used to join on the variants and the variant_impacts tables.</p>
<div class="section" id="using-the-gene-detailed-gene-summary-tables">
<h2>Using the <tt class="docutils literal"><span class="pre">gene_detailed</span></tt> &amp; <tt class="docutils literal"><span class="pre">gene_summary</span></tt> tables<a class="headerlink" href="#using-the-gene-detailed-gene-summary-tables" title="Permalink to this headline">¶</a></h2>
<div class="section" id="query-the-gene-detailed-table-with-a-join-on-variants-table">
<h3>Query the gene_detailed table with a join on variants table:<a class="headerlink" href="#query-the-gene-detailed-table-with-a-join-on-variants-table" title="Permalink to this headline">¶</a></h3>
<p>E.g. Get additional transcript info for the most severe impact transcript
e.g. transcript status, transcript start,end and the protein length</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header -q <span class="s2">&quot;select v.variant_id, v.gene, \</span>
<span class="s2">                   v.impact, g.transcript_status, g.transcript, \</span>
<span class="s2">                               g.transcript_start, g.transcript_end, g.protein_length, \</span>
<span class="s2">                               from variants v, gene_detailed g \</span>

<span class="s2">                               WHERE v.chrom = g.chrom AND \</span>
<span class="s2">                                             v.gene = g.gene AND \</span>
<span class="s2">                                             v.transcript = g.transcript AND \</span>
<span class="s2">                                             v.impact_severity=&#39;HIGH&#39;&quot;</span> test.query.db

    variant_id      gene    impact  transcript_status       transcript      transcript_start        transcript_end  protein_length
    46      SAMD11  frame_shift     KNOWN   ENST00000342066 861118  879955  681
    578     TNFRSF18        frame_shift     PUTATIVE        ENST00000486728 1139224 1141060 169
    733     SCNN1D  stop_gain       NOVEL   ENST00000470022 1217305 1221548 138
</pre></div>
</div>
</div>
<div class="section" id="query-the-gene-detailed-table-with-a-join-on-the-variant-impacts-table">
<h3>Query the gene_detailed table with a join on the variant_impacts table:<a class="headerlink" href="#query-the-gene-detailed-table-with-a-join-on-the-variant-impacts-table" title="Permalink to this headline">¶</a></h3>
<p>E.g. Get the transcript status for all transcripts of the SCNN1D gene where
impact severity is not &#8216;LOW&#8217;.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header -q <span class="s2">&quot;select v.gene, g.transcript_status, g.transcript, \</span>
<span class="s2">               v.impact from variant_impacts v, gene_detailed g \</span>

<span class="s2">                               WHERE v.transcript = g.transcript AND \</span>
<span class="s2">                     v.gene = g.gene AND \</span>
<span class="s2">                                             v.gene = &#39;SCNN1D&#39; \</span>
<span class="s2">                         v.impact_severity!=&#39;LOW&#39;&quot;</span> test.query.db

    gene    transcript_status       transcript      impact
    SCNN1D  NOVEL   ENST00000470022 non_syn_coding
    SCNN1D  NOVEL   ENST00000470022 frame_shift
    SCNN1D  KNOWN   ENST00000325425 frame_shift
    SCNN1D  KNOWN   ENST00000379116 non_syn_coding
    SCNN1D  KNOWN   ENST00000338555 non_syn_coding
    SCNN1D  KNOWN   ENST00000400928 non_syn_coding
</pre></div>
</div>
</div>
<div class="section" id="query-the-gene-summary-table-with-a-join-on-the-variants-table">
<h3>Query the gene_summary table with a join on the variants table:<a class="headerlink" href="#query-the-gene-summary-table-with-a-join-on-the-variants-table" title="Permalink to this headline">¶</a></h3>
<p>E.g. Get the synonym/alternate names, RVIS percentile scores and the min-max
start-end of transcripts for genes that have a severely affected transcript
of a &#8216;HIGH&#8217; order.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header -q <span class="s2">&quot;select v.chrom, v.gene, g.transcript_min_start, \</span>
<span class="s2">                   g.transcript_max_end, g.synonym, g.rvis_pct, v.impact from \</span>
<span class="s2">                               variants v, gene_summary g \</span>

<span class="s2">                               WHERE v.chrom = g.chrom AND \</span>
<span class="s2">                     v.gene = g.gene AND \</span>
<span class="s2">                     v.impact_severity=&#39;HIGH&#39;&quot;</span> test.query.db

    chrom   gene    transcript_min_start    transcript_max_end      synonym rvis_pct        impact
    chr1    SAMD11  860260  879955  MGC45873        None    frame_shift
    chr1    TNFRSF18        1138888 1142071 AITR,CD357,GITR None    frame_shift
    chr1    SCNN1D  1215816 1227409 ENaCdelta,dNaCh 96.77990092     stop_gain
</pre></div>
</div>
</div>
<div class="section" id="query-the-gene-summary-table-with-a-join-on-the-variant-impacts-table">
<h3>Query the gene_summary table with a join on the variant_impacts table:<a class="headerlink" href="#query-the-gene-summary-table-with-a-join-on-the-variant-impacts-table" title="Permalink to this headline">¶</a></h3>
<p>E.g. Get all variants of a gene, the affected transcripts and impacts, where
a mammalian phenotype ID is available for the <tt class="docutils literal"><span class="pre">mouse</span> <span class="pre">phenotype</span></tt>.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header -q <span class="s2">&quot;select v.variant_id, v.chrom, v.gene, i.impact, \</span>
<span class="s2">                   i.transcript, g.mam_phenotype_id from variants v, \</span>
<span class="s2">                               variant_impacts i, gene_summary g \</span>

<span class="s2">                               WHERE v.variant_id=i.variant_id \</span>
<span class="s2">                               i.gene=g.gene AND \</span>
<span class="s2">                               v.chrom=g.chrom AND \</span>
<span class="s2">                               g.mam_phenotype_id !=&#39;None&#39;&quot;</span> test.query.db

    variant_id      chrom   gene    impact  transcript      mam_phenotype_id
    334     chr1    TNFRSF18        non_syn_coding  ENST00000328596 MP:0005397,MP:0005384,MP:0005387
    378     chr1    TNFRSF18        frame_shift     ENST00000486728 MP:0005397,MP:0005384,MP:0005387
    483     chr1    AGRN    synonymous_coding       ENST00000379370 MP:0005378,MP:0005386,MP:0005388,MP:0005367,MP:0005369,MP:0005371,MP:0003631,MP:0002873,MP:0010768
    484     chr1    AGRN    exon    ENST00000461111 MP:0005378,MP:0005386,MP:0005388,MP:0005367,MP:0005369,MP:0005371,MP:0003631,MP:0002873,MP:0010768
    478     chr1    AGRN    intron  ENST00000461111 MP:0005378,MP:0005386,MP:0005388,MP:0005367,MP:0005369,MP:0005371,MP:0003631,MP:0002873,MP:0010768
    479     chr1    AGRN    downstream      ENST00000492947 MP:0005378,MP:0005386,MP:0005388,MP:0005367,MP:0005369,MP:0005371,MP:0003631,MP:0002873,MP:0010768
</pre></div>
</div>
</div>
</div>
<div class="section" id="restrict-analysis-to-transcripts-with-a-valid-ccds-id">
<h2>Restrict analysis to transcripts with a valid <tt class="docutils literal"><span class="pre">CCDS_ID</span></tt><a class="headerlink" href="#restrict-analysis-to-transcripts-with-a-valid-ccds-id" title="Permalink to this headline">¶</a></h2>
<p>Since the current available transcript sets are more than one (e.g. RefSeq, ENSEMBL and UCSC)
we support information (e.g pathways tool) for the ENSEMBL transcripts but provide a mapping of
these transcripts to the consensus set agreed upon by all the above three mentioned groups viz.
<tt class="docutils literal"><span class="pre">transcripts</span> <span class="pre">having</span> <span class="pre">a</span> <span class="pre">valid</span> <span class="pre">CCDS_ID</span></tt>. Here we show, how we can return variants and their
impacts for only these restricted set of transcripts using the gene_detailed table.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query --header -q <span class="s2">&quot;select i.var_id, i.gene, i.impact, i.transcript, \</span>
<span class="s2">                g.transcript_status, ccds_id, g.rvis_pct from \</span>
<span class="s2">                                variant_impacts i, gene_detailed g where \</span>
<span class="s2">                                i.transcript=g.transcript and i.gene=g.gene and\</span>
<span class="s2">                                impact_severity=&#39;HIGH&#39; and g.ccds_id!=&#39;None&#39;&quot;</span> test.query.db

variant_id      gene    impact  transcript      transcript_status       ccds_id rvis_pct
2051    SAMD11  frame_shift     ENST00000342066 KNOWN   CCDS2   None
3639    CCNL2   splice_acceptor ENST00000408918 KNOWN   CCDS30558       53.98089172
3639    CCNL2   splice_acceptor ENST00000400809 KNOWN   CCDS30557       53.98089172
13221   SMIM1   frame_shift     ENST00000444870 NOVEL   CCDS57966       None
21881   NPHP4   splice_acceptor ENST00000378156 KNOWN   CCDS44052       81.78815758
</pre></div>
</div>
</div>
<div class="section" id="what-if-i-don-t-see-my-gene-in-the-database">
<h2>What if I don&#8217;t see my <tt class="docutils literal"><span class="pre">gene</span></tt> in the database?<a class="headerlink" href="#what-if-i-don-t-see-my-gene-in-the-database" title="Permalink to this headline">¶</a></h2>
<p>Most genes are popular by their common names while the representation of gene names
in the GEMINI database is mostly HGNC. For e.g <tt class="docutils literal"><span class="pre">ARTEMIS</span></tt> would be <tt class="docutils literal"><span class="pre">DCLRE1C</span></tt> in
the GEMINI database. As such one may miss out on variants if looking for specific
genes by their common names. While, joining the main tables with the gene tables for
synonym information would be useful (as shown in the previous examples), the gene
tables may also serve as a quick look up for alternate names of a gene, which
could then be looked up in the database.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select synonym from gene_summary where \</span>
<span class="s2">                   gene=&#39;ARTEMIS&#39;&quot;</span> test.query.db
A-SCID,FLJ11360,SNM1C,DCLRE1C,SCIDA

or,

<span class="nv">$ </span>gemini query -q <span class="s2">&quot;select gene from gene_summary where synonym \</span>
<span class="s2">                   like &#39;%ARTEMIS%&#39; and \</span>
<span class="s2">                   is_HGNC=&#39;1&#39;&quot;</span> test.query.db
DCLRE1C

<span class="c">#looking up for DCLRE1C in the database</span>
<span class="nv">$ </span>gemini query -q <span class="s2">&quot;select variant_id, chrom, start, end, impact \</span>
<span class="s2">                   from variants where \</span>
<span class="s2">                   gene=&#39;DCLRE1C&#39;&quot;</span> test.query.db
</pre></div>
</div>
</div>
</div>




  

<div id="disqus_thread"></div>
 <script type="text/javascript">
     /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
     var disqus_shortname = 'geminidocs'; // required: replace example with your forum shortname

     /* * * DON'T EDIT BELOW THIS LINE * * */
     (function() {
         var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
         dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
         (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
     })();
 </script>
 <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
 <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>






          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/gemini.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Querying the GEMINI database</a><ul>
<li><a class="reference internal" href="#basic-queries">Basic queries</a></li>
<li><a class="reference internal" href="#selecting-sample-genotypes">Selecting sample genotypes</a></li>
<li><a class="reference internal" href="#selecting-sample-genotypes-based-on-wildcards">Selecting sample genotypes based on &#8220;wildcards&#8221;.</a><ul>
<li><a class="reference internal" href="#the-wildcard">The &#8220;*&#8221; wildcard</a></li>
<li><a class="reference internal" href="#wildcards-based-on-sample-attributes">Wildcards based on sample attributes</a></li>
</ul>
</li>
<li><a class="reference internal" href="#gt-filter-filtering-on-genotypes"><tt class="docutils literal"><span class="pre">--gt-filter</span></tt> Filtering on genotypes</a></li>
<li><a class="reference internal" href="#gt-filter-wildcard-filtering-on-genotype-columns"><tt class="docutils literal"><span class="pre">--gt-filter</span></tt> Wildcard filtering on genotype columns.</a><ul>
<li><a class="reference internal" href="#the-basics">The basics.</a></li>
<li><a class="reference internal" href="#the-all-operator">The <tt class="docutils literal"><span class="pre">all</span></tt> operator</a></li>
<li><a class="reference internal" href="#the-any-operator">The <tt class="docutils literal"><span class="pre">any</span></tt> operator</a></li>
<li><a class="reference internal" href="#the-none-operator">The <tt class="docutils literal"><span class="pre">none</span></tt> operator</a></li>
<li><a class="reference internal" href="#the-count-operator">The <tt class="docutils literal"><span class="pre">count</span></tt> operator</a></li>
<li><a class="reference internal" href="#example-scenario">Example scenario.</a></li>
<li><a class="reference internal" href="#more-complexity">More complexity.</a></li>
<li><a class="reference internal" href="#other-details">Other details.</a></li>
</ul>
</li>
<li><a class="reference internal" href="#show-samples-finding-out-which-samples-have-a-variant"><tt class="docutils literal"><span class="pre">--show-samples</span></tt> Finding out which samples have a variant</a></li>
<li><a class="reference internal" href="#show-samples-format-sampledetail-provide-a-flattened-view-of-samples"><tt class="docutils literal"><span class="pre">--show-samples</span> <span class="pre">--format</span> <span class="pre">sampledetail</span></tt> Provide a flattened view of samples</a></li>
<li><a class="reference internal" href="#show-families-finding-out-which-families-have-a-variant"><tt class="docutils literal"><span class="pre">--show-families</span></tt> Finding out which families have a variant</a></li>
<li><a class="reference internal" href="#region-restrict-a-query-to-a-specified-region"><tt class="docutils literal"><span class="pre">--region</span></tt> Restrict a query to a specified region</a></li>
<li><a class="reference internal" href="#sample-filter-restrict-a-query-to-specified-samples"><tt class="docutils literal"><span class="pre">--sample-filter</span></tt> Restrict a query to specified samples</a></li>
<li><a class="reference internal" href="#sample-delim-changing-the-sample-list-delimiter"><tt class="docutils literal"><span class="pre">--sample-delim</span></tt> Changing the sample list delimiter</a></li>
<li><a class="reference internal" href="#format-reporting-query-output-in-an-alternate-format"><tt class="docutils literal"><span class="pre">--format</span></tt> Reporting query output in an alternate format.</a></li>
<li><a class="reference internal" href="#carrier-summary-by-phenotype-summarize-carrier-status"><tt class="docutils literal"><span class="pre">--carrier-summary-by-phenotype</span></tt> Summarize carrier status</a></li>
</ul>
</li>
<li><a class="reference internal" href="#querying-the-gene-tables">Querying the gene tables</a><ul>
<li><a class="reference internal" href="#using-the-gene-detailed-gene-summary-tables">Using the <tt class="docutils literal"><span class="pre">gene_detailed</span></tt> &amp; <tt class="docutils literal"><span class="pre">gene_summary</span></tt> tables</a><ul>
<li><a class="reference internal" href="#query-the-gene-detailed-table-with-a-join-on-variants-table">Query the gene_detailed table with a join on variants table:</a></li>
<li><a class="reference internal" href="#query-the-gene-detailed-table-with-a-join-on-the-variant-impacts-table">Query the gene_detailed table with a join on the variant_impacts table:</a></li>
<li><a class="reference internal" href="#query-the-gene-summary-table-with-a-join-on-the-variants-table">Query the gene_summary table with a join on the variants table:</a></li>
<li><a class="reference internal" href="#query-the-gene-summary-table-with-a-join-on-the-variant-impacts-table">Query the gene_summary table with a join on the variant_impacts table:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#restrict-analysis-to-transcripts-with-a-valid-ccds-id">Restrict analysis to transcripts with a valid <tt class="docutils literal"><span class="pre">CCDS_ID</span></tt></a></li>
<li><a class="reference internal" href="#what-if-i-don-t-see-my-gene-in-the-database">What if I don&#8217;t see my <tt class="docutils literal"><span class="pre">gene</span></tt> in the database?</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="preprocessing.html"
                        title="previous chapter">Preprocessing and Loading a VCF file into GEMINI</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="tools.html"
                        title="next chapter">Built-in analysis tools</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/querying.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="tools.html" title="Built-in analysis tools"
             >next</a> |</li>
        <li class="right" >
          <a href="preprocessing.html" title="Preprocessing and Loading a VCF file into GEMINI"
             >previous</a> |</li>
        <li><a href="../index.html">gemini v0.12.1</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2012,2013.
      Last updated on Mar 19, 2015.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>

<!-- 
<div id="editor-trap">
      <h3>Edit and improve this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/arq5x/gemini/blob/master/docs/content/querying.rst">
          Querying the GEMINI database
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/arq5x/gemini/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>
 -->
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', UA-39150563-1]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>


  </body>
</html>