<html>

<head>
<meta http-equiv=Content-Type content="text/html; charset=windows-1252">
<meta name=Generator content="Microsoft Word 15 (filtered)">
<style>
<!--
 /* Font Definitions */
 @font-face
	{font-family:Wingdings;
	panose-1:5 0 0 0 0 0 0 0 0 0;}
@font-face
	{font-family:"Cambria Math";
	panose-1:2 4 5 3 5 4 6 3 2 4;}
@font-face
	{font-family:"Calibri Light";
	panose-1:2 15 3 2 2 2 4 3 2 4;}
@font-face
	{font-family:Calibri;
	panose-1:2 15 5 2 2 2 4 3 2 4;}
 /* Style Definitions */
 p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:0cm;
	line-height:107%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
h1
	{mso-style-link:"Heading 1 Char";
	margin-top:12.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	line-height:107%;
	page-break-after:avoid;
	font-size:16.0pt;
	font-family:"Calibri Light",sans-serif;
	color:#2E74B5;
	font-weight:normal;}
h2
	{mso-style-link:"Heading 2 Char";
	margin-top:2.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	line-height:107%;
	page-break-after:avoid;
	font-size:13.0pt;
	font-family:"Calibri Light",sans-serif;
	color:#2E74B5;
	font-weight:normal;}
h3
	{mso-style-link:"Heading 3 Char";
	margin-top:2.0pt;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:0cm;
	margin-bottom:.0001pt;
	line-height:107%;
	page-break-after:avoid;
	font-size:12.0pt;
	font-family:"Calibri Light",sans-serif;
	color:#1F4D78;
	font-weight:normal;}
p.MsoHeader, li.MsoHeader, div.MsoHeader
	{mso-style-link:"Header Char";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoFooter, li.MsoFooter, div.MsoFooter
	{mso-style-link:"Footer Char";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoTitle, li.MsoTitle, div.MsoTitle
	{mso-style-link:"Title Char";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:28.0pt;
	font-family:"Calibri Light",sans-serif;
	letter-spacing:-.5pt;}
p.MsoTitleCxSpFirst, li.MsoTitleCxSpFirst, div.MsoTitleCxSpFirst
	{mso-style-link:"Title Char";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:28.0pt;
	font-family:"Calibri Light",sans-serif;
	letter-spacing:-.5pt;}
p.MsoTitleCxSpMiddle, li.MsoTitleCxSpMiddle, div.MsoTitleCxSpMiddle
	{mso-style-link:"Title Char";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:28.0pt;
	font-family:"Calibri Light",sans-serif;
	letter-spacing:-.5pt;}
p.MsoTitleCxSpLast, li.MsoTitleCxSpLast, div.MsoTitleCxSpLast
	{mso-style-link:"Title Char";
	margin:0cm;
	margin-bottom:.0001pt;
	font-size:28.0pt;
	font-family:"Calibri Light",sans-serif;
	letter-spacing:-.5pt;}
p.MsoSubtitle, li.MsoSubtitle, div.MsoSubtitle
	{mso-style-link:"Subtitle Char";
	margin-top:0cm;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:0cm;
	line-height:107%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;
	color:#5A5A5A;
	letter-spacing:.75pt;}
a:link, span.MsoHyperlink
	{color:#0563C1;
	text-decoration:underline;}
a:visited, span.MsoHyperlinkFollowed
	{color:#954F72;
	text-decoration:underline;}
span.MsoPlaceholderText
	{color:gray;}
p.MsoListParagraph, li.MsoListParagraph, div.MsoListParagraph
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:36.0pt;
	line-height:107%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoListParagraphCxSpFirst, li.MsoListParagraphCxSpFirst, div.MsoListParagraphCxSpFirst
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:36.0pt;
	margin-bottom:.0001pt;
	line-height:107%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoListParagraphCxSpMiddle, li.MsoListParagraphCxSpMiddle, div.MsoListParagraphCxSpMiddle
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:0cm;
	margin-left:36.0pt;
	margin-bottom:.0001pt;
	line-height:107%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
p.MsoListParagraphCxSpLast, li.MsoListParagraphCxSpLast, div.MsoListParagraphCxSpLast
	{margin-top:0cm;
	margin-right:0cm;
	margin-bottom:8.0pt;
	margin-left:36.0pt;
	line-height:107%;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
span.Heading2Char
	{mso-style-name:"Heading 2 Char";
	mso-style-link:"Heading 2";
	font-family:"Calibri Light",sans-serif;
	color:#2E74B5;}
span.Heading3Char
	{mso-style-name:"Heading 3 Char";
	mso-style-link:"Heading 3";
	font-family:"Calibri Light",sans-serif;
	color:#1F4D78;}
span.TitleChar
	{mso-style-name:"Title Char";
	mso-style-link:Title;
	font-family:"Calibri Light",sans-serif;
	letter-spacing:-.5pt;}
span.SubtitleChar
	{mso-style-name:"Subtitle Char";
	mso-style-link:Subtitle;
	font-family:"Times New Roman",serif;
	color:#5A5A5A;
	letter-spacing:.75pt;}
span.HeaderChar
	{mso-style-name:"Header Char";
	mso-style-link:Header;}
span.FooterChar
	{mso-style-name:"Footer Char";
	mso-style-link:Footer;}
span.Heading1Char
	{mso-style-name:"Heading 1 Char";
	mso-style-link:"Heading 1";
	font-family:"Calibri Light",sans-serif;
	color:#2E74B5;}
.MsoChpDefault
	{font-family:"Calibri",sans-serif;}
.MsoPapDefault
	{margin-bottom:8.0pt;
	line-height:107%;}
 /* Page Definitions */
 @page WordSection1
	{size:595.3pt 841.9pt;
	margin:70.85pt 70.85pt 2.0cm 70.85pt;}
div.WordSection1
	{page:WordSection1;}
 /* List Definitions */
 ol
	{margin-bottom:0cm;}
ul
	{margin-bottom:0cm;}
-->
</style>

</head>

<body lang=EN-US link="#0563C1" vlink="#954F72">

<div class=WordSection1>

<p class=MsoTitle><span lang=FR-CH>DynaMet DOCUMENTATION</span></p>

<p class=MsoNormal><span lang=FR-CH>&nbsp;</span></p>

<h2 style='margin-left:18.0pt;text-indent:-18.0pt'><span lang=DE-CH
style='font-size:18.0pt;line-height:107%'>0.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;
</span></span><span lang=DE-CH style='font-size:18.0pt;line-height:107%'>General
</span></h2>

<p class=MsoNormal style='margin-left:18.0pt'>DynaMet provides a fully
automated workflow for liquid chromatography mass spectrometry (LC-MS) raw data
analyses allowing for metabolome-wide investigations of dynamic isotope
labeling experiments. DynaMet enables untargeted extraction of labeling
profiles by grouping metabolite features in different samples with isotopic
patterns changing over time. Moreover, integrated tools for expressive data
visualization enhance result inspection. DynaMet was developed for Python based
LC-MS data analysis framework eMZed 2. </p>

<p class=MsoNormal>&nbsp;</p>

<h2 style='margin-left:18.0pt;text-indent:-18.0pt'><span style='font-size:18.0pt;
line-height:107%'>1.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp; </span></span><span
style='font-size:18.0pt;line-height:107%'>Installation</span></h2>

<p class=MsoNormal style='margin-left:18.0pt'>DynaMet requires eMZed 2.3.1 or
higher. If you have not yet installed eMZed you can find latest emzed2 version
including installation instructions <span lang=DE-CH><a
href="http://emzed.ethz.ch/installation.html"><span lang=EN-US>here</span></a></span>.
</p>

<p class=MsoNormal style='margin-left:18.0pt'>After eMZed installation you have
to install two additional packages ‘hires’ and ‘pacer’. To install the packages
start eMZed and type into IPython console:</p>

<p class=MsoNormal style='margin-left:18.0pt'><img border=0 width=210
height=14 id="Picture 21"
src="20141030DYNAMET%20DOCUMENTATION_files/image001.png"> press enter</p>

<p class=MsoNormal style='margin-left:18.0pt'><img border=0 width=216
height=19 id="Picture 38"
src="20141030DYNAMET%20DOCUMENTATION_files/image002.png"> press enter</p>

<p class=MsoNormal style='margin-left:18.0pt'>Dynamet is distributed as eMZed
extension called package or apps. To install the package type
emzed.project.install_wheel() into eMZed IPython console.</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=22 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=605 height=415
  src="20141030DYNAMET%20DOCUMENTATION_files/image003.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>After installation open a new IPython console and DynaMet
can be started by entering the command emzed.app.dynamet().</p>

<span style='font-size:11.0pt;line-height:107%;font-family:"Calibri",sans-serif'><br
clear=all style='page-break-before:always'>
</span>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>&nbsp;</p>

<h2 style='margin-left:18.0pt;text-indent:-18.0pt'><span style='font-size:18.0pt;
line-height:107%'>2.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp; </span></span><span
style='font-size:18.0pt;line-height:107%'>Requirements</span></h2>

<p class=MsoNormal style='margin-left:18.0pt'><b><i><u>Samples and Experimental
Design:</u></i></b> Data analysis needs a series of samples from a time course
labeling experiment where the carbon source is shifted from natural labeled
carbon source to <sup>13</sup>C labeled carbon source. Details about the design
of such experiments for different cell types can be found in the literature.</p>

<p class=MsoNormal style='margin-left:18.0pt'><b><i><u>LC-MS Data:</u></i></b> The
workflow was developed for high mass resolution spectra. Since peak grouping is
based on element specific m/z value differences of isotopic peaks it requires
high mass resolution spectra with mass accuracy of at least 0.008 for those
delta values. Quality of isotope analysis will increase with increasing
separation of mass isotopic peaks mass and thus with mass resolution. We tested
the approach on an Orbitrap instrument acquiring at R = 60000. However the tool
comprises a routine called ‘suitability test’ to evaluate whether the data set
is suited for the workflow. In general, best results will be obtained when
measuring the complete sample set in a single batch on a new column. DynaMet
accepts LC-MS data in mzXML and mzML format (peakmaps). If you do not have a converter
for MS data of your instrument you can find more information <span lang=DE-CH><a
href="http://en.wikipedia.org/wiki/Mass_spectrometry_data_format#cite_note-autogenerated1-26"><span
lang=EN-US>here</span></a></span>. <b>Before starting DynaMet create a new
project folder and paste the peakmaps of your data set into the folder. </b></p>

<span style='font-size:11.0pt;line-height:107%;font-family:"Calibri",sans-serif'><br
clear=all style='page-break-before:always'>
</span>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>&nbsp;</p>

<h1 style='margin-left:18.0pt;text-indent:-18.0pt'><span style='font-size:18.0pt;
line-height:107%'>3.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp; </span></span><span
style='font-size:18.0pt;line-height:107%'>Running DynaMet with eMZed 2</span></h1>

<p class=MsoNormal style='margin-left:18.0pt'><img border=0 width=605
height=369 id="Picture 56"
src="20141030DYNAMET%20DOCUMENTATION_files/image004.jpg"></p>

<p class=MsoNormal style='margin-left:18.0pt'>Start eMZed and type emzed.app.dynamet()
into command line of the iPython shell and press <i>Enter</i> . </p>

<p class=MsoNormal style='margin-left:18.0pt'><img border=0 width=316
height=33 id="Picture 39"
src="20141030DYNAMET%20DOCUMENTATION_files/image005.jpg"></p>

<p class=MsoNormal style='margin-left:18.0pt'>The DynaMet main window will pop
up.</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal style='margin-left:18.0pt'> <img border=0 width=270
height=287 id="Picture 65"
src="20141030DYNAMET%20DOCUMENTATION_files/image006.png"></p>

<p class=MsoNormal style='margin-left:18.0pt'>When analyzing a date set for the
first time create a folder containing your LC-MS data (peakmaps).  Select this
folder as project_path by pressing on the folder button (top, right).  When
starting the analysis subfolders are created and your peakmaps will be moved in
the PEAKMAPS subfolder. For any further analysis or for data inspection always choose
the project_path and <b>not </b>a subfolder.</p>

<p class=MsoNormal>&nbsp;</p>

<h2 style='margin-left:18.0pt;text-indent:-18.0pt'><span lang=FR-CH
style='font-size:18.0pt;line-height:107%'>4.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;
</span></span><span lang=FR-CH style='font-size:18.0pt;line-height:107%'>Configure
DynaMet Parameters:</span></h2>

<p class=MsoNormal style='margin-left:18.0pt'>To Configure the pipeline press
the  <i>change_pipline_config</i> button.</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=25 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=270 height=287
  src="20141030DYNAMET%20DOCUMENTATION_files/image007.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal><b><i><span style='font-size:12.0pt;line-height:107%'>Feature
detection</span></i></b></p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=11 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=216 height=199
  src="20141030DYNAMET%20DOCUMENTATION_files/image008.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal><i>Change_pipeline_config -&gt; feature_detection</i></p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=2 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=317 height=307
  src="20141030DYNAMET%20DOCUMENTATION_files/image009.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='text-indent:18.0pt'>Feature detection configuration
comprises for different parts. Only selected parts are displayed for parameter
modification.</p>

<p class=MsoListParagraphCxSpFirst style='text-indent:-18.0pt'><b>1.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></b> <b>Peakmap_processing
:</b></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;text-indent:-18.0pt'>a.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>ignore_blanks:
all sample files (peakmaps) containing label blank, 'Blank' or 'BLANK' in filename
(_label_) or at the end of filename (_label.) are ignored.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;text-indent:-18.0pt'>b.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>orbitrap_data:
Allows removing shoulder artefact peaks around intense peaks. Intense peaks in
mass spectra not only show the detected mass but a number of so-called
side-lobes (A. Kaufmann, P. Butcher, K. Maden, S. Walker, M. Widmer. Analyst
2011, 126, 1898.) or shoulder peaks. Those peaks significantly hamper analysis.
If you choose the Orbitrap data options those peaks are autamically removed
from spectra.</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt'><b>2.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span></b><b>Alignment:</b></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;text-indent:-18.0pt'>a.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span><b><i>retention
time:</i></b> uses emzed rtAlign module providing clustering based retention
time alignment algorithm developed by Lange (Lange E, et al. A geometric
approach for the alignment of liquid chromatography-mass spectrometry data.
Bioinformatics. 2007;23:I273–I281)</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><img border=0
width=350 height=187 id="Picture 20"
src="20141030DYNAMET%20DOCUMENTATION_files/image010.png"></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><b><i><span
style='font-size:12.0pt;line-height:107%'>Parameters:</span></i></b></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><i>rt_alignment</i>:
Retention time alignment will be performed if values is set to on 'on'. You can
switch it off by selecting 'off'.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;text-indent:34.2pt'><i>maxMzDifferencePairfinder</i>:
max allowed difference in m/z values for pair finding.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;text-indent:34.2pt'><i>mz_diff</i>:
max allowed difference in m/z values for super imposer.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;text-indent:34.2pt'><i>rt_diff</i>:
max allowed difference in rt values for searching matching features.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;text-indent:-18.0pt'>b.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span><b><i>mass
alignment:</i></b> Performs affine linear mz-correction for a feature table.
The approach is based on spiking mass calibrants into all samples to correct
for mass drifts occurring overtime. This table needs columns ``mz_hypot`` for
the m/z value calculated from the mass of the isotope, ``rtmin``, ``rtmax`` for
the retention time window where the peak is expected to elute from the column
in order to restrict the match of the table against the ``mz_reference_table`.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><img border=0
width=293 height=262 id="Picture 8"
src="20141030DYNAMET%20DOCUMENTATION_files/image011.png"></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><b><i>Parameters:</i></b></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><i>mz_alignment</i>:
Mz alignment will be performed if values is set to on 'on'. You can switch it
off by selecting 'off'.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><i>process_calibration_table</i>:
For mz calibration, a table for targeted extraction is needed. You can choose
between 'default', 'load different', 'inspect / modify', and 'build new'. </p>

<p class=MsoListParagraphCxSpLast style='margin-left:90.0pt;text-indent:-18.0pt'>-<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span>Default is the calibration table of current config. If you run the tool
for the first time, default table is as shown below. </p>

<p class=MsoNormal style='margin-left:70.8pt;text-align:justify'>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=96 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=531 height=484
  src="20141030DYNAMET%20DOCUMENTATION_files/image012.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='margin-left:70.8pt;text-align:justify'><b>Figure:</b>
Default mass calibration table.  A mass calibration table needs columns
``mz_hypot`` for the mz value calculated from the mass of the isotope,  ``rtmin``,
``rtmax`` for the retention time window where the peak is expected to elute
from the column in order to restrict the match of the table against the  ``mz_reference_table`.</p>

<p class=MsoListParagraphCxSpFirst style='margin-left:90.0pt'>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=122 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=341 height=142
  src="20141030DYNAMET%20DOCUMENTATION_files/image013.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt;text-indent:-18.0pt'>-<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span>'Load different' allows loading any table. When selecting a dialog
window occurs which allows selecting a mass calibration table. By default, an
empty table is provided. </p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>When pressing
the ' help' button requirements to mass calibration tables are specified: </p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'><img border=0
width=516 height=164 id="Picture 26"
src="20141030DYNAMET%20DOCUMENTATION_files/image014.png"></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>If you choose
the empty table, the tool continues with 'build_new'. Else it continues with
the inspect / modify part. This avoids choosing calibration tables, which
contains not the compounds spiked into the samples.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt;text-indent:-18.0pt'>-<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span>'inspect / modify': This allows you adapting retention time windows of
the calibration table to samples of your data set. Select a peakmap to check
for retention time windows by pressing the right bottom. </p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'><img border=0
width=422 height=113 id="Picture 27"
src="20141030DYNAMET%20DOCUMENTATION_files/image015.png"></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=95 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=576 height=360
  src="20141030DYNAMET%20DOCUMENTATION_files/image016.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>After choosing a
peakmap, an information window pops up specifying how to adapt retention time
windows within Tables.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'><img border=0
width=516 height=134 id="Picture 28"
src="20141030DYNAMET%20DOCUMENTATION_files/image017.png"></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>To continue
-&gt; press ok.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=120 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=494 height=319
  src="20141030DYNAMET%20DOCUMENTATION_files/image018.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'> <b>Figure:</b>
Calibration table with selected peakmap. </p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>After pressing
ok table explorer opens calibration table with selected peakmap and you can
inspect and modify retention times. We recommend removing all peaks from
calibration table where no match is found.  In addition, it is possible to
delete rows of the table or to add compounds by cloning rows. To this end right
click on button at the beginning of the row of interest and a window pops up.
Choose the 'Clone row' or 'Delete row'. The right click also allows undoing
manual manipulations.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>'build new':
This option allows building a new calibration table from scratch. If you build
a new calibration table we recommend to choose the compounds in a way that it
covers the complete m/z range measured and that number of calibrants fulfill
the minPoint criterion, e.g. if minPoint is 5 you have to define at least 5
peaks in your calibration table.  When selecting a window pops up afterwards remembering
minimal requirements to the calibration table.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'><img border=0
width=516 height=149 id="Picture 30"
src="20141030DYNAMET%20DOCUMENTATION_files/image019.png"></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>To continue
-&gt; press ok.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'><img border=0
width=462 height=310 id="Picture 31"
src="20141030DYNAMET%20DOCUMENTATION_files/image020.png"></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>An empty Table
pops up where you can enter the compound values manually. If the number of
provided rows is not sufficient, you can add additional rows by right clicking
the button at the beginning of each row. To continue simply close the table
(click on x in upper right corner). </p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'>Remark: rtmin
and rtmax values are calculated in seconds, but are automatically displayed in
minute format. Thus if you enter 60 in a rtmin cell and press enter, the value
1.00m is displayed.  You can enter the same value in minutes directly by typing
1.0m.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:90.0pt'><i>To continue
-&gt; close the table. </i></p>

<p class=MsoListParagraphCxSpLast style='margin-left:90.0pt'>Further steps are
now as describes under 'modify / inspect' (see above).</p>

<p class=MsoNormal>                               <i>mztol:</i> maximal
tolerated mass difference to match a peak with calibrant (units)</p>

<p class=MsoNormal>                               <i>minR2</i>: Stop criterion
when removing outlier points. Values are in Range [0 - 1]. </p>

<p class=MsoNormal style='margin-left:18.0pt'>                       minPoints:
Minimal number of points for calibration curve fitting. You need at least as
many peaks in your calibration table!</p>

<p class=MsoNormal style='margin-left:18.0pt'>                       <i>maxtol</i>:
Maximal tolerated mass deviation after calibration. Stop criterion when removing
outlier points.</p>

<p class=MsoNormal style='margin-left:70.5pt'><i>interactive</i>: if 'True'
manual inspection and data point removal is enabled. For automatic data processing
this is not recommended!</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoListParagraphCxSpFirst style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt'>3.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span><b>Peak_detection:</b>
Extracted ion chromatogram peaks (EIC-peaks) are detected using openMS feature
finder metabo (Kenart E., et al. Molecular &amp; Cellular Proteomics. 2014; 13:
348-359.)</p>

<p class=MsoListParagraphCxSpMiddle><b>&nbsp;</b></p>

<p class=MsoListParagraphCxSpMiddle>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=97 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=384 height=287
  src="20141030DYNAMET%20DOCUMENTATION_files/image021.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:18.0pt'><b>TIP:</b> You
obtain direct help for a selected parameter when remaining with mouse pointer
on parameter for several seconds.</p>

<p class=MsoListParagraphCxSpMiddle>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;text-indent:-18.0pt'>a.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span>Default
setting display a reduced number of parameter settings that allow adapting the
finder to your LC-MS system</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><img border=0
width=329 height=236 id="Picture 19"
src="20141030DYNAMET%20DOCUMENTATION_files/image022.png"></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><i>common_noise_threshold_int</i>:
intensity threshold below which peaks are regarded</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'> as noise. For
given example all peaks with max intensity below 1000 counts are ignored.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><i>common_chrom_peak_snr</i>: 
minimum signal-to-noise a mass trace should have. For instance the definintion
of LOD is snr&gt;3.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><i>common_chrom_fwhm</i>: 
typical peak width (full width at half maximum).</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><i>mtd_mass_error_ppm:
</i>Allowed mass deviation. The value should not be confused </p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>with
mass_accuracy. It corresponds to the boundaries of mz values (mzmin detected,
mz mass detected) observed for exctracted ion chromatographic peaks. </p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><i>mtd_reestimate_mt_sd</i>:
 enables dynamic re-estimatation of m/z variance during mass trace collection
state. It is recommended to use this parameter.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><i>epdet_width_filtering</i>:
enable filtering of unlikely peaks width.  If 'on' tool filters peaks with the
5% and 95% quantiles of the peak width distribution. </p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt;text-indent:-18.0pt'>b.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span><b>This
option is only suited for experts which are familiar with the peak detection
approach.</b> Besides parameters mentioned under a further parameters are
provided for modification.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'><img border=0
width=331 height=441 id="Picture 22"
src="20141030DYNAMET%20DOCUMENTATION_files/image023.png"></p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'><i>mtd_trace_terminaton_criterion</i>:
Termination criterion for the extension of mass traces. In `outlier` mode, trace
extension cancels if a pre-defined number of consecutive outliers are found (see
<i>trace_termination_outliers</i> parameter). In 'sample_rate` mode, trace
extension in both direction stops if ratio of found peaks versus visited
spectra falls below `<i>min_sample_rate</i>` threshold.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'><i>mtd_trace_termination_ouliers</i>: 
mass trace extension in one direction cancels if set value of consecutive
spectra without detected peaks is reached.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'><i>mtd_min_sample_rate</i>: 
Minimum fraction of scans along the mass trace that must contain a peak.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'><i>mtd_min_trace_length:
</i>Minimum expected length of a mass trace (in seconds).</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'><i>mtd_max_trace_length:</i>
Maximum expected length of a mass trace (in seconds).</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'><i>epdet_width_filtering:
</i>Same as described for default settings except that<i> </i>in addition to
'auto' mode (called on in default) a fixes mode is also possible.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'><i>epdet_min_fwhm:</i>
Minimum full-width-at-half-maximum of chromatographic peak (in seconds). Ignored
if parameter <i>epd_width_filtering</i> is off or auto.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'><i>epdet_min_fwhm:</i>
Maximum full-width-at-half-maximum of chromatographic peak (in seconds). Ignored
if parameter <i>epd_width_filtering</i> is off or auto.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:70.8pt'><i>epdet_masstrace_snr_filtering:</i>
apply post-filtering by signal-to-noise ratio after smoothing.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:72.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpMiddle style='text-indent:-18.0pt'>4.<span
style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </span><b>Feature_grouping:</b>
a graph based feature grouper for high-resolution MS data. Feauture_grouping is
a core development of the pipeline since it is a key step to detect features
with significant labeling incorporation.  </p>

<p class=MsoListParagraphCxSpMiddle>&nbsp;</p>

<p class=MsoListParagraphCxSpLast style='margin-left:72.0pt'>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=76 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=301 height=213
  src="20141030DYNAMET%20DOCUMENTATION_files/image024.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='margin-left:70.8pt'><i>isolation width</i>: Maximal m/z
difference between mass traces  of different samples tolerated as same m/z
value (value in U).</p>

<p class=MsoNormal style='margin-left:70.8pt'><i>Charge_lower_bond</i>: Minimal
charge state z to consider.</p>

<p class=MsoNormal style='margin-left:70.8pt'><i>Charge_upper_bond</i>: Maximal
charge state z to consider.</p>

<p class=MsoNormal style='margin-left:70.8pt'><i>max_c_gap</i>: Maximal allowed
carbon gap width n. Here, n corresponds to a multitude of the mass difference
of <sup>13</sup>C and <sup>12</sup>C times the charge state of the feature.
Depending on chosen labeling strategy the maximal distance correspond to the
number of carbon atoms of a metabolite. In case of one-carbon compound as sole
carbon source max_c_gap is one for most core metabolites and corresponds to the
size of precursors originating from central metabolism; e.g. biosynthesis of
ATP requires ribose-5-phosphate which is fully labeled before significant label
accumulates in ATP leading to a max_c_gap of 5 for ATP. However, the higher
max_v_gap the higher the probability of mismatches.</p>

<p class=MsoNormal style='margin-left:70.8pt'><i>rel_min_area</i>: Lowest peak
area relative to the base peak area accepted. </p>

<p class=MsoNormal><b><i><span lang=FR-CH style='font-size:12.0pt;line-height:
107%'>&nbsp;</span></i></b></p>

<p class=MsoNormal><b><i><span lang=FR-CH style='font-size:12.0pt;line-height:
107%'>5.Identification</span></i></b></p>

<p class=MsoNormal style='text-indent:35.4pt'><i><span lang=FR-CH>Change_pipeline_config
-&gt; identification</span></i></p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=41 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=216 height=199
  src="20141030DYNAMET%20DOCUMENTATION_files/image025.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=41 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=288 height=210
  src="20141030DYNAMET%20DOCUMENTATION_files/image026.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i><span style='font-size:
12.0pt;line-height:107%'>Parameters</span></i></b></p>

<p class=MsoNormal style='margin-left:35.4pt'><i>instr_linear_error</i>: Absolute
instrumental linear error on area measurement    'e.g. LTQ-Orbitrap classic instrument
0.03 (3 %).</p>

<p class=MsoNormal style='margin-left:35.4pt'><i>idms_sample</i>: Select
whether data set contains an isotope dilution sample, which is a sample
composed of 1 : 1 mixture of a natural labeled cell extract, and a cell extract
from cells cultivated on [U-13C] labeled carbon source for at least 5
generations. Ideally, carbon source is 99% [U-13C] and thus all metabolites are
labeled as carbon source. In addition, both cell extracts originate from the
same strain or cell line cultivated applying same growth condition.<i> </i></p>

<p class=MsoNormal style='margin-left:35.4pt'><i>c_source_labeling</i>: Enter
value for labeled fraction of applied substrate for dynamic labeling
experiment, e.g. 0.99. The correct value is important for the estimation of the
number of carbon atoms. </p>

<p class=MsoNormal style='margin-left:35.4pt'><i>data base: </i>Choose an emzed
integrated data base, or use your own data base. Assure that your data base is
compatible with pipeline. Dynemet provides emzed integrates data bases KEGG,
Pubchem, and human metabolome data base. In case you are choosing 'other' a
Dialog opens which allows for loading your own data base:</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0>
 <tr>
  <td width=47 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=257 height=113
  src="20141030DYNAMET%20DOCUMENTATION_files/image027.png"></td>
 </tr>
</table>

<br clear=ALL>
                </p>

<p class=MsoNormal>                After pressing the file select button a dialog
box opens</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=42 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=555 height=347
  src="20141030DYNAMET%20DOCUMENTATION_files/image028.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='margin-left:35.4pt'>Database formats table and csv
are accepted. The column names 'mf' containg compounds  molecular formula and
'm0' containing corresponding monoisotopic mass are mandatory. If you load one
data base in csv format the file requires column name is in the first row of
your database. Column names have to be unique (don't forget to switch file type
filter to  *.csv). Currently, identification is only based on exact m/z values
combined with possible adducts to assign a data base entry to a feature. In
case features could be grouped by assigned adducts corresponding mass value is
used for database assigning. </p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoListParagraph style='text-indent:-18.0pt'><b><i><span
style='font-size:12.0pt;line-height:107%'>5.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></i></b><b><i><span style='font-size:12.0pt;line-height:107%'>Feature_analyis</span></i></b></p>

<p class=MsoNormal style='text-indent:18.0pt'><i>Change_pipeline_config -&gt;
data_analysis</i></p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=17 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=216 height=199
  src="20141030DYNAMET%20DOCUMENTATION_files/image029.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=17 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=268 height=161
  src="20141030DYNAMET%20DOCUMENTATION_files/image030.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='text-indent:18.0pt'>Data analysis parameters define
minimal quality criteria for feature selection. </p>

<p class=MsoNormal style='text-indent:18.0pt'><b><i><span style='font-size:
12.0pt;line-height:107%'>Parameters</span></i></b></p>

<p class=MsoNormal style='margin-left:18.0pt'><i>min_labeling</i>: threshold
value for feature selection: Minimal labeled carbon fraction of one feature
measured in at least one sample out of complete set of samples. If not
fulfilled feature will be excluded</p>

<p class=MsoNormal style='margin-left:18.0pt'><i>feature_frequency</i>: threshold
value for feature selection: frequency of feature occurrence in sample set.
Example 0.5 : feature is detected in at least every second sample. If not
fulfilled, feature will be excluded</p>

<p class=MsoNormal style='margin-left:18.0pt'><i>max_nrmse</i>: upper limit
value for feature selection: Maximal normalized root-mean-square-error accepted
for labeling profile fitting. If not fulfilled, no feature labeling curve
analysis</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoListParagraph style='text-indent:-18.0pt'><b><i><span
style='font-size:12.0pt;line-height:107%'>6.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
</span></span></i></b><b><i><span style='font-size:12.0pt;line-height:107%'>Suitablity_test</span></i></b></p>

<p class=MsoNormal style='text-indent:18.0pt'><i>Change_pipeline_config -&gt;
suitability_test</i></p>

<p class=MsoNormal style='text-indent:18.0pt'>Configuration and application are
described below.</p>

<p class=MsoNormal>&nbsp;</p>

<h2 style='margin-left:18.0pt;text-indent:-18.0pt'><span style='font-size:18.0pt;
line-height:107%'>5.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp; </span></span><span
style='font-size:18.0pt;line-height:107%'>Suitablility Test:</span></h2>

<p class=MsoListParagraph>&nbsp;</p>

<p class=MsoNormal style='margin-left:18.0pt'>Dynamet's feature grouping
algorithm relies on high mass accuracy data. Based on accurate m/z distances of
mass isotopologues, feature grouper proposes possible elements (CHNOPS) that
explain measured mz distances. Currently the grouper has only been tested in
the negative ESI mode. Suitability test verifies for a set of compounds whether
data fulfills requirements or not. The tool provides a default set of compounds
(features) but data can also be tested for any user-defined compounds.
Compounds must be natural labeled and at least two mass isotopologues for each
compound should be measured in the sample. We recommend testing at least 10
different features distributed over complete m/z range measured.</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal style='margin-left:18.0pt'><b><span style='font-size:14.0pt;
line-height:107%'>Configure suitability test</span></b><span style='font-size:
14.0pt;line-height:107%'> </span></p>

<p class=MsoNormal style='margin-left:18.0pt'>To modify the setting click '<i>change_pipline_config</i>'
-&gt; '<i>suitablility_test</i>'</p>

<p class=MsoListParagraph>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=24 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=286 height=158
  src="20141030DYNAMET%20DOCUMENTATION_files/image031.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal style='margin-left:18.0pt'>The test takes into account the
mass resolution of your instrument.  Enter the mass Redolution R you applied
for the measuring your samples. In case of Orbitrap instruments resolution R is
calculated for each m/z value from unit Resolution. To do so the 'R at mz'
parameter is required. For Instrument of the QExactive series R is defined at
m/z 200 whereas in case of all other Orbitrap instruments it is defined at m/z
400. The parameter is ignored if field <i>'Orbitrap_instrument</i>' is not
selected.</p>

<p class=MsoNormal><b>&nbsp;</b></p>

<p class=MsoNormal style='text-indent:18.0pt'><b>Run suitability test:</b></p>

<h3 style='text-indent:18.0pt'>CHOOSE TEST_TABLE</h3>

<p class=MsoNormal style='text-indent:18.0pt'>To run the test click '<i>suitability
test</i>' in the main window</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=19 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=216 height=109
  src="20141030DYNAMET%20DOCUMENTATION_files/image032.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal style='margin-left:18.0pt'>As test table you can use the
current table of pipeline configuration (the default is build from compounds of
mass calibration table), or load a table from folder (see example below). If
you load your own table required tale columns are:</p>

<p class=MsoNormal style='margin-left:18.0pt'> <i>id</i>: unique identifier; <i>name</i>:
compound name (int);  </p>

<p class=MsoNormal style='margin-left:18.0pt'><i>mf</i> : molecular formula
(string);</p>

<p class=MsoNormal style='margin-left:18.0pt'><i>mf_ion</i>: molecular formula
corrected for adduct, example in case of M-H a H is substracted from formula
(string); </p>

<p class=MsoNormal style='margin-left:18.0pt'><i>rtmin, rtmax</i>: bounderies
of retention time window within peak elution is expected (float); </p>

<p class=MsoNormal style='margin-left:18.0pt'><i>adduct_name</i>: name of
adduct (for adduct assignment the Metlin nomenclature is applied: [M-H]<sup>-</sup>
-&gt; M-H) (string); </p>

<p class=MsoNormal style='margin-left:18.0pt'><i>polarity</i>: applied ESI
mode: +/- (string);</p>

<p class=MsoNormal style='margin-left:18.0pt'>z: charge state of ion (z&gt;=1,
int)</p>

<p class=MsoNormal style='margin-left:18.0pt'>&nbsp;</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0>
 <tr>
  <td width=17 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=605 height=260
  src="20141030DYNAMET%20DOCUMENTATION_files/image033.jpg"></td>
 </tr>
</table>

<br clear=ALL>
<b>    </b></p>

<p class=MsoNormal style='text-indent:18.0pt'><b>Build a new test table from
scratch</b></p>

<p class=MsoNormal style='text-indent:18.0pt'>Choose <i>build new</i> and press
ok. </p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=22 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=462 height=310
  src="20141030DYNAMET%20DOCUMENTATION_files/image034.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='margin-left:18.0pt'>An empty Table pops up where you
can enter the compound values manually. If the number of provided rows is not
sufficient, you can add additional rows by right clicking the button at the
beginning of each row. To continue simply close the table (click on x in upper
right corner). </p>

<p class=MsoNormal style='margin-left:18.0pt'><b>Remark:</b> rtmin and rtmax
values are calculated in seconds, but are automatically displayed in minute format.
Thus if you enter <i>60</i> in a rtmin cell and press enter, the value <i>1.00m</i>
is displayed.  You can enter the same value in minutes directly by typing <i>1.0m.</i>
</p>

<p class=MsoNormal style='margin-left:18.0pt'>Next for each compound a window
pops up that allows choosing adduct observed for compound. If you want to use
more than adduct for the same compound you have to enter it twice in table with
different id. </p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=25 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=216 height=109
  src="20141030DYNAMET%20DOCUMENTATION_files/image035.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<h2>&nbsp;</h2>

<h3 style='text-indent:18.0pt'>Choose peakmap (sample) from data set</h3>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=20 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=285 height=136
  src="20141030DYNAMET%20DOCUMENTATION_files/image036.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='margin-left:18.0pt'>In the next step, all peaks
specified in test-table are extracted from sample file. Choose a file
containing all or at least most of compounds. When the <i>inspect peaks</i> is
selected, a table with peak extraction results will pop up. We recommend to
always check peak extraction since it is possible to adapt retention time     windows
manually. A pop up window will remind you that. </p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=34 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=516 height=134
  src="20141030DYNAMET%20DOCUMENTATION_files/image037.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=36 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=605 height=247
  src="20141030DYNAMET%20DOCUMENTATION_files/image038.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal><b>&nbsp;</b></p>

<p class=MsoNormal style='margin-left:18.0pt'>To modify the retention time
window left click on the boundary dot and keep the mouse button pressed. You
can now move the window boundary to the left or right. As soon as the window is
placed correctly, press on the integrate button and rtmin and rtmax will be set
to boundary values. All peaks that cannot be detected are removed automatically:
</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=42 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=353 height=149
  src="20141030DYNAMET%20DOCUMENTATION_files/image039.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal style='text-indent:18.0pt'><b>Accurate peak selection is
mandatory for reliable results !!</b></p>

<p class=MsoNormal style='text-indent:18.0pt'><b>TIP: </b>you can also add or remove
ions at that level by cloning or deleting rows. To add the new compound, enter
mzmin, and mzmax values of the ion, select the peak with the integration
bounderies, reintegrate it and, correct all other values ( adduct_name, mf,….).</p>

<p class=MsoNormal>&nbsp;</p>

<h3 style='text-indent:18.0pt'>Start test</h3>

<p class=MsoNormal style='text-indent:18.0pt'>The test will start automatically.
A pop up window will inform you about the results</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=36 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=461 height=149
  src="20141030DYNAMET%20DOCUMENTATION_files/image040.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>In addition details of the test result will be displayed in
emzed iPython console. </p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0>
 <tr>
  <td width=23 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=568 height=305
  src="20141030DYNAMET%20DOCUMENTATION_files/image041.png"></td>
 </tr>
</table>

<br clear=ALL>
                </p>

<p class=MsoNormal>&nbsp;</p>

<h2 style='margin-left:18.0pt;text-indent:-18.0pt'><span style='font-size:18.0pt;
line-height:107%'>6.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp; </span></span><span
style='font-size:18.0pt;line-height:107%'>Run Analysis:</span></h2>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal style='text-indent:18.0pt'>Remark: To get comfortable with
DynaMet the package also contains a reduced example data set ‘Dynamet_test_example.zip’
comprising 9 mzXML files originating from a dynamic 13C labeling incorporation
experiment of <i>Bacillus methanolicus</i>  MGA3 grown on methanol.  </p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=27 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=270 height=287
  src="20141030DYNAMET%20DOCUMENTATION_files/image042.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='margin-left:18.0pt'>Before starting analysis make
sure that (i) the correct project path is selected, (ii) sample files are of
type mzXML or mzML and belong to a time series of a dynamic labeling
experiment, and (iii) all parameters are chosen correctly. If you are not sure whether
your LC-MS data fulfill the minimal criteria, perform a suitability test first.
To start the workflow, press the button 'run analysis' in the main window. When
running the analysis for the first time workflow will request additional data. 
</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=27 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=292 height=346
  src="20141030DYNAMET%20DOCUMENTATION_files/image043.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='text-indent:18.0pt'>From the data LC-MS data files
(peakmaps) a list is creacted and you can select those samples belonging to the
dynamic labeling experiment. </p>

<p class=MsoNormal style='text-indent:18.0pt'><i>To continue -&gt; press ok</i></p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=19 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=281 height=134
  src="20141030DYNAMET%20DOCUMENTATION_files/image044.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='text-indent:18.0pt'>Next, a window pops up. After
pressing 'ok' a table pops up which allows defining sample order and sampling
time points of each sample. Time points are calculated in seconds. Default time
values are 0.0s, default sample order is derived from sorted sample file names
(ascending)</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=23 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=298 height=326
  src="20141030DYNAMET%20DOCUMENTATION_files/image045.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='text-indent:18.0pt'>For given example, time points
are: </p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=19 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=297 height=325
  src="20141030DYNAMET%20DOCUMENTATION_files/image046.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='text-indent:18.0pt'>After finishing, close the table.</p>

<p class=MsoNormal style='text-indent:18.0pt'>If you chose the option <i>Change_pipeline_config
-&gt; identification -&gt; idms_sample, </i>a further selection windows pops up
which allows choosing an idms sample (mixture of natural labeled and uniformly <sup>13</sup>C
labeled cell extract).</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=21 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=236 height=203
  src="20141030DYNAMET%20DOCUMENTATION_files/image047.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='text-indent:18.0pt'>To start DATA ANALYSIS press ok.</p>

<p class=MsoNormal style='text-indent:18.0pt'>When analysis is done a window
pops up informing about end of process.</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=20 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=137 height=130
  src="20141030DYNAMET%20DOCUMENTATION_files/image048.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='margin-left:18.0pt'><b>Remarks:</b> DynaMet uses
'pacer' a lightweight Python package for implementing distributed data
processing workflows. In general, it manages, enhance and accelerate data
analysis. For details see <span lang=DE-CH><a
href="https://pypi.python.org/pypi/pacer"><span lang=EN-US>here</span></a></span>.
Since all intermediate results from processing steps are cached, pacer is able
to determine which steps are concerned by parameter change and re-running the
analysis only executes those steps. </p>

<p class=MsoNormal>&nbsp;</p>

<h2 style='margin-left:18.0pt;text-indent:-18.0pt'><span style='font-size:18.0pt;
line-height:107%'>7.<span style='font:7.0pt "Times New Roman"'>&nbsp;&nbsp; </span></span><span
style='font-size:18.0pt;line-height:107%'>Inspecting results</span></h2>

<p class=MsoListParagraphCxSpFirst style='margin-left:18.0pt'>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=52 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=630 height=472
  src="20141030DYNAMET%20DOCUMENTATION_files/image049.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:18.0pt;text-indent:17.4pt'><b>Figure:</b>
Presentation of results. All results are combined in a single table including
plots, identification results or details of selected features.</p>

<p class=MsoListParagraphCxSpMiddle style='margin-left:18.0pt'>&nbsp;</p>

<p class=MsoListParagraphCxSpLast style='margin-left:35.4pt'>Dynamet creates an
explorable emzed Table object for all extracted features (see Figure). Plots
and subtables can be opened by double clicking on the cell. Identification
results and feature details are provided as feature-wise subtables. If you use implemented
data base e.g. Kegg identification results contain column with direct link of of
assigned compound to pubchem data base. Plots show mass isotopologue 
distribution heatmap and if possible a fit curve of labeling incorporation into
metabolite pools.</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=94 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=270 height=287
  src="20141030DYNAMET%20DOCUMENTATION_files/image050.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='text-indent:35.4pt'>To inspect results press the
inspect_result button. </p>

<p class=MsoNormal style='text-indent:35.4pt'>

<table cellpadding=0 cellspacing=0>
 <tr>
  <td width=51 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=216 height=139
  src="20141030DYNAMET%20DOCUMENTATION_files/image051.png"></td>
 </tr>
</table>

<br clear=ALL>
A dialog box opens which allows you to select the latest result table (current)
or to choose a result table from all result tables in the project RESULT folder
(DynaMet automatically adds date and time to the result file name with while
saving. Older results are not overwritten).  </p>

<p class=MsoNormal style='text-indent:35.4pt'>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=56 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=216 height=139
  src="20141030DYNAMET%20DOCUMENTATION_files/image052.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='text-indent:35.4pt'>When Choosing <b>all </b>a box
opens that allows you select result file out of all present in the result
folder. </p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=56 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=316 height=274
  src="20141030DYNAMET%20DOCUMENTATION_files/image053.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='margin-left:35.4pt'> <b>Explaining the result table
columns:</b></p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>feature_id</i></b>: Unique
identifier number for each feature.</p>

<p class=MsoNormal style='margin-left:35.4pt'><b>adduct_group</b>: Co-eluting
features with m/z value differences which can be explained by different adducts
of the same compound are  grouped.</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>z</i></b>: feature charge
state.</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>rt</i></b>: feature
retention time in minutes.</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>mz0</i></b><i>:</i> Mass of
monoisotopic peak. Value is only assigned if feature was detected in natural
labeled sample.</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>min_mz</i></b>: Lowest m/z
value of one feature out of all samples .</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>possible_m0</i></b>: value
calculated if adduct assignment was possible and m/z 0 was determined (see
above).</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>num_c</i></b>: Number of
estimated carbon atoms (for details see <i>details</i> subtable).</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>flcluster_id</i></b>:
Features are clustered by fitting parameters t50 and std_c_13_fraction .
Grouped features have the same <i>flcuster_id</i></p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>label_t50_sec</i></b>: Time
constant T<sub>50</sub> (time required to reach 50 % of carbon atoms being
labeled) resulting from parameters of first order fitting curves.</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>std_label_t50_sec</i></b>:
Standard deviation of estimated time constant T<sub>50</sub>.</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>std_c13_fraction_calc</i></b>:
gain k of first order fitting curve.</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>pool_t50_sec</i></b>: Time
constant T<sub>50</sub> (half life time) of metabolite pool based on resulting
from parameters of first order fitting curve of isotopologue M0 depletion.</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>std_pool_t50_sec</i></b>:
Standard deviation of estimated pool half life T<sub>50</sub>.</p>

<p class=MsoNormal style='text-indent:35.4pt'><b><i>nrmse</i></b>: Normalized
root mean square error of fitting curve.</p>

<p class=MsoNormal style='text-indent:35.4pt'><b><i>fit_model</i></b>: Two first
order fitting models are applied by workflow: logistic and pt1. </p>

<p class=MsoNormal style='text-indent:35.4pt'><b><i>dli_label_plots</i></b>:
Dynamic labeling incorporation plots. </p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=50 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=605 height=259
  src="20141030DYNAMET%20DOCUMENTATION_files/image054.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal style='margin-left:35.4pt'>For each feature a heatmap is
provided showing distribution of mass isotopologue abundances corrected for
natural labeling. M<sub>i</sub> correspondsd to i<sup>th</sup> mass
isotopologue and S<sub>j</sub> to the j<sup>th</sup> sample. If fitting was
possible, a fitting curve is provided showing the number of incorporated
labeled carbon atoms over time. </p>

<p class=MsoNormal style='text-indent:35.4pt'><b><i><span lang=FR-CH>M0_dilution_plots</span></i></b><span
lang=FR-CH>: Monisotopic isotopologue dilution plots. </span></p>

<p class=MsoNormal><span lang=FR-CH>                </span><img border=0
width=418 height=358 id="Picture 62"
src="20141030DYNAMET%20DOCUMENTATION_files/image055.png"></p>

<p class=MsoNormal style='margin-left:35.4pt'>If fitting was possible for a
feature a fitting curve is provided showing quaktiy of M0 depletion fit.</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal style='text-indent:35.4pt'><b><i>feature_clustering_plot</i></b>:
</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=42 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=536 height=402
  src="20141030DYNAMET%20DOCUMENTATION_files/image056.png"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>                                Feature clustering plot
provides an overview of grouped features.</p>

<p class=MsoNormal style='text-indent:35.4pt'><b><i>details</i></b>: Subtable
with detailed information of one feature:</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=62 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=605 height=369
  src="20141030DYNAMET%20DOCUMENTATION_files/image057.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>                Details subtable provides feature-wise 
detailed results. The columns are:</p>

<p class=MsoNormal style='margin-bottom:0cm;margin-bottom:.0001pt;line-height:
normal'><b>                </b><i><u>adduct_group</u></i>: Co-eluting features
with m/z value differences which can be explained by </p>

<p class=MsoNormal style='margin-bottom:0cm;margin-bottom:.0001pt;line-height:
normal'>                different adducts of the same compound are  grouped.</p>

<p class=MsoNormal style='margin-bottom:0cm;margin-bottom:.0001pt;line-height:
normal'>&nbsp;</p>

<p class=MsoNormal>                <i><u>possible_adducts</u></i>: Assigned
adduct e.g. M-H, M-2H</p>

<p class=MsoNormal>                <i><u>feature_id</u></i>: Unique feature
identification number.</p>

<p class=MsoNormal>                <i><u>mz</u></i>: Measured m/z value of mass
peak.</p>

<p class=MsoNormal>                <i><u>mzmin</u>:</i> Lower boundary of
feature peak mass trace .</p>

<p class=MsoNormal>                <i><u>mzmax</u>:</i> Upper boundary of
feature peak mass trace</p>

<p class=MsoNormal>                <i><u>z</u></i>: charge state</p>

<p class=MsoNormal>                <i><u>possible_m0</u></i>: value calculated if
adduct assignment was possible and m/z 0 was determined</p>

<p class=MsoNormal>                <i><u>rt</u></i>: feature retention time in
minutes</p>

<p class=MsoNormal>                <i><u>rtmin</u></i>: Lower retention time
window boundary</p>

<p class=MsoNormal>                <i><u>rtmax</u></i>: Upper retention time
window boundary</p>

<p class=MsoNormal>                <i><u>fwhm</u></i>: Full width half maximum
of peak</p>

<p class=MsoNormal style='margin-left:35.4pt'><i><u>method</u></i>: Peak
integration method; emzed provides a number of different integration </p>

<p class=MsoNormal style='margin-left:35.4pt'>algorithms by default the tool
applies exponential modified gauss EMG integrator (emg_exact) and trapez.</p>

<p class=MsoNormal>                <i><u>area</u></i>: Peak area counts *s</p>

<p class=MsoNormal>                <i><u>rmse</u></i>: root mean square error
of peak area detect</p>

<p class=MsoNormal>                <i><u>peakmap</u></i>: LC-MS underlying
sample peapmap containing all measured spectra. Double </p>

<p class=MsoNormal>                clicking  opens peakmap explorer:</p>

<p class=MsoNormal>                <img border=0 width=407 height=311
id="Picture 53" src="20141030DYNAMET%20DOCUMENTATION_files/image058.jpg"></p>

<p class=MsoNormal>                Details about how to work with peakmaps can
be found <span lang=DE-CH><a
href="http://emzed.ethz.ch/tour.html#working-with-peak-maps"><span lang=EN-US>here</span></a></span>.</p>

<p class=MsoNormal>                <i><u>source</u></i>: File name of the
peakmap.</p>

<p class=MsoNormal style='text-indent:35.4pt'><i><u>mz0</u></i>: m/z value of
monoisotopic peak</p>

<p class=MsoNormal>                <i><u>min_mz</u></i>: Lowest m/z value of
one feature out of all samples </p>

<p class=MsoNormal style='margin-left:35.4pt'>

<table cellpadding=0 cellspacing=0>
 <tr>
  <td width=180 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=234 height=178
  src="20141030DYNAMET%20DOCUMENTATION_files/image059.jpg"></td>
 </tr>
</table>

<br clear=ALL>
<i><u>time</u></i>: Sampling time point. Tipp if you want to display an overlay
of all feature peaks of one sample choose ‘expand selection by: time’ and click
on one row to select all peaks of the same time.</p>

<p class=MsoNormal>                <i><u>order</u></i>: Sample order as entered
by the user</p>

<p class=MsoNormal>                <i><u>num_isotopes</u></i>: number i of mass
isotopologue Mi</p>

<p class=MsoNormal style='text-indent:35.4pt'><i><u>num_c</u></i>: Number of
estimated carbon atoms of compound (for details see <i><u>q_score</u></i>). </p>

<p class=MsoNormal>                <i><u>min_num_c</u></i>: Lower boundary of
carbon atoms estimation</p>

<p class=MsoNormal>                <i><u>max_num_c</u></i>: Upper boundary of
carbon atoms estimation</p>

<p class=MsoNormal style='margin-left:35.4pt'><i><u>q_score</u></i>: Quality
scoring value of carbon number estimation. Up to four different approaches are
used and combined estimate the number of carbon atoms see below </p>

<p class=MsoNormal style='margin-left:35.4pt'><i><u>origin_of_c_estimation</u></i>:
(1) <b>by_nl</b>: From mass isotopologue distribution of natural labeled
compound. (2) <b>by_idms:</b> If an idms sample is in the data set, the m/z
distance between monoisotopic mass peak M<sub>0</sub> and uniformly <sup>13</sup>C
labeled peak M<sub>UL</sub> is used (by_idms). The fraction of <sup>13</sup>C
in labeled carbon source is taken into account. The approach is most reliable
for substrates with uniformly <sup>13</sup>C labeled fraction &gt; 95%. (3) <b>by_dli:</b>
The number of highest Mi detected defines the lower limit <i><u>min_num_c</u></i>.
(4)  Finally, global carbon atoms bounderies are estimated from provided database
by matching molecular formulas within a mass range of 25 units. If no database was
provided Pubchem database is used pubchem db</p>

<p class=MsoNormal style='text-indent:35.4pt'><i><u>mi_fraction</u></i>: Area
fraction of i<sup>th</sup> mass isotopologue of total feature area. </p>

<p class=MsoNormal style='text-indent:35.4pt'><i><u>mi_frac_corr</u></i>:
mi_fraction corrected for natural carbon isotope abundance.</p>

<p class=MsoNormal style='text-indent:35.4pt'><i><u>no_C13</u></i>: Number of
labeled carbon atoms detected in feature at time x. </p>

<p class=MsoNormal style='text-indent:35.4pt'><i><u>C13_fraction</u></i>:
Fraction of 13C in compound at time x.</p>

<span style='font-size:11.0pt;line-height:107%;font-family:"Calibri",sans-serif'><br
clear=all style='page-break-before:always'>
</span>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal style='text-indent:35.4pt'>&nbsp;</p>

<p class=MsoNormal style='margin-left:35.4pt'><b><i>identification_results</i></b>:
</p>

<p class=MsoNormal style='margin-left:35.4pt'>A sub-table containing most
feature information joined with database. In case data bases provided by emzed
were selected  identification table provides urls to Pubchem DB. You can click
on the link to open corresponding Pubchem entry. </p>

<p class=MsoNormal>

<table cellpadding=0 cellspacing=0 align=left>
 <tr>
  <td width=47 height=0></td>
 </tr>
 <tr>
  <td></td>
  <td><img width=605 height=141
  src="20141030DYNAMET%20DOCUMENTATION_files/image060.jpg"></td>
 </tr>
</table>

<br clear=ALL>
</p>

<p class=MsoNormal>&nbsp;</p>

<p class=MsoNormal>&nbsp;</p>

</div>

</body>

</html>
